// @spartan-ng/cli/src/generators/base/lib/styles/transform-style-map.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformStyleMap = void 0;
const tailwind_merge_1 = require("tailwind-merge");
const ts_morph_1 = require("ts-morph");
const ALLOWLIST = new Set(['spartan-menu-target', 'spartan-logical-sides', 'spartan-rtl-flip', 'spartan-invalid']);
function isStringLiteralLike(node) {
    return ts_morph_1.Node.isStringLiteral(node) || ts_morph_1.Node.isNoSubstitutionTemplateLiteral(node);
}
const transformStyleMap = async ({ sourceFile, styleMap }) => {
    const matchedClasses = new Set();
    applyToHlmCalls(sourceFile, styleMap, matchedClasses);
    applyToClassesCalls(sourceFile, styleMap, matchedClasses);
    applyToHtmlInStrings(sourceFile, styleMap, matchedClasses);
    applyToRawHtml(sourceFile, styleMap, matchedClasses);
    applyToCvaCalls(sourceFile, styleMap, matchedClasses);
    return sourceFile;
};
exports.transformStyleMap = transformStyleMap;
function applyToRawHtml(sourceFile, styleMap, matchedClasses) {
    const text = sourceFile.getFullText();
    if (!text.includes('class="'))
        return;
    const updated = text.replace(/class="([^"]+)"/g, (_, classValue) => {
        const classes = extractSpartanClasses(classValue);
        if (classes.length === 0)
            return `class="${classValue}"`;
        const unmatched = classes.filter((c) => !matchedClasses.has(c));
        const tailwindClasses = unmatched.map((c) => styleMap[c]).filter((v) => Boolean(v));
        let updatedClass = classValue;
        if (tailwindClasses.length) {
            updatedClass = mergeClasses(tailwindClasses.join(' '), classValue);
            unmatched.forEach((c) => matchedClasses.add(c));
        }
        updatedClass = removeSpartanClasses(updatedClass);
        return `class="${updatedClass}"`;
    });
    if (updated !== text) {
        sourceFile.replaceWithText(updated);
    }
}
function applyToHtmlInStrings(sourceFile, styleMap, matchedClasses) {
    sourceFile.forEachDescendant((node) => {
        if (!isStringLiteralLike(node))
            return;
        const text = node.getLiteralText();
        if (!text.includes('class="'))
            return;
        const updated = text.replace(/class="([^"]+)"/g, (_, classValue) => {
            const classes = extractSpartanClasses(classValue);
            if (classes.length === 0)
                return `class="${classValue}"`;
            const unmatched = classes.filter((c) => !matchedClasses.has(c));
            const tailwindClasses = unmatched.map((c) => styleMap[c]).filter((v) => Boolean(v));
            let updatedClass = classValue;
            if (tailwindClasses.length) {
                updatedClass = mergeClasses(tailwindClasses.join(' '), classValue);
                unmatched.forEach((c) => matchedClasses.add(c));
            }
            updatedClass = removeSpartanClasses(updatedClass);
            return `class="${updatedClass}"`;
        });
        if (updated !== text) {
            node.setLiteralValue(updated);
        }
    });
}
function applyToClassesCalls(sourceFile, styleMap, matchedClasses) {
    sourceFile.forEachDescendant((node) => {
        if (!ts_morph_1.Node.isCallExpression(node))
            return;
        const expression = node.getExpression();
        if (!ts_morph_1.Node.isIdentifier(expression) || expression.getText() !== 'classes')
            return;
        const firstArg = node.getArguments()[0];
        if (!firstArg || !ts_morph_1.Node.isArrowFunction(firstArg))
            return;
        const body = firstArg.getBody();
        if (isStringLiteralLike(body)) {
            applyStyle(body, styleMap, matchedClasses);
            return;
        }
        if (ts_morph_1.Node.isArrayLiteralExpression(body)) {
            applyToArrayLiteral(body, styleMap, matchedClasses);
        }
    });
}
function applyToArrayLiteral(arrayNode, styleMap, matchedClasses) {
    for (const element of arrayNode.getElements()) {
        if (isStringLiteralLike(element)) {
            applyStyle(element, styleMap, matchedClasses);
            continue;
        }
        element.forEachDescendant((node) => {
            if (isStringLiteralLike(node)) {
                applyStyle(node, styleMap, matchedClasses);
            }
        });
    }
}
function applyToHlmCalls(sourceFile, styleMap, matchedClasses) {
    sourceFile.forEachDescendant((node) => {
        if (!ts_morph_1.Node.isCallExpression(node))
            return;
        const expression = node.getExpression();
        if (!ts_morph_1.Node.isIdentifier(expression) || expression.getText() !== 'hlm')
            return;
        for (const arg of node.getArguments()) {
            if (!isStringLiteralLike(arg))
                continue;
            applyStyle(arg, styleMap, matchedClasses);
        }
        removeEmptyArgumentsFromHlm(node);
    });
}
function applyStyle(stringNode, styleMap, matchedClasses) {
    const value = stringNode.getLiteralText();
    const classes = extractSpartanClasses(value);
    if (!classes.length)
        return;
    const unmatched = classes.filter((c) => !matchedClasses.has(c));
    const tailwind = unmatched.map((c) => styleMap[c]).filter((v) => Boolean(v));
    let updated = value;
    if (tailwind.length) {
        updated = mergeClasses(tailwind.join(' '), value);
        unmatched.forEach((c) => matchedClasses.add(c));
    }
    updated = removeSpartanClasses(updated);
    stringNode.setLiteralValue(updated);
}
function extractSpartanClasses(str) {
    const matches = str.matchAll(/\bspartan-[\w-]+\b/g);
    return Array.from(matches, (m) => m[0]);
}
function removeSpartanClasses(str) {
    return str
        .replace(/\bspartan-[\w-]+\b/g, (match) => {
        if (ALLOWLIST.has(match))
            return match;
        return '';
    })
        .replace(/\s+/g, ' ')
        .trim();
}
function removeEmptyArgumentsFromHlm(call) {
    const args = call.getArguments();
    const filtered = args.filter((arg) => {
        if (isStringLiteralLike(arg)) {
            return arg.getLiteralText().trim() !== '';
        }
        return true;
    });
    if (filtered.length !== args.length) {
        const argTexts = filtered.map((a) => a.getText());
        call.replaceWithText(`hlm(${argTexts.join(', ')})`);
    }
}
function mergeClasses(newClasses, existing) {
    return (0, tailwind_merge_1.twMerge)(newClasses, existing);
}
function applyToCvaCalls(sourceFile, styleMap, matchedClasses) {
    sourceFile.forEachDescendant((node) => {
        if (!ts_morph_1.Node.isCallExpression(node)) {
            return;
        }
        const expression = node.getExpression();
        if (!ts_morph_1.Node.isIdentifier(expression) || expression.getText() !== 'cva') {
            return;
        }
        const baseArg = node.getArguments()[0];
        if (ts_morph_1.Node.isStringLiteral(baseArg)) {
            applyStyleToCvaString(baseArg, styleMap, matchedClasses);
        }
        const configArg = node.getArguments()[1];
        if (!configArg || !ts_morph_1.Node.isObjectLiteralExpression(configArg)) {
            return;
        }
        const variantsProp = configArg
            .getProperties()
            .find((prop) => ts_morph_1.Node.isPropertyAssignment(prop) &&
            ts_morph_1.Node.isIdentifier(prop.getNameNode()) &&
            prop.getNameNode().getText() === 'variants');
        if (!variantsProp || !ts_morph_1.Node.isPropertyAssignment(variantsProp)) {
            return;
        }
        const variantsObj = variantsProp.getInitializer();
        if (!variantsObj || !ts_morph_1.Node.isObjectLiteralExpression(variantsObj)) {
            return;
        }
        variantsObj.getProperties().forEach((typeProp) => {
            if (!ts_morph_1.Node.isPropertyAssignment(typeProp)) {
                return;
            }
            const typeObj = typeProp.getInitializer();
            if (!typeObj || !ts_morph_1.Node.isObjectLiteralExpression(typeObj)) {
                return;
            }
            typeObj.getProperties().forEach((variantProp) => {
                if (!ts_morph_1.Node.isPropertyAssignment(variantProp)) {
                    return;
                }
                const variantValue = variantProp.getInitializer();
                if (variantValue && ts_morph_1.Node.isStringLiteral(variantValue)) {
                    applyStyleToCvaString(variantValue, styleMap, matchedClasses);
                }
            });
        });
    });
}
function applyStyleToCvaString(stringNode, styleMap, matchedClasses) {
    const stringValue = stringNode.getLiteralText();
    const hlmClasses = extractCnClasses(stringValue);
    if (hlmClasses.length === 0) {
        return;
    }
    // Process all spartan-* classes, not just the first one
    const unmatchedClasses = hlmClasses.filter((hlmClass) => !matchedClasses.has(hlmClass));
    if (unmatchedClasses.length === 0) {
        // All classes already matched, just clean up non-allowlisted ones
        const updated = removeCnClasses(stringValue);
        stringNode.setLiteralValue(updated);
        return;
    }
    const tailwindClassesToApply = unmatchedClasses
        .map((hlmClass) => styleMap[hlmClass])
        .filter((classes) => Boolean(classes));
    if (tailwindClassesToApply.length > 0) {
        const mergedClasses = tailwindClassesToApply.join(' ');
        const updated = removeCnClasses(mergeClasses(mergedClasses, stringValue));
        stringNode.setLiteralValue(updated);
        unmatchedClasses.forEach((hlmClass) => matchedClasses.add(hlmClass));
    }
    else {
        // No styles to apply, but still need to clean up non-allowlisted classes
        const updated = removeCnClasses(stringValue);
        stringNode.setLiteralValue(updated);
    }
}
function extractCnClasses(str) {
    const matches = str.matchAll(/\bspartan-[\w-]+\b/g);
    return Array.from(matches, (match) => match[0]);
}
function removeCnClasses(str) {
    return str
        .replace(/\bspartan-[\w-]+\b/g, (match) => {
        // Preserve allowlisted classes
        if (ALLOWLIST.has(match)) {
            return match;
        }
        return '';
    })
        .replace(/\s+/g, ' ')
        .trim();
}
