// @spartan-ng/cli/src/generators/migrate-brain-accordion-trigger/utils/html-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HTML_PARSE_CONFIG = void 0;
exports.parseHtmlForAccordionMigration = parseHtmlForAccordionMigration;
exports.findAccordionButtons = findAccordionButtons;
exports.isInHeading = isInHeading;
exports.hasSignificantSiblings = hasSignificantSiblings;
exports.getButtonText = getButtonText;
exports.wrapButtonInHeading = wrapButtonInHeading;
const tslib_1 = require("tslib");
const node_html_parser_1 = tslib_1.__importStar(require("node-html-parser"));
exports.HTML_PARSE_CONFIG = {
    lowerCaseTagName: false,
    comment: false,
    voidTag: {
        tags: [
            'img',
            'br',
            'hr',
            'area',
            'base',
            'col',
            'embed',
            'input',
            'link',
            'meta',
            'param',
            'source',
            'track',
            'wbr',
        ],
    },
};
function parseHtmlForAccordionMigration(content) {
    return (0, node_html_parser_1.default)(content, exports.HTML_PARSE_CONFIG);
}
function findAccordionButtons(root) {
    return root.querySelectorAll('button[brnAccordionTrigger], button[hlmAccordionTrigger]');
}
function isInHeading(button) {
    const parent = button.parentNode;
    return !!(parent && /^h[1-6]$/i.test(parent.tagName));
}
function hasSignificantSiblings(button) {
    const parent = button.parentNode;
    if (!parent)
        return false;
    const siblings = parent.childNodes.filter((child) => {
        if (child === button)
            return false;
        if (child.nodeType === node_html_parser_1.NodeType.TEXT_NODE) {
            return child.text?.trim().length > 0;
        }
        return true;
    });
    return siblings.length > 0;
}
function getButtonText(button, maxLength = 50) {
    return button.innerText?.trim()?.substring(0, maxLength) || '';
}
function wrapButtonInHeading(button, useContentsClass = true) {
    const h3Html = useContentsClass ? '<h3 class="contents"></h3>' : '<h3></h3>';
    const h3 = parseHtmlForAccordionMigration(h3Html).firstChild;
    const parent = button.parentNode;
    button.remove();
    h3.appendChild(button);
    if (parent) {
        parent.prepend(h3);
    }
}
