// @spartan-ng/cli/src/generators/base/generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hlmBaseGenerator = hlmBaseGenerator;
const tslib_1 = require("tslib");
const devkit_1 = require("@nx/devkit");
const js_1 = require("@nx/js");
const path = tslib_1.__importStar(require("node:path"));
const tsconfig_1 = require("../../utils/tsconfig");
const version_utils_1 = require("../../utils/version-utils");
const build_dependency_array_1 = require("./lib/build-dependency-array");
const deleteFiles_1 = require("./lib/deleteFiles");
const get_target_library_directory_1 = require("./lib/get-target-library-directory");
const initialize_angular_library_1 = require("./lib/initialize-angular-library");
const generators_1 = require("@nx/angular/generators");
const fs_1 = require("fs");
const single_lib_name_1 = require("./lib/single-lib-name");
const create_style_map_1 = require("./lib/styles/create-style-map");
const transform_1 = require("./lib/styles/transform");
const versions_1 = require("./versions");
const styleMapCache = new Map();
async function getStyleMap(style) {
    if (styleMapCache.has(style)) {
        return styleMapCache.get(style);
    }
    try {
        const cssPath = path.join(__dirname, '..', 'ui', `style-${style}.css`);
        const css = await fs_1.promises.readFile(cssPath, 'utf-8');
        const styleMap = (0, create_style_map_1.createStyleMap)(css);
        styleMapCache.set(style, styleMap);
        return styleMap;
    }
    catch {
        return {};
    }
}
function isAlreadyInstalled(tree, alias) {
    const existingPaths = (0, tsconfig_1.readTsConfigPathsFromTree)(tree);
    return alias in existingPaths;
}
function setupAngularCliProject(tree, alias, targetLibDir) {
    (0, js_1.addTsConfigPath)(tree, alias, [`./${(0, devkit_1.joinPathFragments)(targetLibDir, 'src', 'index.ts').replace(/\\/g, '/')}`]);
}
async function generateEntrypointFiles(tree, alias, options) {
    const targetLibDir = `${options.directory}/${options.name}/src`;
    if (options.buildable) {
        await (0, generators_1.librarySecondaryEntryPointGenerator)(tree, {
            name: options.name,
            library: single_lib_name_1.singleLibName,
            skipFormat: true,
            skipModule: true,
        });
    }
    else {
        (0, devkit_1.updateJson)(tree, 'tsconfig.base.json', (json) => {
            json.compilerOptions ||= {};
            json.compilerOptions.paths ||= {};
            json.compilerOptions.paths[alias] = [`./${(0, devkit_1.joinPathFragments)(targetLibDir, 'index.ts').replace(/\\/g, '/')}`];
            return json;
        });
    }
    (0, devkit_1.generateFiles)(tree, path.join(__dirname, '..', 'ui', 'libs', options.name, 'files'), targetLibDir, options);
    return targetLibDir;
}
function generateLibraryFiles(tree, targetLibDir, options) {
    const deletePath = (0, devkit_1.joinPathFragments)(options.directory, options.name, 'src', 'lib', options.name);
    (0, deleteFiles_1.deleteFiles)(tree, deletePath);
    (0, devkit_1.generateFiles)(tree, path.join(__dirname, '..', 'ui', 'libs', options.name, 'files'), (0, devkit_1.joinPathFragments)(targetLibDir, 'src'), options);
}
function registerDependencies(tree, options) {
    const cdkVersion = (0, version_utils_1.getInstalledPackageVersion)(tree, '@angular/cdk', versions_1.FALLBACK_ANGULAR_CDK_VERSION, true);
    const dependencies = (0, build_dependency_array_1.buildDependencyArray)(tree, options, cdkVersion);
    const devDependencies = (0, build_dependency_array_1.buildDevDependencyArray)(tree);
    return (0, devkit_1.addDependenciesToPackageJson)(tree, dependencies, devDependencies);
}
async function hlmBaseGenerator(tree, options) {
    const tasks = [];
    const targetLibDir = (0, get_target_library_directory_1.getTargetLibraryDirectory)(options, tree);
    const tsConfigAlias = `${options.importAlias}/${options.name}`;
    if (isAlreadyInstalled(tree, tsConfigAlias)) {
        console.log(`Skipping ${tsConfigAlias}. It's already installed!`);
        return (0, devkit_1.runTasksInSerial)(...tasks);
    }
    if (options.angularCli) {
        setupAngularCliProject(tree, tsConfigAlias, targetLibDir);
    }
    else if (options.generateAs === 'library') {
        tasks.push(await (0, initialize_angular_library_1.initializeAngularLibrary)(tree, options));
    }
    let filesPath = targetLibDir;
    if (options.generateAs === 'entrypoint') {
        filesPath = await generateEntrypointFiles(tree, tsConfigAlias, options);
    }
    else {
        generateLibraryFiles(tree, targetLibDir, options);
    }
    // TODO: add style to component.json
    const styleMap = await getStyleMap('vega');
    const generatedFiles = [];
    (0, devkit_1.visitNotIgnoredFiles)(tree, filesPath, (filePath) => {
        generatedFiles.push(filePath);
    });
    for (const filePath of generatedFiles) {
        const content = tree.read(filePath, 'utf-8');
        if (!content)
            continue;
        const transformed = await (0, transform_1.transformStyle)(content, { styleMap });
        tree.write(filePath, transformed);
    }
    tasks.push(registerDependencies(tree, options));
    return (0, devkit_1.runTasksInSerial)(...tasks);
}
exports.default = hlmBaseGenerator;
