// @spartan-ng/cli/src/generators/ui/generator.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = hlmUIGenerator;
exports.createPrimitiveLibraries = createPrimitiveLibraries;
const devkit_1 = require("@nx/devkit");
const enquirer_1 = require("enquirer");
const config_1 = require("../../utils/config");
const initialize_angular_library_1 = require("../base/lib/initialize-angular-library");
const single_lib_name_1 = require("../base/lib/single-lib-name");
const add_dependent_primitive_1 = require("./add-dependent-primitive");
async function hlmUIGenerator(tree, options) {
    const tasks = [];
    const config = await (0, config_1.getOrCreateConfig)(tree, {
        componentsPath: options.directory,
        angularCli: options.angularCli,
    });
    const availablePrimitives = await Promise.resolve().then(() => __importStar(require('./supported-ui-libraries.json'))).then((m) => m.default);
    const availablePrimitiveNames = Object.keys(availablePrimitives);
    let response = { primitives: [] };
    if (options.name && availablePrimitiveNames.includes(options.name)) {
        response.primitives.push(options.name);
    }
    else {
        if (options.name) {
            console.log('Cannot resolve primitive with name:', options.name);
        }
        response = await (0, enquirer_1.prompt)({
            type: 'multiselect',
            required: true,
            name: 'primitives',
            message: 'Choose which primitives you want to copy',
            choices: ['all', ...availablePrimitiveNames.sort()],
        });
    }
    tasks.push(...(await createPrimitiveLibraries(response, availablePrimitiveNames, availablePrimitives, tree, options, config)));
    return (0, devkit_1.runTasksInSerial)(...tasks);
}
async function createPrimitiveLibraries(response, availablePrimitiveNames, availablePrimitives, tree, options, config) {
    const allPrimitivesSelected = response.primitives.includes('all');
    const primitivesToCreate = allPrimitivesSelected ? availablePrimitiveNames : response.primitives;
    const tasks = [];
    const installPeerDependencies = options.installPeerDependencies === undefined
        ? true
        : typeof options.installPeerDependencies === 'string'
            ? options.installPeerDependencies === 'true'
            : options.installPeerDependencies;
    if (!response.primitives.includes('all') && installPeerDependencies) {
        await (0, add_dependent_primitive_1.addDependentPrimitives)(primitivesToCreate, false);
    }
    const projects = (0, devkit_1.getProjects)(tree);
    if (typeof config.generateAs === 'string' &&
        config.generateAs === 'entrypoint' &&
        !options.angularCli &&
        !projects.has(single_lib_name_1.singleLibName)) {
        const task = await (0, initialize_angular_library_1.initializeAngularEntrypoint)(tree, {
            tags: options.tags,
            directory: options.directory ?? config.componentsPath,
            buildable: config.buildable,
            importAlias: config.importAlias,
        });
        tasks.push(task);
    }
    // Use Promise.all() to handle parallel execution of primitive library creation tasks
    const installTasks = await Promise.all(primitivesToCreate.map(async (primitiveName) => {
        const name = availablePrimitives[primitiveName].name;
        const peerDependencies = removeHelmKeys(availablePrimitives[primitiveName].peerDependencies);
        const { generator } = await Promise.resolve(`${`./libs/${name}/generator`}`).then(s => __importStar(require(s)));
        return generator(tree, {
            name: '',
            peerDependencies,
            directory: options.directory ?? config.componentsPath,
            tags: options.tags,
            rootProject: options.rootProject,
            angularCli: options.angularCli,
            buildable: options.buildable ?? config.buildable,
            generateAs: options.generateAs ?? config.generateAs ?? 'library',
            importAlias: options.importAlias ?? config.importAlias ?? `@spartan-ng/helm`,
        });
    }));
    tasks.push(...installTasks.filter(Boolean));
    return tasks;
}
const removeHelmKeys = (obj) => Object.fromEntries(Object.entries(obj).filter(([key]) => !key.toLowerCase().includes('helm')));
