// @spartan-ng/cli/src/generators/ui/primitive-deps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDependentPrimitives = exports.primitiveDependencies = void 0;
exports.primitiveDependencies = {
    accordion: ['utils', 'icon'],
    alert: ['utils', 'icon'],
    'alert-dialog': ['utils', 'button'],
    'aspect-ratio': ['utils'],
    autocomplete: ['utils', 'popover', 'icon', 'input-group'],
    avatar: ['utils'],
    badge: ['utils'],
    breadcrumb: ['utils', 'icon'],
    button: ['utils'],
    'button-group': ['utils', 'button'],
    calendar: ['utils', 'button', 'icon', 'select'],
    card: ['utils'],
    carousel: ['utils', 'button', 'icon'],
    checkbox: ['utils', 'icon'],
    collapsible: ['utils'],
    combobox: ['utils', 'input-group', 'button', 'icon'],
    command: ['utils', 'button', 'icon'],
    'context-menu': ['utils', 'dropdown-menu'],
    'date-picker': ['utils', 'calendar', 'icon', 'popover'],
    dialog: ['utils', 'icon'],
    'dropdown-menu': ['utils', 'icon'],
    empty: ['utils'],
    field: ['utils', 'label', 'separator'],
    'hover-card': ['utils'],
    icon: [],
    input: ['utils'],
    'input-group': ['utils', 'button', 'input', 'textarea'],
    'input-otp': ['utils', 'icon'],
    item: ['utils', 'separator'],
    kbd: ['utils'],
    label: ['utils'],
    menubar: ['utils', 'dropdown-menu'],
    'native-select': ['utils', 'icon'],
    'navigation-menu': ['utils'],
    pagination: ['utils', 'button', 'icon', 'select'],
    popover: ['utils'],
    progress: ['utils'],
    'radio-group': ['utils'],
    resizable: ['utils'],
    'scroll-area': ['utils'],
    select: ['utils', 'icon'],
    separator: ['utils'],
    sheet: ['utils', 'icon', 'button'],
    sidebar: ['utils', 'button', 'icon', 'input', 'separator', 'sheet', 'skeleton', 'tooltip'],
    skeleton: ['utils'],
    slider: ['utils'],
    sonner: ['utils'],
    spinner: ['utils'],
    switch: ['utils'],
    table: ['utils'],
    tabs: ['utils', 'icon'],
    textarea: ['utils'],
    toggle: ['utils'],
    'toggle-group': ['utils', 'toggle'],
    tooltip: [],
    typography: ['utils'],
    utils: [],
};
const getDependentPrimitives = (primitives) => {
    const dependentPrimitives = new Set();
    for (const primitive of primitives) {
        const deps = exports.primitiveDependencies[primitive] ?? [];
        for (const dep of deps) {
            if (!primitives.includes(dep)) {
                // only add the dependent primitive if it's not already in the list of primitives to create
                dependentPrimitives.add(dep);
                (0, exports.getDependentPrimitives)([dep])
                    .filter((primitive) => !dependentPrimitives.has(primitive) && !primitives.includes(primitive))
                    .forEach((d) => dependentPrimitives.add(d));
            }
        }
    }
    return Array.from(dependentPrimitives);
};
exports.getDependentPrimitives = getDependentPrimitives;
