// @spartan-ng/cli/src/utils/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NXConfigSchema = exports.AngularCliConfigSchema = void 0;
exports.getImportAlias = getImportAlias;
exports.getOrCreateConfig = getOrCreateConfig;
const tslib_1 = require("tslib");
const devkit_1 = require("@nx/devkit");
const enquirer_1 = require("enquirer");
const zod_1 = tslib_1.__importStar(require("zod"));
const generate_as_1 = require("../generators/base/lib/generate-as");
const configPath = 'components.json';
exports.AngularCliConfigSchema = zod_1.default.object({
    componentsPath: zod_1.default.string().optional().default('libs/ui'),
    importAlias: zod_1.default
        .string()
        .optional()
        .default('@spartan-ng/helm')
        .refine((val) => !val.endsWith('/'), { message: 'importAlias should not end with a slash' }),
});
exports.NXConfigSchema = zod_1.default.object({
    componentsPath: zod_1.default.string().optional().default('libs/ui'),
    buildable: zod_1.default.boolean().optional().default(true),
    generateAs: zod_1.default.enum(generate_as_1.generateOptions).optional().default('library'),
    importAlias: zod_1.default
        .string()
        .optional()
        .default('@spartan-ng/helm')
        .refine((val) => !val.endsWith('/'), { message: 'importAlias should not end with a slash' }),
});
const getConfig = async (tree, isAngularCli) => {
    const raw = await (0, devkit_1.readJson)(tree, configPath);
    try {
        if (isAngularCli) {
            return exports.AngularCliConfigSchema.parse(raw);
        }
        else {
            return exports.NXConfigSchema.parse(raw);
        }
    }
    catch (e) {
        if (e instanceof zod_1.ZodError) {
            const rows = e.issues.map((issue) => {
                let expected;
                if ('expected' in issue) {
                    expected = String(issue.expected);
                }
                else if ('options' in issue) {
                    expected = issue.options.join(' | ');
                }
                return {
                    path: issue.path.join('.') || 'root',
                    message: issue.message,
                    expected: expected ?? '-',
                };
            });
            console.table(rows, ['path', 'expected', 'message']);
            throw new Error('Config validation failed. Please fix the issues above.');
        }
        else {
            throw e;
        }
    }
};
async function getImportAlias(tree, isAngularCli) {
    try {
        const config = await getConfig(tree, isAngularCli);
        return config.importAlias;
    }
    catch (error) {
        // components.json does not exist or is invalid
        return '@spartan-ng/helm';
    }
}
async function getOrCreateConfig(tree, defaults) {
    if (tree.exists(configPath)) {
        return getConfig(tree, defaults?.angularCli ?? false);
    }
    const questions = [
        {
            type: 'input',
            required: true,
            name: 'componentsPath',
            message: 'Choose a directory to place your spartan libraries, e.g. libs/ui',
            initial: defaults?.componentsPath ?? 'libs/ui',
            skip: !!defaults?.componentsPath,
        },
        // Conditionally add NX specific questions
        ...(!defaults?.angularCli
            ? [
                {
                    type: 'confirm',
                    name: 'buildable',
                    message: 'Should the libraries be buildable?',
                    initial: defaults?.buildable ?? true,
                    skip: typeof defaults?.buildable === 'boolean',
                },
                {
                    type: 'select',
                    choices: ['library', 'entrypoint'],
                    name: 'generateAs',
                    message: 'How should we generate the components?',
                    initial: 0,
                    skip: typeof defaults?.generateAs === 'string',
                },
            ]
            : []),
        {
            type: 'input',
            name: 'importAlias',
            message: 'What import alias should be used for the components?',
            initial: defaults?.importAlias ?? '@spartan-ng/helm',
            skip: !!defaults?.importAlias,
        },
    ];
    console.log('Configuration file not found, creating a new one...');
    const { componentsPath, buildable, generateAs, importAlias } = (await (0, enquirer_1.prompt)(questions));
    const config = { componentsPath, buildable, generateAs, importAlias };
    tree.write(configPath, JSON.stringify(config, null, 2));
    return config;
}
