// @spartan-ng/brain/fesm2022/spartan-ng-brain-resizable.mjs
import * as i0 from '@angular/core';
import { input, inject, ElementRef, numberAttribute, booleanAttribute, signal, computed, Directive, contentChildren, output, model, DOCUMENT, NgZone, afterNextRender } from '@angular/core';

let nextId$1 = 0;
class BrnResizablePanel {
    /** Unique ID for the panel.   */
    id = input(`brn-resizable-panel-${++nextId$1}`, ...(ngDevMode ? [{ debugName: "id" }] : []));
    /** Reference to the parent {@link BrnResizableGroup}. */
    _panelGroup = inject(BrnResizableGroup);
    /** Host DOM element reference. */
    el = inject((ElementRef));
    /**
     * The default size of the panel (percentage of container space).
     * - `undefined` → group decides initial size.
     * - Number → interpreted as percentage (0–100).
     */
    defaultSize = input(undefined, ...(ngDevMode ? [{ debugName: "defaultSize", transform: numberAttribute }] : [{ transform: numberAttribute }]));
    /** The minimum size this panel can shrink to (percentage). */
    minSize = input(0, ...(ngDevMode ? [{ debugName: "minSize", transform: numberAttribute }] : [{ transform: numberAttribute }]));
    /**	 The maximum size this panel can grow to (percentage).   */
    maxSize = input(100, ...(ngDevMode ? [{ debugName: "maxSize", transform: numberAttribute }] : [{ transform: numberAttribute }]));
    /** Whether this panel can be collapsed entirely. */
    collapsible = input(true, ...(ngDevMode ? [{ debugName: "collapsible", transform: booleanAttribute }] : [{ transform: booleanAttribute }]));
    /** Reactive signal holding the current size of the panel. */
    _panelSize = signal(100, ...(ngDevMode ? [{ debugName: "_panelSize" }] : []));
    /**
     * CSS flex style for this panel, derived from its current size.
     * Format: `"flex-grow flex-shrink flex-basis"`.
     *
     * Example: `"25 1 0"` means 25% width (or height in vertical layout).
     */
    _flex = computed(() => `${this._panelSize()} 1 0`, ...(ngDevMode ? [{ debugName: "_flex" }] : []));
    /**
     * Sets the size of the panel.
     * @param size New size (percentage of container space).
     */
    setSize(size) {
        this._panelSize.set(size);
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnResizablePanel, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.3.17", type: BrnResizablePanel, isStandalone: true, selector: "brn-resizable-panel, [brnResizablePanel]", inputs: { id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null }, defaultSize: { classPropertyName: "defaultSize", publicName: "defaultSize", isSignal: true, isRequired: false, transformFunction: null }, minSize: { classPropertyName: "minSize", publicName: "minSize", isSignal: true, isRequired: false, transformFunction: null }, maxSize: { classPropertyName: "maxSize", publicName: "maxSize", isSignal: true, isRequired: false, transformFunction: null }, collapsible: { classPropertyName: "collapsible", publicName: "collapsible", isSignal: true, isRequired: false, transformFunction: null } }, host: { attributes: { "data-panel": "", "data-slot": "resizable-panel" }, properties: { "attr.data-panel-group-id": "_panelGroup.id()", "attr.data-panel-id": "id()", "attr.data-panel-size": "_panelSize()", "id": "id()", "style.flex": "_flex()", "style.overflow": "\"hidden\"" } }, exportAs: ["brnResizablePanel"], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnResizablePanel, decorators: [{
            type: Directive,
            args: [{
                    selector: 'brn-resizable-panel, [brnResizablePanel]',
                    exportAs: 'brnResizablePanel',
                    host: {
                        '[attr.data-panel-group-id]': '_panelGroup.id()',
                        '[attr.data-panel-id]': 'id()',
                        '[attr.data-panel-size]': '_panelSize()',
                        '[id]': 'id()',
                        '[style.flex]': '_flex()',
                        '[style.overflow]': '"hidden"',
                        'data-panel': '',
                        'data-slot': 'resizable-panel',
                    },
                }]
        }], propDecorators: { id: [{ type: i0.Input, args: [{ isSignal: true, alias: "id", required: false }] }], defaultSize: [{ type: i0.Input, args: [{ isSignal: true, alias: "defaultSize", required: false }] }], minSize: [{ type: i0.Input, args: [{ isSignal: true, alias: "minSize", required: false }] }], maxSize: [{ type: i0.Input, args: [{ isSignal: true, alias: "maxSize", required: false }] }], collapsible: [{ type: i0.Input, args: [{ isSignal: true, alias: "collapsible", required: false }] }] } });

let nextId = 0;
class BrnResizableGroup {
    /** The id of the BrnResizableGroup */
    id = input(`brn-resizable-group-${++nextId}`, ...(ngDevMode ? [{ debugName: "id" }] : []));
    /** Host element reference. */
    _el = inject((ElementRef));
    /** Group orientation */
    direction = input('horizontal', ...(ngDevMode ? [{ debugName: "direction" }] : []));
    /** @internal Access all the panels within the group */
    panels = contentChildren(BrnResizablePanel, ...(ngDevMode ? [{ debugName: "panels" }] : []));
    /** event when resize starts */
    dragStart = output();
    /** event when resize ends */
    dragEnd = output();
    /** Resize panel group to the specified layout ([1 - 100, ...]). */
    layout = model([], ...(ngDevMode ? [{ debugName: "layout" }] : []));
    /** Called when group layout changes */
    layoutChanged = output();
    _document = inject(DOCUMENT);
    _zone = inject(NgZone);
    _resizeRaf = null;
    _pendingSizes = null;
    constructor() {
        afterNextRender(() => {
            this._initializePanelSizes();
        });
    }
    _initializePanelSizes() {
        const panels = this.panels();
        const totalPanels = panels.length;
        if (totalPanels === 0)
            return;
        const sizes = [];
        panels.forEach((panel, index) => {
            const defaultSize = panel.defaultSize() ?? this.layout()[index];
            const size = defaultSize !== undefined ? defaultSize : 100 / totalPanels;
            sizes.push(size);
            panel.setSize(size);
        });
        if (this.layout().toString() !== sizes.toString()) {
            this._setLayout(sizes);
        }
    }
    startResize(handleIndex, event) {
        event.preventDefault();
        const cursor = this.direction() === 'vertical' ? 'ns-resize' : 'ew-resize';
        this._document.body.style.cursor = `${cursor}`;
        const sizes = [...this.layout()];
        this.dragStart.emit();
        const startPosition = this._getEventPosition(event);
        const startSizes = [...sizes];
        const handleMove = (moveEvent) => {
            this._zone.runOutsideAngular(() => {
                this._handleResize(moveEvent, handleIndex, startPosition, startSizes);
            });
        };
        const handleEnd = () => {
            this._zone.run(() => this._endResize());
            this._document.body.style.cursor = 'default';
            this._document.removeEventListener('mousemove', handleMove);
            this._document.removeEventListener('touchmove', handleMove);
            this._document.removeEventListener('mouseup', handleEnd);
            this._document.removeEventListener('touchend', handleEnd);
        };
        this._zone.runOutsideAngular(() => {
            this._document.addEventListener('mousemove', handleMove);
            this._document.addEventListener('touchmove', handleMove);
            this._document.addEventListener('mouseup', handleEnd);
            this._document.addEventListener('touchend', handleEnd);
        });
    }
    _handleResize(event, handleIndex, startPosition, startSizes) {
        const currentPosition = this._getEventPosition(event);
        const delta = currentPosition - startPosition;
        const containerSize = this._getContainerSize();
        const deltaPercentage = (delta / containerSize) * 100;
        const newSizes = [...startSizes];
        const panels = this.panels();
        const leftPanel = panels[handleIndex];
        const rightPanel = panels[handleIndex + 1];
        if (!leftPanel || !rightPanel)
            return;
        const leftMin = leftPanel.minSize();
        const leftMax = leftPanel.maxSize();
        const rightMin = rightPanel.minSize();
        const rightMax = rightPanel.maxSize();
        let newLeftSize = startSizes[handleIndex] + deltaPercentage;
        let newRightSize = startSizes[handleIndex + 1] - deltaPercentage;
        newLeftSize = Math.max(leftMin, Math.min(leftMax, newLeftSize));
        newRightSize = Math.max(rightMin, Math.min(rightMax, newRightSize));
        const totalSize = newLeftSize + newRightSize;
        const originalTotal = startSizes[handleIndex] + startSizes[handleIndex + 1];
        if (Math.abs(totalSize - originalTotal) < 0.01) {
            newSizes[handleIndex] = newLeftSize;
            newSizes[handleIndex + 1] = newRightSize;
            // batch update
            this._queueResize(newSizes);
        }
    }
    _queueResize(sizes) {
        this._pendingSizes = sizes;
        if (this._resizeRaf !== null)
            return;
        this._resizeRaf = requestAnimationFrame(() => {
            this._resizeRaf = null;
            if (this._pendingSizes) {
                this._setLayout(this._pendingSizes);
                this.updatePanelStyles();
                this._pendingSizes = null;
            }
        });
    }
    _endResize() {
        if (this._resizeRaf !== null) {
            cancelAnimationFrame(this._resizeRaf);
            this._resizeRaf = null;
            this._pendingSizes = null;
        }
        this.dragEnd.emit();
    }
    updatePanelStyles() {
        const panels = this.panels();
        const sizes = this.layout();
        panels.forEach((panel, index) => {
            const size = sizes[index];
            if (size !== undefined) {
                panel.setSize(size);
            }
        });
    }
    _getEventPosition(event) {
        const layout = this.direction();
        if (event instanceof MouseEvent) {
            return layout === 'vertical' ? event.clientY : event.clientX;
        }
        else {
            const touch = event.touches[0];
            return layout === 'vertical' ? touch.clientY : touch.clientX;
        }
    }
    _getContainerSize() {
        const element = this._el.nativeElement;
        const layout = this.direction();
        return layout === 'vertical' ? element.offsetHeight : element.offsetWidth;
    }
    collapsePanel(index) {
        const panels = this.panels();
        const panel = panels[index];
        if (!panel || !panel.collapsible())
            return;
        const sizes = [...this.layout()];
        const isCollapsed = sizes[index] === 0;
        if (isCollapsed) {
            const panelDefaultSize = panel.defaultSize();
            const defaultSize = panelDefaultSize !== undefined ? panelDefaultSize / panels.length : 100 / panels.length;
            sizes[index] = defaultSize;
            const totalOthers = sizes.reduce((sum, size, i) => (i !== index ? sum + size : sum), 0);
            const scale = (100 - defaultSize) / totalOthers;
            sizes.forEach((size, i) => {
                if (i !== index) {
                    sizes[i] = size * scale;
                }
            });
        }
        else {
            const collapsedSize = sizes[index];
            sizes[index] = 0;
            const totalOthers = sizes.reduce((sum, size, i) => (i !== index ? sum + size : sum), 0);
            const scale = (totalOthers + collapsedSize) / totalOthers;
            sizes.forEach((size, i) => {
                if (i !== index) {
                    sizes[i] = size * scale;
                }
            });
        }
        this._setLayout(sizes);
        this.updatePanelStyles();
    }
    _setLayout(sizes) {
        this.layout.set(sizes);
        this.layoutChanged.emit(sizes);
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnResizableGroup, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.2.0", version: "20.3.17", type: BrnResizableGroup, isStandalone: true, selector: "brn-resizable-group, [brnResizableGroup]", inputs: { id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null }, direction: { classPropertyName: "direction", publicName: "direction", isSignal: true, isRequired: false, transformFunction: null }, layout: { classPropertyName: "layout", publicName: "layout", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { dragStart: "dragStart", dragEnd: "dragEnd", layout: "layoutChange", layoutChanged: "layoutChanged" }, host: { attributes: { "data-panel-group": "", "data-slot": "resizable-panel-group" }, properties: { "attr.data-panel-group-direction": "direction()", "attr.data-panel-group-id": "id()", "id": "id()" } }, queries: [{ propertyName: "panels", predicate: BrnResizablePanel, isSignal: true }], exportAs: ["brnResizableGroup"], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnResizableGroup, decorators: [{
            type: Directive,
            args: [{
                    selector: 'brn-resizable-group, [brnResizableGroup]',
                    exportAs: 'brnResizableGroup',
                    host: {
                        '[attr.data-panel-group-direction]': 'direction()',
                        '[attr.data-panel-group-id]': 'id()',
                        '[id]': 'id()',
                        'data-panel-group': '',
                        'data-slot': 'resizable-panel-group',
                    },
                }]
        }], ctorParameters: () => [], propDecorators: { id: [{ type: i0.Input, args: [{ isSignal: true, alias: "id", required: false }] }], direction: [{ type: i0.Input, args: [{ isSignal: true, alias: "direction", required: false }] }], panels: [{ type: i0.ContentChildren, args: [i0.forwardRef(() => BrnResizablePanel), { isSignal: true }] }], dragStart: [{ type: i0.Output, args: ["dragStart"] }], dragEnd: [{ type: i0.Output, args: ["dragEnd"] }], layout: [{ type: i0.Input, args: [{ isSignal: true, alias: "layout", required: false }] }, { type: i0.Output, args: ["layoutChange"] }], layoutChanged: [{ type: i0.Output, args: ["layoutChanged"] }] } });

class BrnResizableHandle {
    /** Parent resizable group. */
    _resizable = inject(BrnResizableGroup);
    /** Host element reference. */
    _el = inject((ElementRef));
    /** Whether a visual handle is rendered inside the separator. */
    withHandle = input(false, ...(ngDevMode ? [{ debugName: "withHandle", transform: booleanAttribute }] : [{ transform: booleanAttribute }]));
    /** Whether the handle is disabled (not interactive). */
    disabled = input(false, ...(ngDevMode ? [{ debugName: "disabled", transform: booleanAttribute }] : [{ transform: booleanAttribute }]));
    /** The direction of the resizable group (`horizontal` or `vertical`). */
    _direction = this._resizable.direction;
    /** Computed layout orientation based on the parent group. */
    _layout = computed(() => this._resizable?.direction() || 'horizontal', ...(ngDevMode ? [{ debugName: "_layout" }] : []));
    /** Index of this handle relative to panels in the group. */
    _handleIndex = 0;
    constructor() {
        afterNextRender(() => {
            this._handleIndex = this._getHandleIndex();
            this._el.nativeElement.onpointerdown = (e) => {
                this._el.nativeElement.setPointerCapture(e.pointerId);
            };
            this._el.nativeElement.onpointerup = (e) => {
                this._el.nativeElement.releasePointerCapture(e.pointerId);
            };
        });
    }
    _getHandleIndex() {
        const host = this._el.nativeElement;
        const parent = host.parentElement;
        if (!parent)
            return -1;
        // Collect all panels under this group
        const panels = Array.from(parent.querySelectorAll('[data-slot="resizable-panel"]'));
        // Find the previous panel (the closest one before this handle)
        const prevPanel = host.previousElementSibling;
        if (!prevPanel)
            return -1;
        // Return its index among all panels
        return panels.indexOf(prevPanel);
    }
    _handleMouseDown(event) {
        if (this.disabled() || !this._resizable)
            return;
        this._resizable.startResize(this._handleIndex, event);
    }
    _handleKeyDown(event) {
        if (this.disabled())
            return;
        const panels = this._resizable.panels();
        const handleIndex = this._handleIndex;
        const layout = this._layout();
        let delta = 0;
        const step = event.shiftKey ? 10 : 1;
        switch (event.key) {
            case 'ArrowLeft':
                if (layout === 'horizontal')
                    delta = -step;
                break;
            case 'ArrowRight':
                if (layout === 'horizontal')
                    delta = step;
                break;
            case 'ArrowUp':
                if (layout === 'vertical')
                    delta = -step;
                break;
            case 'ArrowDown':
                if (layout === 'vertical')
                    delta = step;
                break;
            case 'Home':
                event.preventDefault();
                this._moveToExtreme(true);
                return;
            case 'End':
                event.preventDefault();
                this._moveToExtreme(false);
                return;
            case 'Enter':
            case ' ':
                event.preventDefault();
                if (panels[handleIndex]?.collapsible() || panels[handleIndex + 1]?.collapsible()) {
                    const collapsibleIndex = panels[handleIndex]?.collapsible() ? handleIndex : handleIndex + 1;
                    this._resizable.collapsePanel(collapsibleIndex);
                }
                return;
            default:
                return;
        }
        if (delta !== 0) {
            event.preventDefault();
            this._adjustSizes(delta);
        }
    }
    _getPanelContext() {
        const panels = this._resizable.panels();
        const li = this._handleIndex;
        const ri = li + 1;
        const left = panels[li];
        const right = panels[ri];
        if (!left || !right)
            return null;
        const sizes = [...this._resizable.layout()];
        return {
            sizes,
            li,
            ri,
            left,
            right,
            leftMin: left.minSize(),
            leftMax: left.maxSize(),
            rightMin: right.minSize(),
            rightMax: right.maxSize(),
        };
    }
    _adjustSizes(delta) {
        const ctx = this._getPanelContext();
        if (!ctx)
            return;
        let newLeftSize = ctx.sizes[ctx.li] + delta;
        let newRightSize = ctx.sizes[ctx.ri] - delta;
        newLeftSize = Math.max(ctx.leftMin, Math.min(ctx.leftMax, newLeftSize));
        newRightSize = Math.max(ctx.rightMin, Math.min(ctx.rightMax, newRightSize));
        const totalSize = newLeftSize + newRightSize;
        const originalTotal = ctx.sizes[ctx.li] + ctx.sizes[ctx.ri];
        if (Math.abs(totalSize - originalTotal) < 0.01) {
            ctx.sizes[ctx.li] = newLeftSize;
            ctx.sizes[ctx.ri] = newRightSize;
            this._resizable.layout.set(ctx.sizes);
            this._resizable.updatePanelStyles();
        }
    }
    _moveToExtreme(toMin) {
        const ctx = this._getPanelContext();
        if (!ctx)
            return;
        const totalSize = ctx.sizes[ctx.li] + ctx.sizes[ctx.ri];
        if (toMin) {
            ctx.sizes[ctx.li] = ctx.leftMin;
            ctx.sizes[ctx.ri] = Math.min(totalSize - ctx.leftMin, ctx.rightMax);
        }
        else {
            ctx.sizes[ctx.li] = Math.min(totalSize - ctx.rightMin, ctx.leftMax);
            ctx.sizes[ctx.ri] = ctx.rightMin;
        }
        this._resizable.layout.set(ctx.sizes);
        this._resizable.updatePanelStyles();
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnResizableHandle, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.3.17", type: BrnResizableHandle, isStandalone: true, selector: "brn-resizable-handle, [brnResizeHandle]", inputs: { withHandle: { classPropertyName: "withHandle", publicName: "withHandle", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "disabled", isSignal: true, isRequired: false, transformFunction: null } }, host: { listeners: { "mousedown": "_handleMouseDown($event)", "keydown": "_handleKeyDown($event)" }, properties: { "attr.data-layout": "_layout()", "attr.tabindex": "disabled() ? null : 0", "attr.role": "\"separator\"", "attr.data-panel-group-direction": "_direction()", "attr.aria-orientation": "_layout() === \"vertical\" ? \"horizontal\" : \"vertical\"", "attr.aria-disabled": "disabled()" } }, exportAs: ["brnResizableHandle"], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnResizableHandle, decorators: [{
            type: Directive,
            args: [{
                    selector: 'brn-resizable-handle, [brnResizeHandle]',
                    exportAs: 'brnResizableHandle',
                    host: {
                        '[attr.data-layout]': '_layout()',
                        '[attr.tabindex]': 'disabled() ? null : 0',
                        '[attr.role]': '"separator"',
                        '[attr.data-panel-group-direction]': '_direction()',
                        '[attr.aria-orientation]': '_layout() === "vertical" ? "horizontal" : "vertical"',
                        '[attr.aria-disabled]': 'disabled()',
                        '(mousedown)': '_handleMouseDown($event)',
                        '(keydown)': '_handleKeyDown($event)',
                    },
                }]
        }], ctorParameters: () => [], propDecorators: { withHandle: [{ type: i0.Input, args: [{ isSignal: true, alias: "withHandle", required: false }] }], disabled: [{ type: i0.Input, args: [{ isSignal: true, alias: "disabled", required: false }] }] } });

/**
 * Generated bundle index. Do not edit.
 */

export { BrnResizableGroup, BrnResizableHandle, BrnResizablePanel };
