// @spartan-ng/brain/fesm2022/spartan-ng-brain-field.mjs
import * as i0 from '@angular/core';
import { signal, computed, Injectable, input, booleanAttribute, Directive, InjectionToken, inject, Injector, DestroyRef, effect } from '@angular/core';
import { NgForm, FormGroupDirective, NgControl } from '@angular/forms';
import { ErrorStateMatcher, createStateTracker } from '@spartan-ng/brain/forms';

class BrnFieldA11yService {
    _descriptions = signal([], ...(ngDevMode ? [{ debugName: "_descriptions" }] : []));
    _errors = signal([], ...(ngDevMode ? [{ debugName: "_errors" }] : []));
    describedBy = computed(() => {
        const ids = [...this._descriptions(), ...this._errors()].filter(Boolean);
        const uniqueIds = [...new Set(ids)];
        return uniqueIds.length ? uniqueIds.join(' ') : null;
    }, ...(ngDevMode ? [{ debugName: "describedBy" }] : []));
    registerDescription(id) {
        this._descriptions.update((ids) => (ids.includes(id) ? ids : [...ids, id]));
    }
    unregisterDescription(id) {
        this._descriptions.update((ids) => ids.filter((value) => value !== id));
    }
    registerError(id) {
        this._errors.update((ids) => (ids.includes(id) ? ids : [...ids, id]));
    }
    unregisterError(id) {
        this._errors.update((ids) => ids.filter((value) => value !== id));
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnFieldA11yService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
    /** @nocollapse */ static ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnFieldA11yService });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnFieldA11yService, decorators: [{
            type: Injectable
        }] });

class BrnField {
    _brnFieldControl = signal(null, ...(ngDevMode ? [{ debugName: "_brnFieldControl" }] : []));
    _labelable = signal(null, ...(ngDevMode ? [{ debugName: "_labelable" }] : []));
    /** Whether the field is invalid. Overrides the `data-invalid` attribute. */
    dataInvalid = input(false, ...(ngDevMode ? [{ debugName: "dataInvalid", transform: booleanAttribute,
            alias: 'data-invalid' }] : [{
            transform: booleanAttribute,
            alias: 'data-invalid',
        }]));
    /**
     * Whether to force the field into an invalid state, regardless of the form control's state.
     * Overrides both the `data-invalid` and `data-matches-spartan-invalid` attributes.
     */
    forceInvalid = input(false, ...(ngDevMode ? [{ debugName: "forceInvalid", transform: booleanAttribute }] : [{
            transform: booleanAttribute,
        }]));
    _invalid = computed(() => {
        if (this.forceInvalid() || this.dataInvalid())
            return true;
        const control = this._brnFieldControl();
        if (!control || !control.ngControl)
            return false;
        return control.controlState()?.invalid;
    }, ...(ngDevMode ? [{ debugName: "_invalid" }] : []));
    _spartanInvalid = computed(() => {
        return this.forceInvalid() || (this._brnFieldControl()?.controlState()?.spartanInvalid ?? null);
    }, ...(ngDevMode ? [{ debugName: "_spartanInvalid" }] : []));
    _dirty = computed(() => {
        return this._brnFieldControl()?.controlState()?.dirty ?? null;
    }, ...(ngDevMode ? [{ debugName: "_dirty" }] : []));
    _touched = computed(() => {
        return this._brnFieldControl()?.controlState()?.touched ?? null;
    }, ...(ngDevMode ? [{ debugName: "_touched" }] : []));
    labelableId = computed(() => this._brnFieldControl()?.id?.() ?? this._labelable()?.labelableId(), ...(ngDevMode ? [{ debugName: "labelableId" }] : []));
    errors = computed(() => this._brnFieldControl()?.errors() ?? null, ...(ngDevMode ? [{ debugName: "errors" }] : []));
    controlState = computed(() => this._brnFieldControl()?.controlState() ?? null, ...(ngDevMode ? [{ debugName: "controlState" }] : []));
    registerFieldControl(fieldControl) {
        this._brnFieldControl.set(fieldControl);
    }
    registerLabelable(labelable) {
        this._labelable.set(labelable);
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnField, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.3.17", type: BrnField, isStandalone: true, selector: "[brnField],brn-field", inputs: { dataInvalid: { classPropertyName: "dataInvalid", publicName: "data-invalid", isSignal: true, isRequired: false, transformFunction: null }, forceInvalid: { classPropertyName: "forceInvalid", publicName: "forceInvalid", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "attr.data-invalid": "_invalid() ? \"true\" : null", "attr.data-matches-spartan-invalid": "_spartanInvalid() ? \"true\" : null", "attr.data-touched": "_touched() ? \"true\" : null", "attr.data-dirty": "_dirty() ? \"true\" : null" } }, providers: [BrnFieldA11yService], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnField, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnField],brn-field',
                    providers: [BrnFieldA11yService],
                    host: {
                        '[attr.data-invalid]': '_invalid() ? "true" : null',
                        '[attr.data-matches-spartan-invalid]': '_spartanInvalid() ? "true" : null',
                        '[attr.data-touched]': '_touched() ? "true" : null',
                        '[attr.data-dirty]': '_dirty() ? "true" : null',
                    },
                }]
        }], propDecorators: { dataInvalid: [{ type: i0.Input, args: [{ isSignal: true, alias: "data-invalid", required: false }] }], forceInvalid: [{ type: i0.Input, args: [{ isSignal: true, alias: "forceInvalid", required: false }] }] } });

const BrnLabelable = new InjectionToken('BrnLabelable');
function provideBrnLabelable(labelable) {
    return { provide: BrnLabelable, useExisting: labelable };
}
function injectBrnLabelable() {
    return inject(BrnLabelable, { optional: true });
}

class BrnFieldControl {
    _injector = inject(Injector);
    _errorStateMatcher = inject(ErrorStateMatcher);
    _parentForm = inject(NgForm, { optional: true });
    _parentFormGroup = inject(FormGroupDirective, { optional: true });
    _field = inject(BrnField, { optional: true });
    _destroyRef = inject(DestroyRef);
    _idEffectRef;
    _stateTracker = signal(null, ...(ngDevMode ? [{ debugName: "_stateTracker" }] : []));
    /** Sentinel value to differentiate "never checked" from "control is null". */
    _lastControl = null;
    /** Gets the AbstractControlDirective for this control. */
    ngControl = null;
    id = signal(undefined, ...(ngDevMode ? [{ debugName: "id" }] : []));
    controlState = computed(() => this._stateTracker()?.controlState() ?? null, ...(ngDevMode ? [{ debugName: "controlState" }] : []));
    errors = computed(() => this._stateTracker()?.errors() ?? null, ...(ngDevMode ? [{ debugName: "errors" }] : []));
    dirty = computed(() => this._stateTracker()?.dirty() ?? null, ...(ngDevMode ? [{ debugName: "dirty" }] : []));
    invalid = computed(() => this._stateTracker()?.invalid() ?? null, ...(ngDevMode ? [{ debugName: "invalid" }] : []));
    spartanInvalid = computed(() => this._stateTracker()?.spartanInvalid() ?? null, ...(ngDevMode ? [{ debugName: "spartanInvalid" }] : []));
    touched = computed(() => this._stateTracker()?.touched() ?? null, ...(ngDevMode ? [{ debugName: "touched" }] : []));
    constructor() {
        this._field?.registerFieldControl(this);
        this._destroyRef.onDestroy(() => {
            this._idEffectRef?.destroy();
            this._stateTracker()?.destroy();
        });
    }
    ngOnInit() {
        this.ngControl = this._injector.get(NgControl, null);
        // Try to sync the tracker eagerly.
        this._syncTracker();
        // For template-driven forms and FormControlName, the control is often resolved
        // asynchronously by Angular's directives after our ngOnInit/ngDoCheck.
        // Schedule a one-time microtask to catch the initial resolution.
        if (this.ngControl) {
            Promise.resolve().then(() => {
                this._syncTracker();
            });
        }
        const labelable = this._injector.get(BrnLabelable, null);
        if (labelable) {
            this._idEffectRef = effect(() => {
                this.id.set(labelable.labelableId());
            }, ...(ngDevMode ? [{ debugName: "_idEffectRef", injector: this._injector }] : [{ injector: this._injector }]));
        }
    }
    // Re-evaluate the control reference on every change detection cycle because
    // the underlying AbstractControl may change when [formControl] rebinds to a new instance.
    // When the instance changes we tear down the old tracker and create a fresh one.
    ngDoCheck() {
        this._syncTracker();
    }
    /** @returns true if the control reference changed */
    _syncTracker() {
        if (!this.ngControl)
            return;
        const currentControl = this.ngControl.control ?? null;
        if (currentControl === this._lastControl)
            return;
        this._lastControl = currentControl;
        this._stateTracker()?.destroy();
        this._stateTracker.set(currentControl
            ? createStateTracker(this.ngControl, this._errorStateMatcher, this._parentFormGroup, this._parentForm)
            : null);
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnFieldControl, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "20.3.17", type: BrnFieldControl, isStandalone: true, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnFieldControl, decorators: [{
            type: Directive
        }], ctorParameters: () => [] });

class BrnFieldControlDescribedBy {
    describedBy = input(null, ...(ngDevMode ? [{ debugName: "describedBy", alias: 'aria-describedby' }] : [{ alias: 'aria-describedby' }]));
    _a11y = inject(BrnFieldA11yService, { optional: true });
    _computedDescribedBy = computed(() => {
        const manual = this.describedBy();
        const manualList = manual ? manual.split(/\s+/).filter(Boolean) : [];
        const fieldIds = this._a11y?.describedBy() ?? null;
        const fieldList = fieldIds ? fieldIds.split(/\s+/).filter(Boolean) : [];
        const combined = [...new Set([...manualList, ...fieldList])];
        return combined.length ? combined.join(' ') : null;
    }, ...(ngDevMode ? [{ debugName: "_computedDescribedBy" }] : []));
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnFieldControlDescribedBy, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.3.17", type: BrnFieldControlDescribedBy, isStandalone: true, selector: "[brnFieldControlDescribedBy]", inputs: { describedBy: { classPropertyName: "describedBy", publicName: "aria-describedby", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "attr.aria-describedby": "_computedDescribedBy()" } }, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnFieldControlDescribedBy, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnFieldControlDescribedBy]',
                    host: {
                        '[attr.aria-describedby]': '_computedDescribedBy()',
                    },
                }]
        }], propDecorators: { describedBy: [{ type: i0.Input, args: [{ isSignal: true, alias: "aria-describedby", required: false }] }] } });

const BrnFieldImports = [BrnField, BrnFieldControl, BrnFieldControlDescribedBy];

/**
 * Generated bundle index. Do not edit.
 */

export { BrnField, BrnFieldA11yService, BrnFieldControl, BrnFieldControlDescribedBy, BrnFieldImports, BrnLabelable, injectBrnLabelable, provideBrnLabelable };
