// @spartan-ng/brain/fesm2022/spartan-ng-brain-hover-card.mjs
import * as i0 from '@angular/core';
import { InjectionToken, inject, TemplateRef, Directive, NgZone, signal, Injectable, ViewContainerRef, ElementRef, input, numberAttribute, computed, effect, untracked, contentChild } from '@angular/core';
import { FocusMonitor } from '@angular/cdk/a11y';
import { Overlay, OverlayPositionBuilder } from '@angular/cdk/overlay';
import { TemplatePortal } from '@angular/cdk/portal';
import { toSignal } from '@angular/core/rxjs-interop';
import { provideExposedSideProviderExisting, provideExposesStateProviderExisting, createHoverObservable } from '@spartan-ng/brain/core';
import { Subject, BehaviorSubject, of, merge, fromEvent } from 'rxjs';
import { switchMap, filter, map, distinctUntilChanged, tap, delay, share, takeUntil } from 'rxjs/operators';

const defaultOptions = {
    showDelay: 300,
    hideDelay: 500,
    animationDelay: 100,
    sideOffset: 5,
    align: 'bottom',
};
const BRN_HOVER_CARD_DEFAULT_OPTIONS = new InjectionToken('brn-hover-card-default-options', {
    providedIn: 'root',
    factory: () => defaultOptions,
});
function provideBrnHoverCardDefaultOptions(options) {
    return { provide: BRN_HOVER_CARD_DEFAULT_OPTIONS, useValue: { ...defaultOptions, ...options } };
}
function injectBrnHoverCardDefaultOptions() {
    return inject(BRN_HOVER_CARD_DEFAULT_OPTIONS, { optional: true }) ?? defaultOptions;
}

class BrnHoverCardContent {
    _contentService = inject(BrnHoverCardContentService);
    state = this._contentService.state;
    side = this._contentService.side;
    template = inject(TemplateRef);
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnHoverCardContent, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "20.3.17", type: BrnHoverCardContent, isStandalone: true, selector: "[brnHoverCardContent]", providers: [
            provideExposedSideProviderExisting((() => BrnHoverCardContent)),
            provideExposesStateProviderExisting((() => BrnHoverCardContent)),
        ], exportAs: ["brnHoverCardContent"], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnHoverCardContent, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnHoverCardContent]',
                    exportAs: 'brnHoverCardContent',
                    providers: [
                        provideExposedSideProviderExisting((() => BrnHoverCardContent)),
                        provideExposesStateProviderExisting((() => BrnHoverCardContent)),
                    ],
                }]
        }] });
const topFirstPositions = [
    {
        originX: 'center',
        originY: 'top',
        overlayX: 'center',
        overlayY: 'bottom',
    },
    {
        originX: 'center',
        originY: 'bottom',
        overlayX: 'center',
        overlayY: 'top',
    },
];
const bottomFirstPositions = [
    {
        originX: 'center',
        originY: 'bottom',
        overlayX: 'center',
        overlayY: 'top',
    },
    {
        originX: 'center',
        originY: 'top',
        overlayX: 'center',
        overlayY: 'bottom',
    },
];
const leftFirstPositions = [
    // left side
    {
        originX: 'start',
        originY: 'center',
        overlayX: 'end',
        overlayY: 'center',
    },
    // Fallback 1: right (opposite horizontal)
    {
        originX: 'end',
        originY: 'center',
        overlayX: 'start',
        overlayY: 'center',
    },
    // Fallback 2: bottom (vertical fallback)
    {
        originX: 'center',
        originY: 'bottom',
        overlayX: 'center',
        overlayY: 'top',
    },
    // Fallback 3: top (vertical fallback)
    {
        originX: 'center',
        originY: 'top',
        overlayX: 'center',
        overlayY: 'bottom',
    },
];
const rightFirstPositions = [
    // right side
    {
        originX: 'end',
        originY: 'center',
        overlayX: 'start',
        overlayY: 'center',
    },
    // Fallback 1: left (opposite horizontal)
    {
        originX: 'start',
        originY: 'center',
        overlayX: 'end',
        overlayY: 'center',
    },
    // Fallback 2: bottom (vertical fallback)
    {
        originX: 'center',
        originY: 'bottom',
        overlayX: 'center',
        overlayY: 'top',
    },
    // Fallback 3: top (vertical fallback)
    {
        originX: 'center',
        originY: 'top',
        overlayX: 'center',
        overlayY: 'bottom',
    },
];
const POSITION_MAP = {
    top: topFirstPositions,
    bottom: bottomFirstPositions,
    left: leftFirstPositions,
    right: rightFirstPositions,
};
class BrnHoverCardContentService {
    _overlay = inject(Overlay);
    _zone = inject(NgZone);
    _psBuilder = inject(OverlayPositionBuilder);
    _content = signal(null, ...(ngDevMode ? [{ debugName: "_content" }] : []));
    _state = signal('closed', ...(ngDevMode ? [{ debugName: "_state" }] : []));
    _config = {};
    _overlayRef;
    _positionStrategy;
    _destroyed$ = new Subject();
    _positionChangesObservables$ = new BehaviorSubject(undefined);
    _overlayHoveredObservables$ = new BehaviorSubject(undefined);
    positionChanges$ = this._positionChangesObservables$.pipe(switchMap((positionChangeObservable) => (positionChangeObservable ? positionChangeObservable : of(undefined))), filter((change) => change !== undefined && change !== null));
    hovered$ = this._overlayHoveredObservables$.pipe(switchMap((overlayHoveredObservable) => (overlayHoveredObservable ? overlayHoveredObservable : of(false))));
    state = this._state.asReadonly();
    side = toSignal(this.positionChanges$.pipe(map((change) => 
    // todo: better translation or adjusting hlm to take that into account
    change.connectionPair.originY === 'center'
        ? change.connectionPair.originX === 'start'
            ? 'left'
            : 'right'
        : change.connectionPair.originY)), { initialValue: 'bottom' });
    setConfig(config) {
        this._config = config;
        if (config.attachTo) {
            const align = config.align ?? 'bottom';
            const positions = config.attachPositions ?? POSITION_MAP[align];
            this._positionStrategy = this._psBuilder.flexibleConnectedTo(config.attachTo).withPositions(positions);
            const offset = config.sideOffset ?? 0;
            if (align === 'left' || align === 'right') {
                this._positionStrategy.withDefaultOffsetX(align === 'left' ? -offset : offset);
            }
            else {
                this._positionStrategy.withDefaultOffsetY(align === 'top' ? -offset : offset);
            }
            this._config = {
                ...this._config,
                positionStrategy: this._positionStrategy,
                scrollStrategy: this._overlay.scrollStrategies.reposition(),
            };
            this._positionChangesObservables$.next(this._positionStrategy.positionChanges);
        }
        this._overlayRef = this._overlay.create(this._config);
    }
    setContent(value, vcr) {
        this._content.set(new TemplatePortal(value instanceof TemplateRef ? value : value.template, vcr));
        if (!this._overlayRef) {
            this._overlayRef = this._overlay.create(this._config);
        }
    }
    setState(newState) {
        this._state.set(newState);
    }
    show() {
        const content = this._content();
        if (!content || !this._overlayRef)
            return;
        this._overlayRef?.detach();
        this._overlayRef?.attach(content);
        this._destroyed$ = new Subject();
        this._overlayHoveredObservables$.next(createHoverObservable(this._overlayRef.hostElement, this._zone, this._destroyed$).pipe(map((e) => e.hover)));
    }
    hide() {
        this._overlayRef?.detach();
        this._destroyed$.next();
        this._destroyed$.complete();
        this._destroyed$ = new Subject();
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnHoverCardContentService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
    /** @nocollapse */ static ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnHoverCardContentService });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnHoverCardContentService, decorators: [{
            type: Injectable
        }] });
class BrnHoverCardTrigger {
    _destroy$ = new Subject();
    _vcr = inject(ViewContainerRef);
    _zone = inject(NgZone);
    _el = inject(ElementRef);
    _contentService = inject(BrnHoverCardContentService);
    _focusMonitor = inject(FocusMonitor);
    focused$ = this._focusMonitor.monitor(this._el).pipe(map((e) => e !== null));
    hovered$ = merge(fromEvent(this._el.nativeElement, 'click').pipe(map(() => false)), createHoverObservable(this._el.nativeElement, this._zone, this._destroy$).pipe(map((e) => e.hover)), this._contentService.hovered$, this.focused$).pipe(distinctUntilChanged());
    showing$ = this.hovered$.pipe(
    // we set the state to open here because we are about to open show the content
    tap((visible) => visible && this._contentService.setState('open')), switchMap((visible) => {
        // we are delaying based on the configure-able input
        return of(visible).pipe(delay(visible ? this.showDelay() : this.hideDelay()));
    }), switchMap((visible) => {
        // don't do anything when we are in the process of showing the content
        if (visible)
            return of(visible);
        // we set the state to closed here to trigger any animations for the element leaving
        this._contentService.setState('closed');
        // then delay to wait for the leaving animation to finish
        return of(visible).pipe(delay(this.animationDelay()));
    }), distinctUntilChanged(), share(), takeUntil(this._destroy$));
    _defaultOptions = injectBrnHoverCardDefaultOptions();
    showDelay = input(this._defaultOptions.showDelay, ...(ngDevMode ? [{ debugName: "showDelay", transform: numberAttribute }] : [{
            transform: numberAttribute,
        }]));
    hideDelay = input(this._defaultOptions.hideDelay, ...(ngDevMode ? [{ debugName: "hideDelay", transform: numberAttribute }] : [{
            transform: numberAttribute,
        }]));
    animationDelay = input(this._defaultOptions.animationDelay, ...(ngDevMode ? [{ debugName: "animationDelay", transform: numberAttribute }] : [{
            transform: numberAttribute,
        }]));
    sideOffset = input(this._defaultOptions.sideOffset, ...(ngDevMode ? [{ debugName: "sideOffset", transform: numberAttribute }] : [{
            transform: numberAttribute,
        }]));
    align = input(this._defaultOptions.align, ...(ngDevMode ? [{ debugName: "align" }] : []));
    brnHoverCardTriggerFor = input(undefined, ...(ngDevMode ? [{ debugName: "brnHoverCardTriggerFor" }] : []));
    mutableBrnHoverCardTriggerFor = computed(() => signal(this.brnHoverCardTriggerFor()), ...(ngDevMode ? [{ debugName: "mutableBrnHoverCardTriggerFor" }] : []));
    _brnHoverCardTriggerForState = computed(() => this.mutableBrnHoverCardTriggerFor()(), ...(ngDevMode ? [{ debugName: "_brnHoverCardTriggerForState" }] : []));
    constructor() {
        effect(() => {
            const value = this._brnHoverCardTriggerForState();
            untracked(() => {
                if (value) {
                    this._contentService.setContent(value, this._vcr);
                }
            });
        });
    }
    ngOnInit() {
        this._contentService.setConfig({ attachTo: this._el, align: this.align(), sideOffset: this.sideOffset() });
        this.showing$.subscribe((isHovered) => {
            if (isHovered) {
                this._contentService.show();
            }
            else {
                this._contentService.hide();
            }
        });
    }
    ngOnDestroy() {
        this._destroy$.next();
        this._destroy$.complete();
        this._focusMonitor.stopMonitoring(this._el);
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnHoverCardTrigger, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.3.17", type: BrnHoverCardTrigger, isStandalone: true, selector: "[brnHoverCardTrigger]:not(ng-container),[brnHoverCardTriggerFor]:not(ng-container)", inputs: { showDelay: { classPropertyName: "showDelay", publicName: "showDelay", isSignal: true, isRequired: false, transformFunction: null }, hideDelay: { classPropertyName: "hideDelay", publicName: "hideDelay", isSignal: true, isRequired: false, transformFunction: null }, animationDelay: { classPropertyName: "animationDelay", publicName: "animationDelay", isSignal: true, isRequired: false, transformFunction: null }, sideOffset: { classPropertyName: "sideOffset", publicName: "sideOffset", isSignal: true, isRequired: false, transformFunction: null }, align: { classPropertyName: "align", publicName: "align", isSignal: true, isRequired: false, transformFunction: null }, brnHoverCardTriggerFor: { classPropertyName: "brnHoverCardTriggerFor", publicName: "brnHoverCardTriggerFor", isSignal: true, isRequired: false, transformFunction: null } }, exportAs: ["brnHoverCardTrigger"], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnHoverCardTrigger, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnHoverCardTrigger]:not(ng-container),[brnHoverCardTriggerFor]:not(ng-container)',
                    exportAs: 'brnHoverCardTrigger',
                }]
        }], ctorParameters: () => [], propDecorators: { showDelay: [{ type: i0.Input, args: [{ isSignal: true, alias: "showDelay", required: false }] }], hideDelay: [{ type: i0.Input, args: [{ isSignal: true, alias: "hideDelay", required: false }] }], animationDelay: [{ type: i0.Input, args: [{ isSignal: true, alias: "animationDelay", required: false }] }], sideOffset: [{ type: i0.Input, args: [{ isSignal: true, alias: "sideOffset", required: false }] }], align: [{ type: i0.Input, args: [{ isSignal: true, alias: "align", required: false }] }], brnHoverCardTriggerFor: [{ type: i0.Input, args: [{ isSignal: true, alias: "brnHoverCardTriggerFor", required: false }] }] } });

class BrnHoverCard {
    _trigger = contentChild(BrnHoverCardTrigger, ...(ngDevMode ? [{ debugName: "_trigger" }] : []));
    _content = contentChild(BrnHoverCardContent, ...(ngDevMode ? [{ debugName: "_content" }] : []));
    ngAfterContentInit() {
        if (!this._trigger() || !this._content())
            return;
        this._trigger()?.mutableBrnHoverCardTriggerFor().set(this._content());
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnHoverCard, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.2.0", version: "20.3.17", type: BrnHoverCard, isStandalone: true, selector: "[brnHoverCard],brn-hover-card", providers: [BrnHoverCardContentService], queries: [{ propertyName: "_trigger", first: true, predicate: BrnHoverCardTrigger, descendants: true, isSignal: true }, { propertyName: "_content", first: true, predicate: BrnHoverCardContent, descendants: true, isSignal: true }], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnHoverCard, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnHoverCard],brn-hover-card',
                    providers: [BrnHoverCardContentService],
                }]
        }], propDecorators: { _trigger: [{ type: i0.ContentChild, args: [i0.forwardRef(() => BrnHoverCardTrigger), { isSignal: true }] }], _content: [{ type: i0.ContentChild, args: [i0.forwardRef(() => BrnHoverCardContent), { isSignal: true }] }] } });

const BrnHoverCardImports = [BrnHoverCard, BrnHoverCardContent, BrnHoverCardTrigger];

/**
 * Generated bundle index. Do not edit.
 */

export { BrnHoverCard, BrnHoverCardContent, BrnHoverCardContentService, BrnHoverCardImports, BrnHoverCardTrigger, defaultOptions, injectBrnHoverCardDefaultOptions, provideBrnHoverCardDefaultOptions };
