// @spartan-ng/brain/fesm2022/spartan-ng-brain-label.mjs
import * as i0 from '@angular/core';
import { inject, input, computed, Directive } from '@angular/core';
import { BrnField } from '@spartan-ng/brain/field';

class BrnLabel {
    static _id = 0;
    _brnField = inject(BrnField, { optional: true });
    /** The id of the label. */
    id = input(`brn-label-${++BrnLabel._id}`, ...(ngDevMode ? [{ debugName: "id" }] : []));
    /** The id of the form control this label is associated with. */
    for = input(...(ngDevMode ? [undefined, { debugName: "for" }] : []));
    _for = computed(() => {
        const forValue = this.for();
        if (forValue)
            return forValue;
        return this._brnField?.labelableId();
    }, ...(ngDevMode ? [{ debugName: "_for" }] : []));
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnLabel, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.3.17", type: BrnLabel, isStandalone: true, selector: "[brnLabel]", inputs: { id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null }, for: { classPropertyName: "for", publicName: "for", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "id": "id()", "attr.for": "_for()" } }, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnLabel, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnLabel]',
                    host: {
                        '[id]': 'id()',
                        '[attr.for]': '_for()',
                    },
                }]
        }], propDecorators: { id: [{ type: i0.Input, args: [{ isSignal: true, alias: "id", required: false }] }], for: [{ type: i0.Input, args: [{ isSignal: true, alias: "for", required: false }] }] } });

const BrnLabelImports = [BrnLabel];

/**
 * Generated bundle index. Do not edit.
 */

export { BrnLabel, BrnLabelImports };
