// @spartan-ng/brain/fesm2022/spartan-ng-brain-forms.mjs
import * as i0 from '@angular/core';
import { signal, computed, isSignal, Injectable } from '@angular/core';
import { NgForm } from '@angular/forms';

function controlStateEqual(a, b) {
    return (a === b ||
        (a != null &&
            b != null &&
            a.dirty === b.dirty &&
            a.invalid === b.invalid &&
            a.touched === b.touched &&
            a.spartanInvalid === b.spartanInvalid &&
            a.errors === b.errors));
}

class ReactiveStateTracker {
    _ngControl;
    _matcher;
    _parentFormGroup;
    _parentForm;
    _stateVersion = signal(0, ...(ngDevMode ? [{ debugName: "_stateVersion" }] : []));
    _eventsSubscription;
    controlState = computed(() => {
        const control = this._ngControl.control;
        if (!control)
            return null;
        this._stateVersion();
        const spartanInvalid = this._matcher?.isInvalid(control, this._controlParent) ?? false;
        return {
            dirty: control.dirty,
            errors: control.errors,
            invalid: control.invalid,
            spartanInvalid,
            touched: control.touched,
        };
    }, ...(ngDevMode ? [{ debugName: "controlState", equal: controlStateEqual }] : [{ equal: controlStateEqual }]));
    errors = computed(() => this.controlState()?.errors ?? null, ...(ngDevMode ? [{ debugName: "errors" }] : []));
    dirty = computed(() => this.controlState()?.dirty ?? null, ...(ngDevMode ? [{ debugName: "dirty" }] : []));
    invalid = computed(() => this.controlState()?.invalid ?? null, ...(ngDevMode ? [{ debugName: "invalid" }] : []));
    spartanInvalid = computed(() => this.controlState()?.spartanInvalid ?? null, ...(ngDevMode ? [{ debugName: "spartanInvalid" }] : []));
    touched = computed(() => this.controlState()?.touched ?? null, ...(ngDevMode ? [{ debugName: "touched" }] : []));
    get _controlParent() {
        return this._parentFormGroup || this._parentForm;
    }
    constructor(_ngControl, _matcher, _parentFormGroup, _parentForm) {
        this._ngControl = _ngControl;
        this._matcher = _matcher;
        this._parentFormGroup = _parentFormGroup;
        this._parentForm = _parentForm;
        const control = _ngControl.control;
        if (control) {
            this._eventsSubscription = control.events.subscribe(() => {
                this._stateVersion.update((v) => v + 1);
            });
        }
    }
    destroy() {
        this._eventsSubscription?.unsubscribe();
    }
}

class SignalStateTracker {
    _ngControl;
    _matcher;
    _parentFormGroup;
    _parentForm;
    _control = computed(() => {
        const control = this._ngControl.control;
        if (!control)
            return null;
        return control;
    }, ...(ngDevMode ? [{ debugName: "_control" }] : []));
    // With signal forms, AbstractControl is implemented by InteropNgControl, whose control state
    // properties (e.g. dirty, touched, invalid) are getter functions that internally read FormField state signals.
    // Accessing them inside a computed() therefore creates reactive signal dependencies automatically.
    // See: https://github.com/angular/angular/blob/39e382a756b552d2b7bd3ce2c364daee9d7a0056/packages/forms/signals/src/controls/interop_ng_control.ts#L68-L129
    dirty = computed(() => this._control()?.dirty ?? null, ...(ngDevMode ? [{ debugName: "dirty" }] : []));
    touched = computed(() => this._control()?.touched ?? null, ...(ngDevMode ? [{ debugName: "touched" }] : []));
    invalid = computed(() => this._control()?.invalid ?? null, ...(ngDevMode ? [{ debugName: "invalid" }] : []));
    errors = computed(() => this._control()?.errors ?? null, ...(ngDevMode ? [{ debugName: "errors" }] : []));
    spartanInvalid = computed(() => {
        const control = this._control();
        if (!control) {
            return null;
        }
        return this._matcher?.isInvalid(control, this._controlParent) ?? false;
    }, ...(ngDevMode ? [{ debugName: "spartanInvalid" }] : []));
    controlState = computed(() => {
        const dirty = this.dirty();
        const invalid = this.invalid();
        const touched = this.touched();
        const spartanInvalid = this.spartanInvalid();
        const errors = this.errors();
        if (dirty === null || invalid === null || touched === null || spartanInvalid === null)
            return null;
        return { dirty, errors, invalid, spartanInvalid, touched };
    }, ...(ngDevMode ? [{ debugName: "controlState", equal: controlStateEqual }] : [{ equal: controlStateEqual }]));
    get _controlParent() {
        return this._parentFormGroup || this._parentForm;
    }
    constructor(_ngControl, _matcher, _parentFormGroup, _parentForm) {
        this._ngControl = _ngControl;
        this._matcher = _matcher;
        this._parentFormGroup = _parentFormGroup;
        this._parentForm = _parentForm;
    }
    destroy() {
        // No subscriptions to clean up for signal-based controls
    }
}

function createStateTracker(ngControl, matcher, parentFormGroup, parentForm) {
    if (ngControl.control && 'field' in ngControl.control && isSignal(ngControl.control.field)) {
        return new SignalStateTracker(ngControl, matcher, parentFormGroup, parentForm);
    }
    return new ReactiveStateTracker(ngControl, matcher, parentFormGroup, parentForm);
}

/**
 * Provides a custom {@link ErrorStateMatcher} at the component or application level.
 *
 * @example <caption>Application-level (appConfig)</caption>
 * ```ts
 * provideErrorStateMatcher(ShowOnDirtyErrorStateMatcher)
 * ```
 *
 * @example <caption>Component-level</caption>
 * ```ts
 * providers: [provideErrorStateMatcher(ShowOnDirtyErrorStateMatcher)]
 * ```
 */
function provideErrorStateMatcher(matcher) {
    return { provide: ErrorStateMatcher, useClass: matcher };
}
/** Error state matcher that matches when a control is invalid and dirty. */
class ShowOnDirtyErrorStateMatcher {
    isInvalid(control, form) {
        return !!(control && control.invalid && (control.dirty || (form instanceof NgForm && form.submitted)));
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: ShowOnDirtyErrorStateMatcher, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
    /** @nocollapse */ static ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: ShowOnDirtyErrorStateMatcher });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: ShowOnDirtyErrorStateMatcher, decorators: [{
            type: Injectable
        }] });
/**
 * Determines when a form control should be considered in an invalid (error) state.
 *
 * The return value of `isInvalid` is reflected as the `data-matches-spartan-invalid` attribute
 * on field components, which drives both error styling and the visibility of `HlmFieldError`
 * messages.
 *
 * The default implementation matches when the control is invalid and either touched or, for
 * template-driven forms, the parent `NgForm` has been submitted.
 *
 * Provide a custom implementation (e.g. `ShowOnDirtyErrorStateMatcher`) at the component or
 * application level to change when errors are shown.
 */
class ErrorStateMatcher {
    isInvalid(control, form) {
        return !!(control && control.invalid && (control.touched || (form instanceof NgForm && form.submitted)));
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: ErrorStateMatcher, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
    /** @nocollapse */ static ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: ErrorStateMatcher, providedIn: 'root' });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: ErrorStateMatcher, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { ErrorStateMatcher, ReactiveStateTracker, ShowOnDirtyErrorStateMatcher, SignalStateTracker, controlStateEqual, createStateTracker, provideErrorStateMatcher };
