// @spartan-ng/brain/fesm2022/spartan-ng-brain-avatar.mjs
import * as i0 from '@angular/core';
import { signal, computed, Directive, contentChild, ChangeDetectionStrategy, Component, Pipe } from '@angular/core';

class BrnAvatarImage {
    _loaded = signal(false, ...(ngDevMode ? [{ debugName: "_loaded" }] : []));
    _onError() {
        this._loaded.set(false);
    }
    _onLoad() {
        this._loaded.set(true);
    }
    canShow = computed(() => this._loaded(), ...(ngDevMode ? [{ debugName: "canShow" }] : []));
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnAvatarImage, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "20.3.17", type: BrnAvatarImage, isStandalone: true, selector: "img[brnAvatarImage]", host: { listeners: { "load": "_onLoad()", "error": "_onError()" } }, exportAs: ["avatarImage"], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnAvatarImage, decorators: [{
            type: Directive,
            args: [{
                    selector: 'img[brnAvatarImage]',
                    exportAs: 'avatarImage',
                    host: {
                        '(load)': '_onLoad()',
                        '(error)': '_onError()',
                    },
                }]
        }] });

class BrnAvatar {
    _image = contentChild(BrnAvatarImage, ...(ngDevMode ? [{ debugName: "_image" }] : []));
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnAvatar, deps: [], target: i0.ɵɵFactoryTarget.Component });
    /** @nocollapse */ static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "20.3.17", type: BrnAvatar, isStandalone: true, selector: "brn-avatar", queries: [{ propertyName: "_image", first: true, predicate: BrnAvatarImage, descendants: true, isSignal: true }], ngImport: i0, template: `
		@if (_image()?.canShow()) {
			<ng-content select="[brnAvatarImage]" />
		} @else {
			<ng-content select="[brnAvatarFallback]" />
		}
	`, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnAvatar, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-avatar',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    template: `
		@if (_image()?.canShow()) {
			<ng-content select="[brnAvatarImage]" />
		} @else {
			<ng-content select="[brnAvatarFallback]" />
		}
	`,
                }]
        }], propDecorators: { _image: [{ type: i0.ContentChild, args: [i0.forwardRef(() => BrnAvatarImage), { isSignal: true }] }] } });

class BrnAvatarFallback {
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnAvatarFallback, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "20.3.17", type: BrnAvatarFallback, isStandalone: true, selector: "[brnAvatarFallback]", exportAs: ["avatarFallback"], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnAvatarFallback, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnAvatarFallback]',
                    exportAs: 'avatarFallback',
                }]
        }] });

function hashString(str) {
    let h;
    for (let i = 0; i < str.length; i++)
        h = (Math.imul(31, h || 0) + str.charCodeAt(i)) | 0;
    return h || 0;
}
function hashManyTimes(times, str) {
    let h = hashString(str);
    for (let i = 0; i < times; i++)
        h = hashString(String(h));
    return h;
}
function hexColorFor(str) {
    const hash = str.length <= 2 ? hashManyTimes(5, str) : hashString(str);
    let color = '#';
    for (let i = 0; i < 3; i += 1) {
        const value = (hash >> (i * 8)) & 0xff;
        color += `00${value.toString(16)}`.slice(-2);
    }
    return color;
}

const toInitial = (capitalize = true) => (word) => {
    const initial = word.charAt(0);
    return capitalize ? initial.toLocaleUpperCase() : initial;
};
const firstAndLast = (initials) => `${initials[0]}${initials[initials.length - 1]}`;
class InitialsPipe {
    transform(name, capitalize = true, firstAndLastOnly = true, delimiter = ' ') {
        if (!name)
            return '';
        const initials = name.trim().split(delimiter).filter(Boolean).map(toInitial(capitalize));
        if (firstAndLastOnly && initials.length > 1)
            return firstAndLast(initials);
        return initials.join('');
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: InitialsPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
    /** @nocollapse */ static ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "20.3.17", ngImport: i0, type: InitialsPipe, isStandalone: true, name: "initials" });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: InitialsPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'initials',
                }]
        }] });

const isShortHand = (hex) => hex.length === 3;
const cleanup = (hex) => {
    const noHash = hex.replace('#', '').trim().toLowerCase();
    if (!isShortHand(noHash))
        return noHash;
    return noHash
        .split('')
        .map((char) => char + char)
        .join('');
};
const isBright = (hex) => Number.parseInt(cleanup(hex), 16) > 0xffffff / 1.25;

const BrnAvatarImports = [BrnAvatar, BrnAvatarFallback, BrnAvatarImage];

/**
 * Generated bundle index. Do not edit.
 */

export { BrnAvatar, BrnAvatarFallback, BrnAvatarImage, BrnAvatarImports, InitialsPipe, hexColorFor, isBright };
