// @spartan-ng/brain/fesm2022/spartan-ng-brain-progress.mjs
import * as i0 from '@angular/core';
import { InjectionToken, inject, input, numberAttribute, computed, Directive } from '@angular/core';

const BrnProgressToken = new InjectionToken('BrnProgressComponent');
function provideBrnProgress(progress) {
    return { provide: BrnProgressToken, useExisting: progress };
}
function injectBrnProgress() {
    return inject(BrnProgressToken);
}

class BrnProgress {
    /**
     * The current progress value.
     */
    value = input(undefined, ...(ngDevMode ? [{ debugName: "value", transform: (value) => (value === undefined || value === null ? undefined : Number(value)) }] : [{
            transform: (value) => (value === undefined || value === null ? undefined : Number(value)),
        }]));
    /**
     * The maximum progress value.
     */
    max = input(100, ...(ngDevMode ? [{ debugName: "max", transform: numberAttribute }] : [{ transform: numberAttribute }]));
    /**
     * A function that returns the label for the current progress value.
     */
    getValueLabel = input((value, max) => `${Math.round((value / max) * 100)}%`, ...(ngDevMode ? [{ debugName: "getValueLabel" }] : []));
    _label = computed(() => {
        const value = this.value();
        return value === null || value === undefined ? undefined : this.getValueLabel()(value, this.max());
    }, ...(ngDevMode ? [{ debugName: "_label" }] : []));
    state = computed(() => {
        const value = this.value();
        const max = this.max();
        return value === null || value === undefined ? 'indeterminate' : value === max ? 'complete' : 'loading';
    }, ...(ngDevMode ? [{ debugName: "state" }] : []));
    ngOnChanges(changes) {
        if ('value' in changes || 'max' in changes) {
            this.validate();
        }
    }
    validate() {
        // validate that the value is within the bounds of the max
        const value = this.value();
        const max = this.max();
        if (value === null || value === undefined) {
            return;
        }
        if (value > max || value < 0) {
            throw Error('Value must be 0 or greater and less or equal to max');
        }
        if (max < 0) {
            throw Error('max must be greater than 0');
        }
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnProgress, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.3.17", type: BrnProgress, isStandalone: true, selector: "brn-progress", inputs: { value: { classPropertyName: "value", publicName: "value", isSignal: true, isRequired: false, transformFunction: null }, max: { classPropertyName: "max", publicName: "max", isSignal: true, isRequired: false, transformFunction: null }, getValueLabel: { classPropertyName: "getValueLabel", publicName: "getValueLabel", isSignal: true, isRequired: false, transformFunction: null } }, host: { attributes: { "role": "progressbar" }, properties: { "attr.aria-valuemax": "max()", "attr.aria-valuemin": "0", "attr.aria-valuenow": "value()", "attr.aria-valuetext": "_label()", "attr.data-state": "state()", "attr.data-value": "value()", "attr.data-max": "max()" } }, providers: [provideBrnProgress(BrnProgress)], exportAs: ["brnProgress"], usesOnChanges: true, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnProgress, decorators: [{
            type: Directive,
            args: [{
                    selector: 'brn-progress',
                    exportAs: 'brnProgress',
                    providers: [provideBrnProgress(BrnProgress)],
                    host: {
                        role: 'progressbar',
                        '[attr.aria-valuemax]': 'max()',
                        '[attr.aria-valuemin]': '0',
                        '[attr.aria-valuenow]': 'value()',
                        '[attr.aria-valuetext]': '_label()',
                        '[attr.data-state]': 'state()',
                        '[attr.data-value]': 'value()',
                        '[attr.data-max]': 'max()',
                    },
                }]
        }], propDecorators: { value: [{ type: i0.Input, args: [{ isSignal: true, alias: "value", required: false }] }], max: [{ type: i0.Input, args: [{ isSignal: true, alias: "max", required: false }] }], getValueLabel: [{ type: i0.Input, args: [{ isSignal: true, alias: "getValueLabel", required: false }] }] } });

/* eslint-disable @angular-eslint/directive-selector */
class BrnProgressIndicator {
    _progress = injectBrnProgress();
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnProgressIndicator, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "20.3.17", type: BrnProgressIndicator, isStandalone: true, selector: "brn-progress-indicator", host: { properties: { "attr.data-state": "_progress.state()", "attr.data-value": "_progress.value()", "attr.data-max": "_progress.max()" } }, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnProgressIndicator, decorators: [{
            type: Directive,
            args: [{
                    selector: 'brn-progress-indicator',
                    host: {
                        '[attr.data-state]': '_progress.state()',
                        '[attr.data-value]': '_progress.value()',
                        '[attr.data-max]': '_progress.max()',
                    },
                }]
        }] });

const BrnProgressImports = [BrnProgress, BrnProgressIndicator];

/**
 * Generated bundle index. Do not edit.
 */

export { BrnProgress, BrnProgressImports, BrnProgressIndicator, injectBrnProgress };
