// @spartan-ng/brain/fesm2022/spartan-ng-brain-button.mjs
import * as i0 from '@angular/core';
import { input, booleanAttribute, inject, HOST_TAG_NAME, ElementRef, Directive } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { fromEvent } from 'rxjs';
import { filter } from 'rxjs/operators';

class BrnButton {
    disabled = input(false, ...(ngDevMode ? [{ debugName: "disabled", transform: booleanAttribute }] : [{ transform: booleanAttribute }]));
    _isAnchor = inject(HOST_TAG_NAME) === 'a';
    _elementRef = inject(ElementRef);
    constructor() {
        if (this._isAnchor) {
            fromEvent(this._elementRef.nativeElement, 'click')
                .pipe(filter(() => this.disabled()), takeUntilDestroyed())
                .subscribe((event) => {
                event.preventDefault();
                event.stopImmediatePropagation();
            });
        }
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnButton, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.3.17", type: BrnButton, isStandalone: true, selector: "a[brnButton], button[brnButton]", inputs: { disabled: { classPropertyName: "disabled", publicName: "disabled", isSignal: true, isRequired: false, transformFunction: null } }, host: { properties: { "attr.tabindex": "disabled() ? -1 : undefined", "attr.disabled": "!_isAnchor && disabled() || null", "attr.data-disabled": "disabled() || null" } }, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnButton, decorators: [{
            type: Directive,
            args: [{
                    selector: 'a[brnButton], button[brnButton]',
                    host: {
                        '[attr.tabindex]': 'disabled() ? -1 : undefined',
                        '[attr.disabled]': '!_isAnchor && disabled() || null',
                        '[attr.data-disabled]': 'disabled() || null',
                    },
                }]
        }], ctorParameters: () => [], propDecorators: { disabled: [{ type: i0.Input, args: [{ isSignal: true, alias: "disabled", required: false }] }] } });

const BrnButtonImports = [BrnButton];

/**
 * Generated bundle index. Do not edit.
 */

export { BrnButton, BrnButtonImports };
