// @spartan-ng/brain/fesm2022/spartan-ng-brain-slider.mjs
import { Directionality } from '@angular/cdk/bidi';
import * as i0 from '@angular/core';
import { InjectionToken, inject, forwardRef, Injector, input, model, numberAttribute, booleanAttribute, output, computed, signal, linkedSignal, Directive, ElementRef, PLATFORM_ID, DestroyRef, TemplateRef, Renderer2, ViewContainerRef, effect } from '@angular/core';
import { NG_VALUE_ACCESSOR, NgControl, NgModel } from '@angular/forms';
import * as i1 from '@spartan-ng/brain/field';
import { BrnFieldControl } from '@spartan-ng/brain/field';
import { isPlatformServer } from '@angular/common';
import { injectElementSize } from '@spartan-ng/brain/core';

const BrnSliderToken = new InjectionToken('BrnSliderToken');
function provideBrnSlider(slider) {
    return { provide: BrnSliderToken, useExisting: slider };
}
function injectBrnSlider() {
    return inject(BrnSliderToken);
}

const BRN_SLIDER_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(() => BrnSlider),
    multi: true,
};
let nextId = 0;
class BrnSlider {
    _dir = inject(Directionality);
    _injector = inject(Injector);
    _fieldControl = inject(BrnFieldControl);
    ngControl = null;
    _ariaInvalid = this._fieldControl.invalid;
    _dirty = this._fieldControl.dirty;
    _touched = this._fieldControl.touched;
    _spartanInvalid = this._fieldControl.spartanInvalid;
    /** Unique identifier for the slider element. Auto-generated if not provided. */
    id = input(`brn-slider-${++nextId}`, ...(ngDevMode ? [{ debugName: "id" }] : []));
    /** Accessibility label for the slider. Forwarded to all thumbs. */
    ariaLabel = input(null, ...(ngDevMode ? [{ debugName: "ariaLabel", alias: 'aria-label' }] : [{ alias: 'aria-label' }]));
    /** ID of the element that labels this slider for accessibility. Forwarded to all thumbs. */
    ariaLabelledby = input(null, ...(ngDevMode ? [{ debugName: "ariaLabelledby", alias: 'aria-labelledby' }] : [{ alias: 'aria-labelledby' }]));
    /**
     * The current slider value(s).
     *
     * For single-thumb sliders, this contains one value.
     * For range sliders, values are kept sorted in ascending order.
     */
    value = model([], ...(ngDevMode ? [{ debugName: "value" }] : []));
    /** Minimum allowed slider value. */
    min = input(0, ...(ngDevMode ? [{ debugName: "min", transform: numberAttribute }] : [{
            transform: numberAttribute,
        }]));
    /** Maximum allowed slider value. */
    max = input(100, ...(ngDevMode ? [{ debugName: "max", transform: numberAttribute }] : [{
            transform: numberAttribute,
        }]));
    /** Step increment used when changing values. */
    step = input(1, ...(ngDevMode ? [{ debugName: "step", transform: numberAttribute }] : [{
            transform: numberAttribute,
        }]));
    /** Minimum number of steps required between thumbs in a range slider. */
    minStepsBetweenThumbs = input(0, ...(ngDevMode ? [{ debugName: "minStepsBetweenThumbs", transform: numberAttribute }] : [{
            transform: numberAttribute,
        }]));
    /** Whether the slider is disabled. */
    disabled = input(false, ...(ngDevMode ? [{ debugName: "disabled", transform: booleanAttribute }] : [{
            transform: booleanAttribute,
        }]));
    /** Whether the slider direction is inverted. */
    inverted = input(false, ...(ngDevMode ? [{ debugName: "inverted", transform: booleanAttribute }] : [{
            transform: booleanAttribute,
        }]));
    /** Slider orientation. */
    orientation = input('horizontal', ...(ngDevMode ? [{ debugName: "orientation" }] : []));
    /** Whether tick marks should be displayed. */
    showTicks = input(false, ...(ngDevMode ? [{ debugName: "showTicks", transform: booleanAttribute }] : [{
            transform: booleanAttribute,
        }]));
    /** Maximum number of ticks to render. Excess ticks are evenly distributed. */
    maxTicks = input(25, ...(ngDevMode ? [{ debugName: "maxTicks", transform: numberAttribute }] : [{
            transform: numberAttribute,
        }]));
    /** Interval at which tick labels are shown. A value of `2` shows a label every second tick. */
    tickLabelInterval = input(2, ...(ngDevMode ? [{ debugName: "tickLabelInterval", transform: numberAttribute }] : [{
            transform: numberAttribute,
        }]));
    /** Defines how the tick should be displayed in the UI. */
    formatTick = input((tick) => tick.toString(), ...(ngDevMode ? [{ debugName: "formatTick" }] : []));
    /** Whether dragging the selected range should move all thumbs together. */
    draggableRange = input(false, ...(ngDevMode ? [{ debugName: "draggableRange", transform: booleanAttribute }] : [{
            transform: booleanAttribute,
        }]));
    /** Whether only dragging the range should work (overrides normal track clicks). */
    draggableRangeOnly = input(false, ...(ngDevMode ? [{ debugName: "draggableRangeOnly", transform: booleanAttribute }] : [{
            transform: booleanAttribute,
        }]));
    /** Emits when the value changes. */
    valueChange = output();
    /** @internal Normalized slider values. Values are clamped to `[min, max]` and sorted in ascending order. */
    normalizedValue = computed(() => [...this.value()].sort((a, b) => a - b).map((v) => clamp(v, [this.min(), this.max()])), ...(ngDevMode ? [{ debugName: "normalizedValue", equal: areArrsEqual }] : [{ equal: areArrsEqual }]));
    /** Indexes for all active thumbs. */
    thumbIndexes = computed(() => Array.from({ length: this.normalizedValue().length }, (_, i) => i), ...(ngDevMode ? [{ debugName: "thumbIndexes", equal: (a, b) => a.length === b.length }] : [{
            equal: (a, b) => a.length === b.length,
        }]));
    /** @internal Whether the slider is in range mode and draggable range is enabled */
    isDraggableRange = computed(() => this.draggableRange() && this.normalizedValue().length > 1, ...(ngDevMode ? [{ debugName: "isDraggableRange" }] : []));
    /** @internal Whether dragging should only move the range. */
    isDraggableRangeOnly = computed(() => this.draggableRangeOnly() && this.isDraggableRange(), ...(ngDevMode ? [{ debugName: "isDraggableRangeOnly" }] : []));
    _direction = this._dir.valueSignal;
    /** @internal Logical edge from which the slider value increases. */
    slidingSource = computed(() => {
        const orientation = this.orientation();
        const inverted = this.inverted();
        if (orientation === 'vertical') {
            return inverted ? 'top' : 'bottom';
        }
        const isLtr = this._direction() === 'ltr';
        const fromLeft = (isLtr && !inverted) || (!isLtr && inverted);
        return fromLeft ? 'left' : 'right';
    }, ...(ngDevMode ? [{ debugName: "slidingSource" }] : []));
    /** @internal */
    isHorizontal = computed(() => this.orientation() === 'horizontal', ...(ngDevMode ? [{ debugName: "isHorizontal" }] : []));
    /** @internal Reference to the slider track instance. */
    track = signal(null, ...(ngDevMode ? [{ debugName: "track" }] : []));
    /** @internal Reference to the slider range instance. */
    range = signal(null, ...(ngDevMode ? [{ debugName: "range" }] : []));
    /** @internal All registered slider thumbs. */
    thumbs = signal([], ...(ngDevMode ? [{ debugName: "thumbs" }] : []));
    /** @internal Index of the thumb currently being updated. */
    valueIndexToChange = signal(0, ...(ngDevMode ? [{ debugName: "valueIndexToChange" }] : []));
    /** @internal Visible tick values after applying density reduction. */
    ticks = computed(() => {
        if (!this.showTicks()) {
            return [];
        }
        const min = this.min();
        const max = this.max();
        const step = this.step();
        const maxTicks = this.maxTicks();
        if (step <= 0 || min > max) {
            return [];
        }
        const totalCount = Math.floor((max - min) / step) + 1;
        // No need to reduce
        if (totalCount <= maxTicks) {
            return Array.from({ length: totalCount }, (_, i) => min + i * step);
        }
        const stride = Math.ceil(totalCount / maxTicks);
        return Array.from({ length: totalCount }, (_, i) => min + i * step).filter((_, index) => index % stride === 0);
    }, ...(ngDevMode ? [{ debugName: "ticks" }] : []));
    /** @internal Disabled state that can be controlled internally or externally. */
    mutableDisabled = linkedSignal(() => this.disabled(), ...(ngDevMode ? [{ debugName: "mutableDisabled" }] : []));
    /** @internal Store the on change callback */
    _onChange;
    /** @internal Store the on touched callback */
    _onTouched;
    ngOnInit() {
        this.ngControl = this._injector.get(NgControl, null);
        // If bound to an Angular form control, writeValue() will run after ngOnInit,
        // so avoid initializing defaults here to prevent a transient min-value override.
        if (!this.ngControl) {
            if (!this.value().length) {
                const defaultValue = [this.min()];
                this.value.set(defaultValue);
            }
            const normalizedValue = this.value()
                .map((v) => clamp(v, [this.min(), this.max()]))
                .sort((a, b) => a - b);
            if (!areArrsEqual(normalizedValue, this.value())) {
                this.value.set(normalizedValue);
            }
        }
    }
    registerOnChange(fn) {
        this._onChange = fn;
    }
    registerOnTouched(fn) {
        this._onTouched = fn;
    }
    setDisabledState(isDisabled) {
        this.mutableDisabled.set(isDisabled);
    }
    writeValue(value) {
        if (!Array.isArray(value))
            return;
        if (this.ngControl instanceof NgModel && !this._onChange) {
            // Avoid phantom call for ngModel
            // https://github.com/angular/angular/issues/14988#issuecomment-2946355465
            return;
        }
        const newValue = [...value].sort((a, b) => a - b).map((v) => clamp(v, [this.min(), this.max()]));
        this.value.set(newValue);
    }
    /** Sets a new value for the slider at the given thumb index. */
    setValue(value, atIndex) {
        const decimalCount = getDecimalCount(this.step());
        const snapToStep = roundValue(Math.round((value - this.min()) / this.step()) * this.step() + this.min(), decimalCount);
        value = clamp(snapToStep, [this.min(), this.max()]);
        const newValue = [...this.normalizedValue()];
        newValue[atIndex] = value;
        newValue.sort((a, b) => a - b);
        if (!hasMinStepsBetweenValues(newValue, this.minStepsBetweenThumbs() * this.step()))
            return;
        const newValIndex = newValue.findIndex((val) => val === value);
        this.valueIndexToChange.set(newValIndex);
        if (areArrsEqual(newValue, this.value()))
            return;
        this.value.set(newValue);
        this._onChange?.(newValue);
        this.valueChange.emit(newValue);
        if (this.thumbs()[newValIndex]) {
            this.thumbs()[newValIndex].elementRef.nativeElement.focus();
        }
    }
    /** Moves the entire range by a delta value, snapping to step and preserving spacing. */
    setAllValuesByDelta(delta) {
        const current = this.normalizedValue();
        if (!current.length || delta === 0)
            return;
        const min = this.min();
        const max = this.max();
        const step = this.step();
        const decimalCount = getDecimalCount(step);
        const snappedDelta = roundValue(Math.round(delta / step) * step, decimalCount);
        if (snappedDelta === 0)
            return;
        let next = current.map((v) => v + snappedDelta);
        const rangeMin = Math.min(...next);
        const rangeMax = Math.max(...next);
        if (rangeMin < min) {
            const offset = min - rangeMin;
            next = next.map((v) => v + offset);
        }
        else if (rangeMax > max) {
            const offset = max - rangeMax;
            next = next.map((v) => v + offset);
        }
        next = next.map((v) => roundValue(v, decimalCount));
        if (areArrsEqual(next, this.value()))
            return;
        this.value.set(next);
        this._onChange?.(next);
        this.valueChange.emit(next);
    }
    /** @internal */
    addThumb(thumb) {
        this.thumbs.update((thumbs) => {
            thumbs.push(thumb);
            return [...thumbs];
        });
    }
    /** @internal */
    removeThumb(thumb) {
        this.thumbs.update((thumbs) => thumbs.filter((t) => t !== thumb));
    }
    _onFocusOut(event) {
        const currentTarget = event.currentTarget;
        const focusedEl = event.relatedTarget;
        if (!currentTarget.contains(focusedEl)) {
            this._onTouched?.();
        }
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSlider, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.3.17", type: BrnSlider, isStandalone: true, selector: "[brnSlider]", inputs: { id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null }, ariaLabel: { classPropertyName: "ariaLabel", publicName: "aria-label", isSignal: true, isRequired: false, transformFunction: null }, ariaLabelledby: { classPropertyName: "ariaLabelledby", publicName: "aria-labelledby", isSignal: true, isRequired: false, transformFunction: null }, value: { classPropertyName: "value", publicName: "value", isSignal: true, isRequired: false, transformFunction: null }, min: { classPropertyName: "min", publicName: "min", isSignal: true, isRequired: false, transformFunction: null }, max: { classPropertyName: "max", publicName: "max", isSignal: true, isRequired: false, transformFunction: null }, step: { classPropertyName: "step", publicName: "step", isSignal: true, isRequired: false, transformFunction: null }, minStepsBetweenThumbs: { classPropertyName: "minStepsBetweenThumbs", publicName: "minStepsBetweenThumbs", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "disabled", isSignal: true, isRequired: false, transformFunction: null }, inverted: { classPropertyName: "inverted", publicName: "inverted", isSignal: true, isRequired: false, transformFunction: null }, orientation: { classPropertyName: "orientation", publicName: "orientation", isSignal: true, isRequired: false, transformFunction: null }, showTicks: { classPropertyName: "showTicks", publicName: "showTicks", isSignal: true, isRequired: false, transformFunction: null }, maxTicks: { classPropertyName: "maxTicks", publicName: "maxTicks", isSignal: true, isRequired: false, transformFunction: null }, tickLabelInterval: { classPropertyName: "tickLabelInterval", publicName: "tickLabelInterval", isSignal: true, isRequired: false, transformFunction: null }, formatTick: { classPropertyName: "formatTick", publicName: "formatTick", isSignal: true, isRequired: false, transformFunction: null }, draggableRange: { classPropertyName: "draggableRange", publicName: "draggableRange", isSignal: true, isRequired: false, transformFunction: null }, draggableRangeOnly: { classPropertyName: "draggableRangeOnly", publicName: "draggableRangeOnly", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { value: "valueChange", valueChange: "valueChange" }, host: { attributes: { "data-slot": "slider" }, listeners: { "focusout": "_onFocusOut($event)" }, properties: { "attr.id": "id()", "attr.dir": "_direction()", "attr.aria-disabled": "mutableDisabled() ? \"true\" : null", "attr.data-disabled": "mutableDisabled() ? \"\" : null", "attr.data-inverted": "inverted() ? \"\" : null", "attr.data-orientation": "orientation()", "attr.aria-invalid": "_ariaInvalid() ? \"true\" : null", "attr.data-invalid": "_ariaInvalid() ? \"true\" : null", "attr.data-matches-spartan-invalid": "_ariaInvalid() ? \"true\" : null", "attr.data-dirty": "_dirty() ? \"true\" : null", "attr.data-touched": "_touched() ? \"true\" : null" } }, providers: [BRN_SLIDER_VALUE_ACCESSOR, provideBrnSlider(BrnSlider)], exportAs: ["brnSlider"], hostDirectives: [{ directive: i1.BrnFieldControl }], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSlider, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnSlider]',
                    exportAs: 'brnSlider',
                    providers: [BRN_SLIDER_VALUE_ACCESSOR, provideBrnSlider(BrnSlider)],
                    hostDirectives: [BrnFieldControl],
                    host: {
                        '[attr.id]': 'id()',
                        '[attr.dir]': '_direction()',
                        '[attr.aria-disabled]': 'mutableDisabled() ? "true" : null',
                        '[attr.data-disabled]': 'mutableDisabled() ? "" : null',
                        '[attr.data-inverted]': 'inverted() ? "" : null',
                        '[attr.data-orientation]': 'orientation()',
                        'data-slot': 'slider',
                        '(focusout)': '_onFocusOut($event)',
                        '[attr.aria-invalid]': '_ariaInvalid() ? "true" : null',
                        '[attr.data-invalid]': '_ariaInvalid() ? "true" : null',
                        '[attr.data-matches-spartan-invalid]': '_ariaInvalid() ? "true" : null',
                        '[attr.data-dirty]': '_dirty() ? "true" : null',
                        '[attr.data-touched]': '_touched() ? "true" : null',
                    },
                }]
        }], propDecorators: { id: [{ type: i0.Input, args: [{ isSignal: true, alias: "id", required: false }] }], ariaLabel: [{ type: i0.Input, args: [{ isSignal: true, alias: "aria-label", required: false }] }], ariaLabelledby: [{ type: i0.Input, args: [{ isSignal: true, alias: "aria-labelledby", required: false }] }], value: [{ type: i0.Input, args: [{ isSignal: true, alias: "value", required: false }] }, { type: i0.Output, args: ["valueChange"] }], min: [{ type: i0.Input, args: [{ isSignal: true, alias: "min", required: false }] }], max: [{ type: i0.Input, args: [{ isSignal: true, alias: "max", required: false }] }], step: [{ type: i0.Input, args: [{ isSignal: true, alias: "step", required: false }] }], minStepsBetweenThumbs: [{ type: i0.Input, args: [{ isSignal: true, alias: "minStepsBetweenThumbs", required: false }] }], disabled: [{ type: i0.Input, args: [{ isSignal: true, alias: "disabled", required: false }] }], inverted: [{ type: i0.Input, args: [{ isSignal: true, alias: "inverted", required: false }] }], orientation: [{ type: i0.Input, args: [{ isSignal: true, alias: "orientation", required: false }] }], showTicks: [{ type: i0.Input, args: [{ isSignal: true, alias: "showTicks", required: false }] }], maxTicks: [{ type: i0.Input, args: [{ isSignal: true, alias: "maxTicks", required: false }] }], tickLabelInterval: [{ type: i0.Input, args: [{ isSignal: true, alias: "tickLabelInterval", required: false }] }], formatTick: [{ type: i0.Input, args: [{ isSignal: true, alias: "formatTick", required: false }] }], draggableRange: [{ type: i0.Input, args: [{ isSignal: true, alias: "draggableRange", required: false }] }], draggableRangeOnly: [{ type: i0.Input, args: [{ isSignal: true, alias: "draggableRangeOnly", required: false }] }], valueChange: [{ type: i0.Output, args: ["valueChange"] }] } });
function areArrsEqual(arr1, arr2) {
    return String(arr1) === String(arr2);
}
function roundValue(value, decimalCount) {
    const rounder = Math.pow(10, decimalCount);
    return Math.round(value * rounder) / rounder;
}
function getDecimalCount(value) {
    return (String(value).split('.')[1] || '').length;
}
function clamp(value, [min, max]) {
    return Math.min(max, Math.max(min, value));
}
function getStepsBetweenValues(values) {
    return values.slice(0, -1).map((value, index) => values[index + 1] - value);
}
function hasMinStepsBetweenValues(values, minStepsBetweenValues) {
    if (minStepsBetweenValues > 0) {
        const stepsBetweenValues = getStepsBetweenValues(values);
        const actualMinStepsBetweenValues = Math.min(...stepsBetweenValues);
        return actualMinStepsBetweenValues >= minStepsBetweenValues;
    }
    return true;
}

class BrnSliderRange {
    _slider = injectBrnSlider();
    elementRef = inject(ElementRef);
    constructor() {
        this._slider.range.set(this);
    }
    _rangeInset = computed(() => {
        const thumbs = this._slider.thumbs();
        if (!thumbs.length)
            return;
        const inverted = this._slider.inverted();
        const isHorizontal = this._slider.isHorizontal();
        const start = thumbs[0].percentage();
        const end = thumbs[thumbs.length - 1].percentage();
        const inset = thumbs.length > 1 ? [`${start}%`, `${100 - end}%`] : ['0px', `${100 - start}%`];
        if (!isHorizontal) {
            inset.reverse();
        }
        if (inverted) {
            inset.reverse();
        }
        return inset.join(' ');
    }, ...(ngDevMode ? [{ debugName: "_rangeInset" }] : []));
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSliderRange, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "20.3.17", type: BrnSliderRange, isStandalone: true, selector: "[brnSliderRange]", host: { attributes: { "data-slot": "slider-range" }, properties: { "attr.data-disabled": "_slider.mutableDisabled() ? \"\" : null", "attr.data-orientation": "_slider.orientation()", "attr.data-draggable-range": "_slider.isDraggableRange() ? \"\" : null", "style.inset-inline": "_slider.isHorizontal() ? _rangeInset() : undefined", "style.inset-block": "!_slider.isHorizontal() ? _rangeInset() : undefined" } }, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSliderRange, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnSliderRange]',
                    host: {
                        '[attr.data-disabled]': '_slider.mutableDisabled() ? "" : null',
                        '[attr.data-orientation]': '_slider.orientation()',
                        '[attr.data-draggable-range]': '_slider.isDraggableRange() ? "" : null',
                        'data-slot': 'slider-range',
                        '[style.inset-inline]': '_slider.isHorizontal() ? _rangeInset() : undefined',
                        '[style.inset-block]': '!_slider.isHorizontal() ? _rangeInset() : undefined',
                    },
                }]
        }], ctorParameters: () => [] });

function linearScale(input, output) {
    return (value) => {
        if (input[0] === input[1] || output[0] === output[1])
            return output[0];
        const ratio = (output[1] - output[0]) / (input[1] - input[0]);
        return output[0] + ratio * (value - input[0]);
    };
}

const PAGE_KEYS = ['PageUp', 'PageDown'];
class BrnSliderThumb {
    _platform = inject(PLATFORM_ID);
    _destroyRef = inject(DestroyRef);
    _size = injectElementSize();
    _slider = injectBrnSlider();
    elementRef = inject(ElementRef);
    /**
     * Indicates whether the thumb is ready to be displayed and interactable.
     *
     * This signal returns `true` when the element's size has been measured
     * and the width is greater than 0. It ensures that the thumb is not
     * rendered or interacted with prematurely, which prevents layout
     * issues or first-frame “jerks” when using SSR.
     */
    _thumbReady = computed(() => {
        const size = this._size();
        return !!size && size.width > 0;
    }, ...(ngDevMode ? [{ debugName: "_thumbReady" }] : []));
    _index = computed(() => this._slider.thumbs().findIndex((thumb) => thumb === this), ...(ngDevMode ? [{ debugName: "_index" }] : []));
    percentage = computed(() => {
        const range = this._slider.max() - this._slider.min();
        if (range === 0)
            return 0;
        return ((this._slider.normalizedValue()[this._index()] - this._slider.min()) / range) * 100;
    }, ...(ngDevMode ? [{ debugName: "percentage" }] : []));
    _thumbInBoundsOffset = computed(() => {
        // we can't compute the offset on the server
        if (isPlatformServer(this._platform)) {
            return 0;
        }
        const size = this._slider.isHorizontal() ? this._size()?.width : this._size()?.height;
        if (!size) {
            return 0;
        }
        const halfSize = size / 2;
        const offset = linearScale([0, 50], [0, halfSize]);
        const direction = this._slider.slidingSource() === 'left' || this._slider.slidingSource() === 'top' ? 1 : -1;
        return (halfSize - offset(this.percentage()) * direction) * direction;
    }, ...(ngDevMode ? [{ debugName: "_thumbInBoundsOffset" }] : []));
    /**
     * Offsets the thumb centre point while sliding to ensure it remains
     * within the bounds of the slider when reaching the edges.
     * Based on https://github.com/radix-ui/primitives/blob/main/packages/react/slider/src/slider.tsx
     */
    _thumbOffset = computed(() => {
        // we can't compute the offset on the server
        if (isPlatformServer(this._platform)) {
            return this.percentage() + '%';
        }
        return `calc(${this.percentage()}% + ${this._thumbInBoundsOffset()}px)`;
    }, ...(ngDevMode ? [{ debugName: "_thumbOffset" }] : []));
    _computedAriaLabel = computed(() => this._slider.ariaLabel() || `Value ${this._index() + 1} of ${this._slider.normalizedValue().length}`, ...(ngDevMode ? [{ debugName: "_computedAriaLabel" }] : []));
    _transformValue = computed(() => this._slider.orientation() === 'horizontal' ? 'translateX(-50%)' : 'translateY(-50%)', ...(ngDevMode ? [{ debugName: "_transformValue" }] : []));
    constructor() {
        this._slider.addThumb(this);
        this._destroyRef.onDestroy(() => {
            this._slider.removeThumb(this);
        });
    }
    _onPointerDown(event) {
        this._slider.track()?.onPointerDown(event);
    }
    _onPointerMove(event) {
        this._slider.track()?.onPointerMove(event);
    }
    _onPointerUp(event) {
        this._slider.track()?.onPointerUp(event);
    }
    handleKeydown(event) {
        if (this._slider.mutableDisabled())
            return;
        const step = this._slider.step();
        const min = this._slider.min();
        const max = this._slider.max();
        const multiplier = event.shiftKey || PAGE_KEYS.includes(event.key) ? 10 : 1;
        // Determine delta based on slider orientation
        const dirLR = this._slider.slidingSource() === 'right' ? -1 : 1;
        const dirUD = this._slider.slidingSource() === 'top' ? -1 : 1;
        const deltas = {
            ArrowLeft: -step * multiplier * dirLR,
            ArrowRight: step * multiplier * dirLR,
            ArrowUp: step * multiplier * dirUD,
            ArrowDown: -step * multiplier * dirUD,
            PageUp: step * multiplier,
            PageDown: -step * multiplier,
        };
        // Home/End keys
        if (event.key === 'Home') {
            if (this._slider.isDraggableRangeOnly()) {
                const range = this._slider.normalizedValue();
                const delta = min - range[0];
                this._slider.setAllValuesByDelta(delta);
            }
            else {
                this._slider.setValue(min, this._index());
            }
            event.preventDefault();
            return;
        }
        if (event.key === 'End') {
            if (this._slider.isDraggableRangeOnly()) {
                const range = this._slider.normalizedValue();
                const delta = max - range[range.length - 1];
                this._slider.setAllValuesByDelta(delta);
            }
            else {
                this._slider.setValue(max, this._index());
            }
            event.preventDefault();
            return;
        }
        const delta = deltas[event.key];
        if (delta === undefined)
            return;
        if (this._slider.isDraggableRangeOnly()) {
            this._slider.setAllValuesByDelta(delta);
        }
        else {
            const index = this._index();
            const value = this._slider.normalizedValue()[index];
            this._slider.setValue(Math.min(max, Math.max(min, value + delta)), index);
        }
        event.preventDefault();
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSliderThumb, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "20.3.17", type: BrnSliderThumb, isStandalone: true, selector: "[brnSliderThumb]", host: { attributes: { "role": "slider", "data-slot": "slider-thumb" }, listeners: { "pointerdown": "_onPointerDown($event)", "pointermove": "_onPointerMove($event)", "pointerup": "_onPointerUp($event)", "keydown": "handleKeydown($event)" }, properties: { "attr.aria-label": "_computedAriaLabel()", "attr.aria-labelledby": "_slider.ariaLabelledby() || null", "attr.aria-orientation": "_slider.orientation()", "attr.aria-valuenow": "_slider.normalizedValue()[_index()]", "attr.aria-valuemin": "_slider.min()", "attr.aria-valuemax": "_slider.max()", "attr.tabindex": "_slider.mutableDisabled() ? -1 : 0", "attr.data-disabled": "_slider.mutableDisabled() ? \"\" : null", "attr.data-orientation": "_slider.orientation()", "style.inset-inline-start": "_slider.isHorizontal() ?  _slider.inverted() ? undefined : _thumbOffset() : undefined", "style.inset-inline-end": "_slider.isHorizontal() ? _slider.inverted() ? _thumbOffset() : undefined : undefined", "style.inset-block-end": "!_slider.isHorizontal() ? _slider.inverted() ? undefined : _thumbOffset() : undefined", "style.inset-block-start": "!_slider.isHorizontal() ? _slider.inverted() ? _thumbOffset() : undefined : undefined", "style.visibility": "_thumbReady() ? undefined : \"hidden\"", "style.pointer-events": "_thumbReady() ? undefined : \"none\"", "style.transform": "_transformValue()" } }, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSliderThumb, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnSliderThumb]',
                    host: {
                        role: 'slider',
                        '[attr.aria-label]': `_computedAriaLabel()`,
                        '[attr.aria-labelledby]': '_slider.ariaLabelledby() || null',
                        '[attr.aria-orientation]': '_slider.orientation()',
                        '[attr.aria-valuenow]': '_slider.normalizedValue()[_index()]',
                        '[attr.aria-valuemin]': '_slider.min()',
                        '[attr.aria-valuemax]': '_slider.max()',
                        '[attr.tabindex]': '_slider.mutableDisabled() ? -1 : 0',
                        '[attr.data-disabled]': '_slider.mutableDisabled() ? "" : null',
                        '[attr.data-orientation]': '_slider.orientation()',
                        '[style.inset-inline-start]': '_slider.isHorizontal() ?  _slider.inverted() ? undefined : _thumbOffset() : undefined',
                        '[style.inset-inline-end]': '_slider.isHorizontal() ? _slider.inverted() ? _thumbOffset() : undefined : undefined',
                        '[style.inset-block-end]': '!_slider.isHorizontal() ? _slider.inverted() ? undefined : _thumbOffset() : undefined',
                        '[style.inset-block-start]': '!_slider.isHorizontal() ? _slider.inverted() ? _thumbOffset() : undefined : undefined',
                        '[style.visibility]': '_thumbReady() ? undefined : "hidden"',
                        '[style.pointer-events]': '_thumbReady() ? undefined : "none"',
                        '[style.transform]': '_transformValue()',
                        'data-slot': 'slider-thumb',
                        '(pointerdown)': '_onPointerDown($event)',
                        '(pointermove)': '_onPointerMove($event)',
                        '(pointerup)': '_onPointerUp($event)',
                        '(keydown)': 'handleKeydown($event)',
                    },
                }]
        }], ctorParameters: () => [] });

class BrnSliderTick {
    _slider = injectBrnSlider();
    _templateRef = inject(TemplateRef);
    _renderer = inject(Renderer2);
    _viewContainer = inject(ViewContainerRef);
    _destroyRef = inject(DestroyRef);
    _ticks = [];
    constructor() {
        effect(() => {
            const ticks = this._slider.ticks();
            const tickLabelInterval = this._slider.tickLabelInterval();
            // Remove any existing ticks
            this._ticks.forEach((tick) => this._viewContainer.remove(this._viewContainer.indexOf(tick)));
            // Create new ticks
            this._ticks = [];
            ticks.forEach((tick, index) => {
                const view = this._viewContainer.createEmbeddedView(this._templateRef, {
                    $implicit: tick,
                    index,
                    formattedTick: this._slider.formatTick()(tick),
                });
                const tickEl = view.rootNodes[0];
                if (tickLabelInterval === 0 || index % tickLabelInterval !== 0) {
                    this._renderer.setAttribute(tickEl, 'data-skip', '');
                }
                this._ticks.push(view);
            });
        });
        this._destroyRef.onDestroy(() => {
            this._ticks.forEach((tick) => this._viewContainer.remove(this._viewContainer.indexOf(tick)));
        });
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSliderTick, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "20.3.17", type: BrnSliderTick, isStandalone: true, selector: "[brnSliderTick]", host: { attributes: { "data-slot": "slider-tick" } }, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSliderTick, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnSliderTick]',
                    host: {
                        'data-slot': 'slider-tick',
                    },
                }]
        }], ctorParameters: () => [] });

const BrnSliderTrackToken = new InjectionToken('BrnSliderTrackToken');
function provideBrnSliderTrack(slider) {
    return { provide: BrnSliderTrackToken, useExisting: slider };
}
function injectBrnSliderTrack() {
    return inject(BrnSliderTrackToken);
}

class BrnSliderTrack {
    _elementRef = inject(ElementRef);
    _slider = injectBrnSlider();
    _rangeDragStartPointer = null;
    _rangeDragAccumulatedDelta = 0;
    constructor() {
        this._slider.track.set(this);
    }
    onPointerDown(event) {
        if (this._slider.mutableDisabled())
            return;
        const target = event.target;
        const isTrack = this._isTrack(target);
        // Draggable range only: ignore empty track clicks
        if (isTrack && this._slider.isDraggableRangeOnly())
            return;
        target.setPointerCapture(event.pointerId);
        const isRange = this._isRange(target);
        // Prevent browser focus behaviour because we instead focus a thumb manually when values change.
        if (isTrack || isRange)
            event.preventDefault();
        if ((isRange && this._slider.isDraggableRange()) || this._slider.isDraggableRangeOnly()) {
            this._rangeDragStartPointer = this._getPointerPosition(event);
            return;
        }
        const pointerPosition = this._getPointerPosition(event);
        const value = this._getValueFromPointer(pointerPosition);
        const closestIndex = getClosestValueIndex(this._slider.normalizedValue(), value);
        if (isTrack || isRange) {
            this._slider.setValue(value, closestIndex);
        }
        else {
            this._slider.valueIndexToChange.set(closestIndex);
        }
    }
    onPointerMove(event) {
        if (this._slider.mutableDisabled())
            return;
        const target = event.target;
        if (!target.hasPointerCapture(event.pointerId))
            return;
        if (this._rangeDragStartPointer !== null && this._slider.isDraggableRange()) {
            const currentPointer = this._getPointerPosition(event);
            const pixelDelta = currentPointer - this._rangeDragStartPointer;
            const valueDelta = this._pixelDeltaToValueDelta(pixelDelta);
            // Accumulate sub-step deltas
            this._rangeDragAccumulatedDelta += valueDelta;
            const step = this._slider.step();
            const snappedDelta = Math.trunc(this._rangeDragAccumulatedDelta / step) * step;
            if (snappedDelta !== 0) {
                this._slider.setAllValuesByDelta(snappedDelta);
                // Remove the applied delta, keep remainder
                this._rangeDragAccumulatedDelta -= snappedDelta;
            }
            // Reset start pointer so delta stays incremental
            this._rangeDragStartPointer = currentPointer;
            return;
        }
        const pointerPosition = this._getPointerPosition(event);
        const value = this._getValueFromPointer(pointerPosition);
        this._slider.setValue(value, this._slider.valueIndexToChange());
    }
    onPointerUp(event) {
        const target = event.target;
        if (target.hasPointerCapture(event.pointerId)) {
            target.releasePointerCapture(event.pointerId);
        }
        this._rangeDragStartPointer = null;
        this._rangeDragAccumulatedDelta = 0;
    }
    _getValueFromPointer(pointerPosition) {
        const rect = this._elementRef.nativeElement.getBoundingClientRect();
        const source = this._slider.slidingSource();
        const isVertical = source === 'top' || source === 'bottom';
        const size = isVertical ? rect.height : rect.width;
        const input = [0, size];
        const output = [this._slider.min(), this._slider.max()];
        const value = linearScale(input, output);
        let relativePosition;
        switch (source) {
            case 'left':
                relativePosition = pointerPosition - rect.left;
                break;
            case 'right':
                relativePosition = rect.right - pointerPosition;
                break;
            case 'top':
                relativePosition = pointerPosition - rect.top;
                break;
            case 'bottom':
                relativePosition = rect.bottom - pointerPosition;
                break;
        }
        return value(relativePosition);
    }
    _pixelDeltaToValueDelta(pixelDelta) {
        const rect = this._elementRef.nativeElement.getBoundingClientRect();
        const size = this._slider.isHorizontal() ? rect.width : rect.height;
        const scale = linearScale([0, size], [this._slider.min(), this._slider.max()]);
        // Determine direction multiplier based on sliding source
        let direction = 1;
        switch (this._slider.slidingSource()) {
            case 'right':
            case 'bottom':
                direction = -1;
                break;
            case 'left':
            case 'top':
                direction = 1;
                break;
        }
        return scale(pixelDelta * direction) - scale(0);
    }
    _isTrack(el) {
        return this._elementRef.nativeElement === el;
    }
    _isRange(el) {
        return this._slider.range()?.elementRef.nativeElement === el;
    }
    _getPointerPosition(event) {
        return this._slider.orientation() === 'horizontal' ? event.clientX : event.clientY;
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSliderTrack, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "20.3.17", type: BrnSliderTrack, isStandalone: true, selector: "[brnSliderTrack]", host: { attributes: { "data-slot": "slider-track" }, listeners: { "pointerdown": "onPointerDown($event)", "pointermove": "onPointerMove($event)", "pointerup": "onPointerUp($event)" }, properties: { "attr.data-disabled": "_slider.mutableDisabled() ? \"\" : null", "attr.data-orientation": "_slider.orientation()" } }, providers: [provideBrnSliderTrack(BrnSliderTrack)], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSliderTrack, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnSliderTrack]',
                    providers: [provideBrnSliderTrack(BrnSliderTrack)],
                    host: {
                        '[attr.data-disabled]': '_slider.mutableDisabled() ? "" : null',
                        '[attr.data-orientation]': '_slider.orientation()',
                        'data-slot': 'slider-track',
                        '(pointerdown)': 'onPointerDown($event)',
                        '(pointermove)': 'onPointerMove($event)',
                        '(pointerup)': 'onPointerUp($event)',
                    },
                }]
        }], ctorParameters: () => [] });
function getClosestValueIndex(values, nextValue) {
    if (values.length === 1)
        return 0;
    const distances = values.map((value) => Math.abs(value - nextValue));
    const closestDistance = Math.min(...distances);
    return distances.indexOf(closestDistance);
}

const BrnSliderImports = [BrnSlider, BrnSliderTrack, BrnSliderThumb, BrnSliderRange, BrnSliderTick];

/**
 * Generated bundle index. Do not edit.
 */

export { BRN_SLIDER_VALUE_ACCESSOR, BrnSlider, BrnSliderImports, BrnSliderRange, BrnSliderThumb, BrnSliderTick, BrnSliderTrack, BrnSliderTrackToken, injectBrnSlider, injectBrnSliderTrack, provideBrnSlider, provideBrnSliderTrack };
