// @spartan-ng/brain/fesm2022/spartan-ng-brain-checkbox.mjs
import { FocusMonitor } from '@angular/cdk/a11y';
import { isPlatformBrowser } from '@angular/common';
import * as i0 from '@angular/core';
import { forwardRef, inject, DestroyRef, Renderer2, ElementRef, ChangeDetectorRef, DOCUMENT, PLATFORM_ID, signal, model, output, computed, input, linkedSignal, booleanAttribute, viewChild, afterRenderEffect, ChangeDetectionStrategy, Component } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import * as i1 from '@spartan-ng/brain/field';
import { BrnFieldControl, provideBrnLabelable } from '@spartan-ng/brain/field';

const BRN_CHECKBOX_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(() => BrnCheckbox),
    multi: true,
};
let uniqueIdCounter = 0;
const CONTAINER_POST_FIX = '-checkbox';
class BrnCheckbox {
    _destroyRef = inject(DestroyRef);
    _renderer = inject(Renderer2);
    _elementRef = inject(ElementRef);
    _focusMonitor = inject(FocusMonitor);
    _cdr = inject(ChangeDetectorRef);
    _fieldControl = inject(BrnFieldControl, { optional: true });
    _document = inject(DOCUMENT);
    _isBrowser = isPlatformBrowser(inject(PLATFORM_ID));
    _focusVisible = signal(false, ...(ngDevMode ? [{ debugName: "_focusVisible" }] : []));
    _focused = signal(false, ...(ngDevMode ? [{ debugName: "_focused" }] : []));
    /**
     * The checked state of the checkbox.
     * Can be bound with [(checked)] for two-way binding.
     */
    checked = model(false, ...(ngDevMode ? [{ debugName: "checked" }] : []));
    /** Emits when checked state changes. */
    checkedChange = output();
    /**
     * Read-only signal of current checkbox state.
     * Use this when you only need to read state without changing it.
     */
    isChecked = this.checked.asReadonly();
    /*
     * The indeterminate state of the checkbox.
     * For example, a "select all/deselect all" checkbox may be in the indeterminate state when some but not all of its sub-controls are checked.
     */
    indeterminate = model(false, ...(ngDevMode ? [{ debugName: "indeterminate" }] : []));
    /**
     * Computed data-state attribute value based on checked state.
     * Returns 'checked', 'unchecked', or 'indeterminate'.
     */
    _dataState = computed(() => {
        if (this.indeterminate())
            return 'indeterminate';
        return this.checked() ? 'checked' : 'unchecked';
    }, ...(ngDevMode ? [{ debugName: "_dataState" }] : []));
    /**
     * Computed aria-checked attribute value for accessibility.
     * Returns 'true', 'false', or 'mixed' (for indeterminate).
     */
    _ariaChecked = computed(() => {
        if (this.indeterminate())
            return 'mixed';
        return this.checked() ? 'true' : 'false';
    }, ...(ngDevMode ? [{ debugName: "_ariaChecked" }] : []));
    /**
     * Unique identifier for checkbox component.
     * When provided, inner button gets ID without '-checkbox' suffix.
     * Auto-generates ID if not provided.
     */
    id = input(++uniqueIdCounter + '', ...(ngDevMode ? [{ debugName: "id" }] : []));
    /**
     * Form control name for checkbox.
     * When provided, inner button gets name without '-checkbox' suffix.
     */
    name = input(null, ...(ngDevMode ? [{ debugName: "name" }] : []));
    /**
     * CSS classes applied to inner button element.
     */
    class = input(null, ...(ngDevMode ? [{ debugName: "class" }] : []));
    /**
     * Accessibility label for screen readers.
     * Use when no visible label exists.
     */
    ariaLabel = input(null, ...(ngDevMode ? [{ debugName: "ariaLabel", alias: 'aria-label' }] : [{ alias: 'aria-label' }]));
    /**
     * ID of element that labels this checkbox for accessibility.
     * Auto-set when checkbox is inside label element.
     */
    ariaLabelledby = input(null, ...(ngDevMode ? [{ debugName: "ariaLabelledby", alias: 'aria-labelledby' }] : [{ alias: 'aria-labelledby' }]));
    mutableAriaLabelledby = linkedSignal(() => this.ariaLabelledby(), ...(ngDevMode ? [{ debugName: "mutableAriaLabelledby" }] : []));
    /**
     * ID of element that describes this checkbox for accessibility.
     */
    ariaDescribedby = input(null, ...(ngDevMode ? [{ debugName: "ariaDescribedby", alias: 'aria-describedby' }] : [{ alias: 'aria-describedby' }]));
    /**
     * Whether checkbox is required in a form.
     */
    required = input(false, ...(ngDevMode ? [{ debugName: "required", transform: booleanAttribute }] : [{ transform: booleanAttribute }]));
    /**
     * Whether checkbox is disabled.
     * Disabled checkboxes cannot be toggled and indicate disabled state through data-disabled attribute.
     */
    disabled = input(false, ...(ngDevMode ? [{ debugName: "disabled", transform: booleanAttribute }] : [{ transform: booleanAttribute }]));
    /**
     * Whether to force the field into an invalid state, regardless of the form control's state.
     * Overrides both the `data-invalid` and `data-matches-spartan-invalid` attributes.
     */
    forceInvalid = input(false, ...(ngDevMode ? [{ debugName: "forceInvalid", transform: booleanAttribute }] : [{ transform: booleanAttribute }]));
    /**
     * Computed state for checkbox container and accessibility.
     * Manages ID, name, and disabled state.
     */
    _state = computed(() => {
        const name = this.name();
        const id = this.id();
        return {
            disabled: signal(this.disabled()),
            name: name ? name + CONTAINER_POST_FIX : null,
            id: id ? id + CONTAINER_POST_FIX : null,
        };
    }, ...(ngDevMode ? [{ debugName: "_state" }] : []));
    _buttonId = computed(() => this._getCheckboxButtonId(this._state().id), ...(ngDevMode ? [{ debugName: "_buttonId" }] : []));
    _buttonName = computed(() => this._getCheckboxButtonId(this._state().name), ...(ngDevMode ? [{ debugName: "_buttonName" }] : []));
    labelableId = this._buttonId;
    _dirty = this._fieldControl?.dirty;
    _invalid = computed(() => this.forceInvalid() || (this._fieldControl?.invalid?.() ?? null), ...(ngDevMode ? [{ debugName: "_invalid" }] : []));
    spartanInvalid = computed(() => this.forceInvalid() || (this._fieldControl?.spartanInvalid?.() ?? null), ...(ngDevMode ? [{ debugName: "spartanInvalid" }] : []));
    _controlTouched = this._fieldControl?.touched;
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    _onChange = () => { };
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    _onTouched = () => { };
    /**
     * Reference to the checkbox button element in the template.
     */
    checkbox = viewChild.required('checkBox');
    /**
     * Event emitted when checkbox is blurred (loses focus).
     * Used for form validation.
     */
    touched = output();
    constructor() {
        afterRenderEffect(() => {
            const state = this._state();
            const isDisabled = state.disabled();
            if (!this._elementRef.nativeElement || !this._isBrowser)
                return;
            const newLabelId = state.id + '-label';
            const checkboxButtonId = this._getCheckboxButtonId(state.id);
            const labelElement = this._elementRef.nativeElement.closest('label') ??
                this._document.querySelector(`label[for="${checkboxButtonId}"]`);
            if (!labelElement)
                return;
            const existingLabelId = labelElement.id;
            this._renderer.setAttribute(labelElement, 'data-disabled', isDisabled ? 'true' : 'false');
            this.mutableAriaLabelledby.set(existingLabelId || newLabelId);
            if (!existingLabelId || existingLabelId.length === 0) {
                this._renderer.setAttribute(labelElement, 'id', newLabelId);
            }
        });
    }
    /**
     * Toggles checkbox between checked/unchecked states.
     * If checkbox is indeterminate, sets to checked.
     * Does nothing if checkbox is disabled.
     */
    toggle() {
        if (this._state().disabled())
            return;
        this._onTouched();
        this.touched.emit();
        const newChecked = this.indeterminate() ? true : !this.checked();
        this.indeterminate.set(false);
        this.checkedChange.emit(newChecked);
        this.checked.set(newChecked);
        this._onChange(newChecked);
    }
    ngAfterContentInit() {
        this._focusMonitor
            .monitor(this._elementRef, true)
            .pipe(takeUntilDestroyed(this._destroyRef))
            .subscribe((focusOrigin) => {
            if (focusOrigin)
                this._focused.set(true);
            if (focusOrigin === 'keyboard' || focusOrigin === 'program') {
                this._focusVisible.set(true);
                this._cdr.markForCheck();
            }
            if (!focusOrigin) {
                // When a focused element becomes disabled, the browser *immediately* fires a blur event.
                // Angular does not expect events to be raised during change detection, so any state
                // change (such as a form control's ng-touched) will cause a changed-after-checked error.
                // See https://github.com/angular/angular/issues/17793. To work around this, we defer
                // telling the form control it has been touched until the next tick.
                Promise.resolve().then(() => {
                    this._focusVisible.set(false);
                    this._focused.set(false);
                    this._onTouched();
                    this.touched.emit();
                    this._cdr.markForCheck();
                });
            }
        });
    }
    ngOnDestroy() {
        this._focusMonitor.stopMonitoring(this._elementRef);
    }
    /**
     * Gets proper ID for inner button element.
     * Removes '-checkbox' suffix if present in container ID.
     *
     * @param idPassedToContainer - ID applied to container element
     * @returns ID to use for inner button or null
     */
    _getCheckboxButtonId(idPassedToContainer) {
        return idPassedToContainer ? idPassedToContainer.replace(new RegExp(CONTAINER_POST_FIX + '$'), '') : null;
    }
    /**
     * Updates internal state when control value changes from outside.
     * Handles boolean and 'indeterminate' values.
     * Part of ControlValueAccessor interface.
     *
     * @param value - New checkbox state (true/false/'indeterminate')
     */
    writeValue(value) {
        this.checked.set(value);
    }
    /**
     * Registers callback for value changes.
     * Part of ControlValueAccessor interface.
     *
     * @param fn - Function to call when value changes
     */
    registerOnChange(fn) {
        this._onChange = fn;
    }
    /**
     * Registers callback for touched events.
     * Part of ControlValueAccessor interface.
     *
     * @param fn - Function to call when control is touched
     */
    registerOnTouched(fn) {
        this._onTouched = fn;
    }
    /**
     * Updates disabled state from form control.
     * Part of ControlValueAccessor interface.
     *
     * @param isDisabled - Whether checkbox should be disabled
     */
    setDisabledState(isDisabled) {
        this._state().disabled.set(isDisabled);
        this._cdr.markForCheck();
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnCheckbox, deps: [], target: i0.ɵɵFactoryTarget.Component });
    /** @nocollapse */ static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.2.0", version: "20.3.17", type: BrnCheckbox, isStandalone: true, selector: "brn-checkbox", inputs: { checked: { classPropertyName: "checked", publicName: "checked", isSignal: true, isRequired: false, transformFunction: null }, indeterminate: { classPropertyName: "indeterminate", publicName: "indeterminate", isSignal: true, isRequired: false, transformFunction: null }, id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null }, name: { classPropertyName: "name", publicName: "name", isSignal: true, isRequired: false, transformFunction: null }, class: { classPropertyName: "class", publicName: "class", isSignal: true, isRequired: false, transformFunction: null }, ariaLabel: { classPropertyName: "ariaLabel", publicName: "aria-label", isSignal: true, isRequired: false, transformFunction: null }, ariaLabelledby: { classPropertyName: "ariaLabelledby", publicName: "aria-labelledby", isSignal: true, isRequired: false, transformFunction: null }, ariaDescribedby: { classPropertyName: "ariaDescribedby", publicName: "aria-describedby", isSignal: true, isRequired: false, transformFunction: null }, required: { classPropertyName: "required", publicName: "required", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "disabled", isSignal: true, isRequired: false, transformFunction: null }, forceInvalid: { classPropertyName: "forceInvalid", publicName: "forceInvalid", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { checked: "checkedChange", checkedChange: "checkedChange", indeterminate: "indeterminateChange", touched: "touched" }, host: { properties: { "style": "{display: \"contents\"}", "attr.id": "_state().id", "attr.name": "_state().name", "attr.aria-labelledby": "null", "attr.aria-label": "null", "attr.aria-describedby": "null", "attr.aria-invalid": "_invalid?.() ? \"true\" : null", "attr.data-invalid": "_invalid?.() ? \"true\" : null", "attr.data-matches-spartan-invalid": "spartanInvalid?.() ? \"true\" : null", "attr.data-touched": "_controlTouched?.() ? \"true\" : null", "attr.data-dirty": "_dirty?.() ? \"true\" : null", "attr.data-state": "_dataState()", "attr.data-focus-visible": "_focusVisible()", "attr.data-focus": "_focused()", "attr.data-disabled": "_state().disabled()" } }, providers: [BRN_CHECKBOX_VALUE_ACCESSOR, provideBrnLabelable(BrnCheckbox)], viewQueries: [{ propertyName: "checkbox", first: true, predicate: ["checkBox"], descendants: true, isSignal: true }], hostDirectives: [{ directive: i1.BrnFieldControl }], ngImport: i0, template: `
		<button
			#checkBox
			role="checkbox"
			type="button"
			[attr.id]="_buttonId()"
			[attr.name]="_buttonName()"
			[class]="class()"
			[attr.aria-checked]="_ariaChecked()"
			[attr.aria-label]="ariaLabel() || null"
			[attr.aria-labelledby]="mutableAriaLabelledby() || null"
			[attr.aria-describedby]="ariaDescribedby() || null"
			[attr.data-state]="_dataState()"
			[attr.data-focus-visible]="_focusVisible()"
			[attr.data-focus]="_focused()"
			[attr.data-disabled]="_state().disabled()"
			[disabled]="_state().disabled()"
			[tabIndex]="_state().disabled() ? -1 : 0"
			(click)="$event.preventDefault(); toggle()"
		>
			<ng-content />
		</button>
	`, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnCheckbox, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-checkbox',
                    providers: [BRN_CHECKBOX_VALUE_ACCESSOR, provideBrnLabelable(BrnCheckbox)],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    hostDirectives: [BrnFieldControl],
                    host: {
                        '[style]': '{display: "contents"}',
                        '[attr.id]': '_state().id',
                        '[attr.name]': '_state().name',
                        '[attr.aria-labelledby]': 'null',
                        '[attr.aria-label]': 'null',
                        '[attr.aria-describedby]': 'null',
                        '[attr.aria-invalid]': '_invalid?.() ? "true" : null',
                        '[attr.data-invalid]': '_invalid?.() ? "true" : null',
                        '[attr.data-matches-spartan-invalid]': 'spartanInvalid?.() ? "true" : null',
                        '[attr.data-touched]': '_controlTouched?.() ? "true" : null',
                        '[attr.data-dirty]': '_dirty?.() ? "true" : null',
                        '[attr.data-state]': '_dataState()',
                        '[attr.data-focus-visible]': '_focusVisible()',
                        '[attr.data-focus]': '_focused()',
                        '[attr.data-disabled]': '_state().disabled()',
                    },
                    template: `
		<button
			#checkBox
			role="checkbox"
			type="button"
			[attr.id]="_buttonId()"
			[attr.name]="_buttonName()"
			[class]="class()"
			[attr.aria-checked]="_ariaChecked()"
			[attr.aria-label]="ariaLabel() || null"
			[attr.aria-labelledby]="mutableAriaLabelledby() || null"
			[attr.aria-describedby]="ariaDescribedby() || null"
			[attr.data-state]="_dataState()"
			[attr.data-focus-visible]="_focusVisible()"
			[attr.data-focus]="_focused()"
			[attr.data-disabled]="_state().disabled()"
			[disabled]="_state().disabled()"
			[tabIndex]="_state().disabled() ? -1 : 0"
			(click)="$event.preventDefault(); toggle()"
		>
			<ng-content />
		</button>
	`,
                }]
        }], ctorParameters: () => [], propDecorators: { checked: [{ type: i0.Input, args: [{ isSignal: true, alias: "checked", required: false }] }, { type: i0.Output, args: ["checkedChange"] }], checkedChange: [{ type: i0.Output, args: ["checkedChange"] }], indeterminate: [{ type: i0.Input, args: [{ isSignal: true, alias: "indeterminate", required: false }] }, { type: i0.Output, args: ["indeterminateChange"] }], id: [{ type: i0.Input, args: [{ isSignal: true, alias: "id", required: false }] }], name: [{ type: i0.Input, args: [{ isSignal: true, alias: "name", required: false }] }], class: [{ type: i0.Input, args: [{ isSignal: true, alias: "class", required: false }] }], ariaLabel: [{ type: i0.Input, args: [{ isSignal: true, alias: "aria-label", required: false }] }], ariaLabelledby: [{ type: i0.Input, args: [{ isSignal: true, alias: "aria-labelledby", required: false }] }], ariaDescribedby: [{ type: i0.Input, args: [{ isSignal: true, alias: "aria-describedby", required: false }] }], required: [{ type: i0.Input, args: [{ isSignal: true, alias: "required", required: false }] }], disabled: [{ type: i0.Input, args: [{ isSignal: true, alias: "disabled", required: false }] }], forceInvalid: [{ type: i0.Input, args: [{ isSignal: true, alias: "forceInvalid", required: false }] }], checkbox: [{ type: i0.ViewChild, args: ['checkBox', { isSignal: true }] }], touched: [{ type: i0.Output, args: ["touched"] }] } });

const BrnCheckboxImports = [BrnCheckbox];

/**
 * Generated bundle index. Do not edit.
 */

export { BRN_CHECKBOX_VALUE_ACCESSOR, BrnCheckbox, BrnCheckboxImports };
