// @spartan-ng/brain/fesm2022/spartan-ng-brain-sonner.mjs
import { Directionality } from '@angular/cdk/bidi';
import { NgComponentOutlet, isPlatformBrowser, NgTemplateOutlet } from '@angular/common';
import * as i0 from '@angular/core';
import { input, ChangeDetectionStrategy, Component, booleanAttribute, InjectionToken, inject, Pipe, signal, computed, viewChild, linkedSignal, effect, afterRenderEffect, PLATFORM_ID, DOCUMENT, numberAttribute, contentChild, TemplateRef, DestroyRef, afterNextRender, ViewEncapsulation } from '@angular/core';
import clsx from 'clsx';

class BrnSonnerIcon {
    type = input('default', ...(ngDevMode ? [{ debugName: "type" }] : []));
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSonnerIcon, deps: [], target: i0.ɵɵFactoryTarget.Component });
    /** @nocollapse */ static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "20.3.17", type: BrnSonnerIcon, isStandalone: true, selector: "brn-sonner-icon", inputs: { type: { classPropertyName: "type", publicName: "type", isSignal: true, isRequired: false, transformFunction: null } }, ngImport: i0, template: `
		@switch (type()) {
			@case ('success') {
				<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" height="20" width="20">
					<path
						fill-rule="evenodd"
						d="M10 18a8 8 0 100-16 8 8 0 000 16zm3.857-9.809a.75.75 0 00-1.214-.882l-3.483 4.79-1.88-1.88a.75.75 0 10-1.06 1.061l2.5 2.5a.75.75 0 001.137-.089l4-5.5z"
						clip-rule="evenodd"
					/>
				</svg>
			}
			@case ('error') {
				<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" height="20" width="20">
					<path
						fill-rule="evenodd"
						d="M18 10a8 8 0 11-16 0 8 8 0 0116 0zm-8-5a.75.75 0 01.75.75v4.5a.75.75 0 01-1.5 0v-4.5A.75.75 0 0110 5zm0 10a1 1 0 100-2 1 1 0 000 2z"
						clip-rule="evenodd"
					/>
				</svg>
			}
			@case ('info') {
				<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" height="20" width="20">
					<path
						fill-rule="evenodd"
						d="M18 10a8 8 0 11-16 0 8 8 0 0116 0zm-7-4a1 1 0 11-2 0 1 1 0 012 0zM9 9a.75.75 0 000 1.5h.253a.25.25 0 01.244.304l-.459 2.066A1.75 1.75 0 0010.747 15H11a.75.75 0 000-1.5h-.253a.25.25 0 01-.244-.304l.459-2.066A1.75 1.75 0 009.253 9H9z"
						clip-rule="evenodd"
					/>
				</svg>
			}
			@case ('warning') {
				<svg viewBox="0 0 64 64" fill="currentColor" height="20" width="20" xmlns="http://www.w3.org/2000/svg">
					<path
						d="M32.427,7.987c2.183,0.124 4,1.165 5.096,3.281l17.936,36.208c1.739,3.66 -0.954,8.585 -5.373,8.656l-36.119,0c-4.022,-0.064 -7.322,-4.631 -5.352,-8.696l18.271,-36.207c0.342,-0.65 0.498,-0.838 0.793,-1.179c1.186,-1.375 2.483,-2.111 4.748,-2.063Zm-0.295,3.997c-0.687,0.034 -1.316,0.419 -1.659,1.017c-6.312,11.979 -12.397,24.081 -18.301,36.267c-0.546,1.225 0.391,2.797 1.762,2.863c12.06,0.195 24.125,0.195 36.185,0c1.325,-0.064 2.321,-1.584 1.769,-2.85c-5.793,-12.184 -11.765,-24.286 -17.966,-36.267c-0.366,-0.651 -0.903,-1.042 -1.79,-1.03Z"
					/>
					<path
						d="M33.631,40.581l-3.348,0l-0.368,-16.449l4.1,0l-0.384,16.449Zm-3.828,5.03c0,-0.609 0.197,-1.113 0.592,-1.514c0.396,-0.4 0.935,-0.601 1.618,-0.601c0.684,0 1.223,0.201 1.618,0.601c0.395,0.401 0.593,0.905 0.593,1.514c0,0.587 -0.193,1.078 -0.577,1.473c-0.385,0.395 -0.929,0.593 -1.634,0.593c-0.705,0 -1.249,-0.198 -1.634,-0.593c-0.384,-0.395 -0.576,-0.886 -0.576,-1.473Z"
					/>
				</svg>
			}
		}
	`, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSonnerIcon, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-sonner-icon',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    template: `
		@switch (type()) {
			@case ('success') {
				<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" height="20" width="20">
					<path
						fill-rule="evenodd"
						d="M10 18a8 8 0 100-16 8 8 0 000 16zm3.857-9.809a.75.75 0 00-1.214-.882l-3.483 4.79-1.88-1.88a.75.75 0 10-1.06 1.061l2.5 2.5a.75.75 0 001.137-.089l4-5.5z"
						clip-rule="evenodd"
					/>
				</svg>
			}
			@case ('error') {
				<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" height="20" width="20">
					<path
						fill-rule="evenodd"
						d="M18 10a8 8 0 11-16 0 8 8 0 0116 0zm-8-5a.75.75 0 01.75.75v4.5a.75.75 0 01-1.5 0v-4.5A.75.75 0 0110 5zm0 10a1 1 0 100-2 1 1 0 000 2z"
						clip-rule="evenodd"
					/>
				</svg>
			}
			@case ('info') {
				<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" height="20" width="20">
					<path
						fill-rule="evenodd"
						d="M18 10a8 8 0 11-16 0 8 8 0 0116 0zm-7-4a1 1 0 11-2 0 1 1 0 012 0zM9 9a.75.75 0 000 1.5h.253a.25.25 0 01.244.304l-.459 2.066A1.75 1.75 0 0010.747 15H11a.75.75 0 000-1.5h-.253a.25.25 0 01-.244-.304l.459-2.066A1.75 1.75 0 009.253 9H9z"
						clip-rule="evenodd"
					/>
				</svg>
			}
			@case ('warning') {
				<svg viewBox="0 0 64 64" fill="currentColor" height="20" width="20" xmlns="http://www.w3.org/2000/svg">
					<path
						d="M32.427,7.987c2.183,0.124 4,1.165 5.096,3.281l17.936,36.208c1.739,3.66 -0.954,8.585 -5.373,8.656l-36.119,0c-4.022,-0.064 -7.322,-4.631 -5.352,-8.696l18.271,-36.207c0.342,-0.65 0.498,-0.838 0.793,-1.179c1.186,-1.375 2.483,-2.111 4.748,-2.063Zm-0.295,3.997c-0.687,0.034 -1.316,0.419 -1.659,1.017c-6.312,11.979 -12.397,24.081 -18.301,36.267c-0.546,1.225 0.391,2.797 1.762,2.863c12.06,0.195 24.125,0.195 36.185,0c1.325,-0.064 2.321,-1.584 1.769,-2.85c-5.793,-12.184 -11.765,-24.286 -17.966,-36.267c-0.366,-0.651 -0.903,-1.042 -1.79,-1.03Z"
					/>
					<path
						d="M33.631,40.581l-3.348,0l-0.368,-16.449l4.1,0l-0.384,16.449Zm-3.828,5.03c0,-0.609 0.197,-1.113 0.592,-1.514c0.396,-0.4 0.935,-0.601 1.618,-0.601c0.684,0 1.223,0.201 1.618,0.601c0.395,0.401 0.593,0.905 0.593,1.514c0,0.587 -0.193,1.078 -0.577,1.473c-0.385,0.395 -0.929,0.593 -1.634,0.593c-0.705,0 -1.249,-0.198 -1.634,-0.593c-0.384,-0.395 -0.576,-0.886 -0.576,-1.473Z"
					/>
				</svg>
			}
		}
	`,
                }]
        }], propDecorators: { type: [{ type: i0.Input, args: [{ isSignal: true, alias: "type", required: false }] }] } });

class BrnSonnerLoader {
    isVisible = input.required(...(ngDevMode ? [{ debugName: "isVisible", transform: booleanAttribute }] : [{ transform: booleanAttribute }]));
    _bars = Array(12).fill(0);
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSonnerLoader, deps: [], target: i0.ɵɵFactoryTarget.Component });
    /** @nocollapse */ static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "20.3.17", type: BrnSonnerLoader, isStandalone: true, selector: "brn-sonner-loader", inputs: { isVisible: { classPropertyName: "isVisible", publicName: "isVisible", isSignal: true, isRequired: true, transformFunction: null } }, ngImport: i0, template: `
		<div class="sonner-loading-wrapper" [attr.data-visible]="isVisible()">
			<div class="sonner-spinner">
				@for (_ of _bars; track $index) {
					<div class="sonner-loading-bar"></div>
				}
			</div>
		</div>
	`, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSonnerLoader, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-sonner-loader',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    template: `
		<div class="sonner-loading-wrapper" [attr.data-visible]="isVisible()">
			<div class="sonner-spinner">
				@for (_ of _bars; track $index) {
					<div class="sonner-loading-bar"></div>
				}
			</div>
		</div>
	`,
                }]
        }], propDecorators: { isVisible: [{ type: i0.Input, args: [{ isSignal: true, alias: "isVisible", required: true }] }] } });

const defaultConfig = {
    visibleToastsAmount: 3,
    viewPortOffset: '32px',
    toastLifetime: 4000,
    toastWidth: 356,
    gap: 14,
    swipeThreshold: 20,
    timeBeforeUnmount: 200,
};
const BrnSonnerToasterConfigToken = new InjectionToken('BrnSonnerToasterConfig');
function provideBrnSonnerToasterConfig(config) {
    return { provide: BrnSonnerToasterConfigToken, useValue: { ...defaultConfig, ...config } };
}
function injectBrnSonnerToasterConfig() {
    return inject(BrnSonnerToasterConfigToken, { optional: true }) ?? defaultConfig;
}
const defaultClasses = {
    toast: '',
    title: '',
    description: '',
    loader: '',
    closeButton: '',
    cancelButton: '',
    actionButton: '',
    action: '',
    warning: '',
    error: '',
    success: '',
    default: '',
    info: '',
    loading: '',
};

class AsComponentPipe {
    transform(value) {
        return value;
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: AsComponentPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
    /** @nocollapse */ static ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "20.3.17", ngImport: i0, type: AsComponentPipe, isStandalone: true, name: "asComponent" });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: AsComponentPipe, decorators: [{
            type: Pipe,
            args: [{ name: 'asComponent' }]
        }] });

class IsStringPipe {
    transform(value) {
        return typeof value === 'string';
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: IsStringPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
    /** @nocollapse */ static ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "20.3.17", ngImport: i0, type: IsStringPipe, isStandalone: true, name: "isString" });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: IsStringPipe, decorators: [{
            type: Pipe,
            args: [{ name: 'isString' }]
        }] });

let toastsCounter = 0;
function createToastState() {
    const toasts = signal([], ...(ngDevMode ? [{ debugName: "toasts" }] : []));
    const heights = signal([], ...(ngDevMode ? [{ debugName: "heights" }] : []));
    function addToast(data) {
        toasts.update((prev) => [data, ...prev]);
    }
    function create(data) {
        const { message, ...rest } = data;
        const id = typeof data?.id === 'number' || (data.id && data.id?.length > 0) ? data.id : toastsCounter++;
        const dismissible = data.dismissible ?? true;
        const type = data.type ?? 'default';
        const alreadyExists = toasts().find((toast) => toast.id === id);
        if (alreadyExists) {
            toasts.update((prev) => prev.map((toast) => {
                if (toast.id === id) {
                    return {
                        ...toast,
                        ...data,
                        id,
                        title: message,
                        dismissible,
                        type,
                        updated: true,
                    };
                }
                else
                    return { ...toast, updated: false };
            }));
        }
        else {
            addToast({ ...rest, id, title: message, dismissible: dismissible, type });
        }
        return id;
    }
    function dismiss(id) {
        if (id === undefined) {
            toasts.set([]);
            return;
        }
        toasts.update((prev) => prev.filter((toast) => toast.id !== id));
        return id;
    }
    function message(message, data) {
        return create({ ...data, type: 'default', message });
    }
    function error(message, data) {
        return create({ ...data, type: 'error', message });
    }
    function success(message, data) {
        return create({ ...data, type: 'success', message });
    }
    function info(message, data) {
        return create({ ...data, type: 'info', message });
    }
    function warning(message, data) {
        return create({ ...data, type: 'warning', message });
    }
    function loading(message, data) {
        return create({ ...data, type: 'loading', message });
    }
    function promise(promise, data) {
        if (!data)
            return;
        let id = undefined;
        if (data.loading !== undefined) {
            id = create({
                ...data,
                promise,
                type: 'loading',
                message: data.loading,
            });
        }
        const p = promise instanceof Promise ? promise : promise();
        let shouldDismiss = id !== undefined;
        p.then((response) => {
            // @ts-expect-error: Incorrect response type
            if (response && typeof response.ok === 'boolean' && !response.ok) {
                shouldDismiss = false;
                const message = typeof data.error === 'function'
                    ? // @ts-expect-error: TODO: Better function checking
                        data.error(`HTTP error! status: ${response.status}`)
                    : data.error;
                create({ id, type: 'error', message });
            }
            else if (data.success !== undefined) {
                shouldDismiss = false;
                const message = typeof data.success === 'function'
                    ? // @ts-expect-error: TODO: Better function checking
                        data.success(response)
                    : data.success;
                create({ id, type: 'success', message });
            }
        })
            .catch((error) => {
            if (data.error !== undefined) {
                shouldDismiss = false;
                const message = 
                // @ts-expect-error: TODO: Better function checking
                typeof data.error === 'function' ? data.error(error) : data.error;
                create({ id, type: 'error', message });
            }
        })
            .finally(() => {
            if (shouldDismiss) {
                // Toast is still in load state (and will be indefinitely — dismiss it)
                dismiss(id);
                id = undefined;
            }
            data.finally?.();
        });
        return id;
    }
    function custom(component, data) {
        const id = data?.id ?? toastsCounter++;
        create({ component, id, ...data });
        return id;
    }
    function removeHeight(id) {
        heights.update((prev) => prev.filter((height) => height.toastId !== id));
    }
    function addHeight(height) {
        heights.update((prev) => [height, ...prev].sort(sortHeights));
    }
    const sortHeights = (a, b) => toasts().findIndex((t) => t.id === a.toastId) - toasts().findIndex((t) => t.id === b.toastId);
    function reset() {
        toasts.set([]);
        heights.set([]);
    }
    return {
        //methods
        create,
        addToast,
        dismiss,
        message,
        error,
        success,
        info,
        warning,
        loading,
        promise,
        custom,
        removeHeight,
        addHeight,
        reset,
        // signals
        toasts: toasts.asReadonly(),
        heights: heights.asReadonly(),
    };
}
const toastState = createToastState();
// bind this to the toast function
function toastFunction(message, data) {
    return toastState.create({
        message,
        ...data,
    });
}
const basicToast = toastFunction;
const toast = Object.assign(basicToast, {
    success: toastState.success,
    info: toastState.info,
    warning: toastState.warning,
    error: toastState.error,
    custom: toastState.custom,
    message: toastState.message,
    promise: toastState.promise,
    dismiss: toastState.dismiss,
    loading: toastState.loading,
});

class BrnSonnerToast {
    _config = injectBrnSonnerToasterConfig();
    _toasts = computed(() => toastState.toasts().filter((t) => t.position === this.position()), ...(ngDevMode ? [{ debugName: "_toasts" }] : []));
    _heights = computed(() => toastState.heights().filter((h) => h.position === this.position()), ...(ngDevMode ? [{ debugName: "_heights" }] : []));
    _removeHeight = toastState.removeHeight;
    _addHeight = toastState.addHeight;
    _dismiss = toastState.dismiss;
    toast = input.required(...(ngDevMode ? [{ debugName: "toast" }] : []));
    index = input.required(...(ngDevMode ? [{ debugName: "index" }] : []));
    expanded = input.required(...(ngDevMode ? [{ debugName: "expanded" }] : []));
    invert = input.required(...(ngDevMode ? [{ debugName: "invert" }] : []));
    position = input.required(...(ngDevMode ? [{ debugName: "position" }] : []));
    visibleToasts = input.required(...(ngDevMode ? [{ debugName: "visibleToasts" }] : []));
    expandByDefault = input.required(...(ngDevMode ? [{ debugName: "expandByDefault" }] : []));
    closeButton = input.required(...(ngDevMode ? [{ debugName: "closeButton" }] : []));
    interacting = input.required(...(ngDevMode ? [{ debugName: "interacting" }] : []));
    cancelButtonStyle = input(...(ngDevMode ? [undefined, { debugName: "cancelButtonStyle" }] : []));
    actionButtonStyle = input(...(ngDevMode ? [undefined, { debugName: "actionButtonStyle" }] : []));
    duration = input(this._config.toastLifetime, ...(ngDevMode ? [{ debugName: "duration" }] : []));
    descriptionClass = input('', ...(ngDevMode ? [{ debugName: "descriptionClass" }] : []));
    classes = input({}, ...(ngDevMode ? [{ debugName: "classes" }] : []));
    unstyled = input(false, ...(ngDevMode ? [{ debugName: "unstyled" }] : []));
    userClass = input('', ...(ngDevMode ? [{ debugName: "userClass", alias: 'class' }] : [{ alias: 'class' }]));
    style = input({}, ...(ngDevMode ? [{ debugName: "style" }] : []));
    _mounted = signal(false, ...(ngDevMode ? [{ debugName: "_mounted" }] : []));
    _removed = signal(false, ...(ngDevMode ? [{ debugName: "_removed" }] : []));
    _swiping = signal(false, ...(ngDevMode ? [{ debugName: "_swiping" }] : []));
    _swipeOut = signal(false, ...(ngDevMode ? [{ debugName: "_swipeOut" }] : []));
    _offsetBeforeRemove = signal(0, ...(ngDevMode ? [{ debugName: "_offsetBeforeRemove" }] : []));
    _initialHeight = signal(0, ...(ngDevMode ? [{ debugName: "_initialHeight" }] : []));
    _toastRef = viewChild.required('toastRef');
    _classes = computed(() => ({
        ...defaultClasses,
        ...this.classes(),
    }), ...(ngDevMode ? [{ debugName: "_classes" }] : []));
    _isFront = computed(() => this.index() === 0, ...(ngDevMode ? [{ debugName: "_isFront" }] : []));
    _isVisible = computed(() => this.index() + 1 <= this.visibleToasts(), ...(ngDevMode ? [{ debugName: "_isVisible" }] : []));
    _toastType = computed(() => this.toast().type ?? 'default', ...(ngDevMode ? [{ debugName: "_toastType" }] : []));
    _toastClass = computed(() => this.toast().class ?? '', ...(ngDevMode ? [{ debugName: "_toastClass" }] : []));
    _toastPosition = computed(() => this.toast().position ?? this.position(), ...(ngDevMode ? [{ debugName: "_toastPosition" }] : []));
    _toastDescriptionClass = computed(() => this.toast().descriptionClass ?? '', ...(ngDevMode ? [{ debugName: "_toastDescriptionClass" }] : []));
    _heightIndex = computed(() => this._heights().findIndex((height) => height.toastId === this.toast().id), ...(ngDevMode ? [{ debugName: "_heightIndex" }] : []));
    _offset = linkedSignal(...(ngDevMode ? [{ debugName: "_offset", source: () => ({
                heightIndex: this._heightIndex(),
                toastsHeightBefore: this._toastsHeightBefore(),
            }),
            computation: ({ heightIndex, toastsHeightBefore }) => Math.round(heightIndex * this._config.gap + toastsHeightBefore) }] : [{
            source: () => ({
                heightIndex: this._heightIndex(),
                toastsHeightBefore: this._toastsHeightBefore(),
            }),
            computation: ({ heightIndex, toastsHeightBefore }) => Math.round(heightIndex * this._config.gap + toastsHeightBefore),
        }]));
    _closeTimerStartTimeRef = 0;
    _lastCloseTimerStartTimeRef = 0;
    _pointerStartRef = null;
    _coords = computed(() => this._toastPosition().split('-'), ...(ngDevMode ? [{ debugName: "_coords" }] : []));
    _toastsHeightBefore = computed(() => this._heights().reduce((prev, curr, reducerIndex) => {
        if (reducerIndex >= this._heightIndex())
            return prev;
        return prev + curr.height;
    }, 0), ...(ngDevMode ? [{ debugName: "_toastsHeightBefore" }] : []));
    _invert = computed(() => this.toast().invert ?? this.invert(), ...(ngDevMode ? [{ debugName: "_invert" }] : []));
    _closeButton = computed(() => this.toast().closeButton ?? this.closeButton(), ...(ngDevMode ? [{ debugName: "_closeButton" }] : []));
    _disabled = computed(() => this._toastType() === 'loading', ...(ngDevMode ? [{ debugName: "_disabled" }] : []));
    _timeoutId;
    _remainingTime = 0;
    _isPromiseLoadingOrInfiniteDuration = computed(() => (this.toast().promise && this._toastType() === 'loading') || this.toast().duration === Number.POSITIVE_INFINITY, ...(ngDevMode ? [{ debugName: "_isPromiseLoadingOrInfiniteDuration" }] : []));
    _toastClasses = computed(() => clsx(this.userClass(), this._toastClass(), this._classes().toast, this.toast().classes?.toast, this._classes()[this._toastType()], this.toast().classes?.[this._toastType()]), ...(ngDevMode ? [{ debugName: "_toastClasses" }] : []));
    _toastStyle = computed(() => ({
        '--index': `${this.index()}`,
        '--toasts-before': `${this.index()}`,
        '--z-index': `${this._toasts().length - this.index()}`,
        '--offset': `${this._removed() ? this._offsetBeforeRemove() : this._offset()}px`,
        '--initial-height': this.expandByDefault() ? 'auto' : `${this._initialHeight()}px`,
        ...this.style(),
    }), ...(ngDevMode ? [{ debugName: "_toastStyle" }] : []));
    _actionButtonClasses = computed(() => clsx(this._classes().actionButton, this.toast().classes?.actionButton), ...(ngDevMode ? [{ debugName: "_actionButtonClasses" }] : []));
    _cancelButtonClasses = computed(() => clsx(this._classes().cancelButton, this.toast().classes?.cancelButton), ...(ngDevMode ? [{ debugName: "_cancelButtonClasses" }] : []));
    _toastDescriptionClasses = computed(() => clsx(this.descriptionClass(), this._toastDescriptionClass(), this._classes().description, this.toast().classes?.description), ...(ngDevMode ? [{ debugName: "_toastDescriptionClasses" }] : []));
    _titleClasses = computed(() => clsx(this._classes().title, this.toast().classes?.title), ...(ngDevMode ? [{ debugName: "_titleClasses" }] : []));
    _closeButtonClasses = computed(() => clsx(this._classes().closeButton, this.toast().classes?.closeButton), ...(ngDevMode ? [{ debugName: "_closeButtonClasses" }] : []));
    constructor() {
        effect(() => {
            if (this.toast().updated) {
                // if the toast has been updated after the initial render,
                // we want to reset the timer and set the remaining time to the
                // new duration
                clearTimeout(this._timeoutId);
                this._remainingTime = this.toast().duration ?? this.duration() ?? this._config.toastLifetime;
                this.startTimer();
            }
        });
        afterRenderEffect((onCleanup) => {
            if (!this._isPromiseLoadingOrInfiniteDuration()) {
                if (this.expanded() || this.interacting()) {
                    this.pauseTimer();
                }
                else {
                    this.startTimer();
                }
            }
            onCleanup(() => clearTimeout(this._timeoutId));
        });
        effect(() => {
            if (this.toast().delete) {
                this.deleteToast();
            }
        });
    }
    ngAfterViewInit() {
        this._remainingTime = this.toast().duration ?? this.duration() ?? this._config.toastLifetime;
        this._mounted.set(true);
        const height = this._toastRef().nativeElement.getBoundingClientRect().height;
        this._initialHeight.set(height);
        this._addHeight({ toastId: this.toast().id, height, position: this._toastPosition() });
    }
    ngOnDestroy() {
        clearTimeout(this._timeoutId);
        this._removeHeight(this.toast().id);
    }
    deleteToast() {
        this._removed.set(true);
        this._offsetBeforeRemove.set(this._offset());
        this._removeHeight(this.toast().id);
        setTimeout(() => {
            this._dismiss(this.toast().id);
        }, this._config.timeBeforeUnmount);
    }
    // If toast's duration changes, it will be out of sync with the
    // remainingAtTimeout, so we know we need to restart the timer
    // with the new duration
    // Pause the timer on each hover
    pauseTimer() {
        if (this._lastCloseTimerStartTimeRef < this._closeTimerStartTimeRef) {
            // Get the elapsed time since the timer started
            const elapsedTime = new Date().getTime() - this._closeTimerStartTimeRef;
            this._remainingTime = this._remainingTime - elapsedTime;
        }
        this._lastCloseTimerStartTimeRef = new Date().getTime();
    }
    startTimer() {
        this._closeTimerStartTimeRef = new Date().getTime();
        // Let the toast know it has started
        this._timeoutId = setTimeout(() => {
            this.toast().onAutoClose?.(this.toast());
            this.deleteToast();
        }, this._remainingTime);
    }
    onPointerDown(event) {
        if (this._disabled() || !this.toast().dismissible)
            return;
        this._offsetBeforeRemove.set(this._offset());
        const target = event.target;
        // Ensure we maintain correct pointer capture even when going outside the toast (e.g. when swiping)
        target.setPointerCapture(event.pointerId);
        if (target.tagName === 'BUTTON') {
            return;
        }
        this._swiping.set(true);
        this._pointerStartRef = { x: event.clientX, y: event.clientY };
    }
    onPointerUp() {
        if (this._swipeOut() || !this.toast().dismissible)
            return;
        this._pointerStartRef = null;
        const swipeAmount = Number(this._toastRef().nativeElement.style.getPropertyValue('--swipe-amount').replace('px', '') || 0);
        // Remove only if threshold is met
        if (Math.abs(swipeAmount) >= this._config.swipeThreshold) {
            this._offsetBeforeRemove.set(this._offset());
            this.toast().onDismiss?.(this.toast());
            this.deleteToast();
            this._swipeOut.set(true);
            return;
        }
        this._toastRef().nativeElement.style.setProperty('--swipe-amount', '0px');
        this._swiping.set(false);
    }
    onPointerMove(event) {
        if (!this._pointerStartRef || !this.toast().dismissible)
            return;
        const yPosition = event.clientY - this._pointerStartRef.y;
        const xPosition = event.clientX - this._pointerStartRef.x;
        const clamp = this._coords()[0] === 'top' ? Math.min : Math.max;
        const clampedY = clamp(0, yPosition);
        const swipeStartThreshold = event.pointerType === 'touch' ? 10 : 2;
        const isAllowedToSwipe = Math.abs(clampedY) > swipeStartThreshold;
        if (isAllowedToSwipe) {
            this._toastRef().nativeElement.style.setProperty('--swipe-amount', `${yPosition}px`);
        }
        else if (Math.abs(xPosition) > swipeStartThreshold) {
            // User is swiping in wrong direction, so we disable swipe gesture
            // for the current pointer down interaction
            this._pointerStartRef = null;
        }
    }
    onCloseButtonClick() {
        if (this._disabled() || !this.toast().dismissible)
            return;
        this.deleteToast();
        this.toast().onDismiss?.(this.toast());
    }
    onCancelClick() {
        const toast = this.toast();
        if (!toast.dismissible)
            return;
        this.deleteToast();
        if (toast.cancel?.onClick) {
            toast.cancel.onClick();
        }
    }
    onActionClick(event) {
        const toast = this.toast();
        toast.action?.onClick(event);
        if (event.defaultPrevented)
            return;
        this.deleteToast();
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSonnerToast, deps: [], target: i0.ɵɵFactoryTarget.Component });
    /** @nocollapse */ static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "20.3.17", type: BrnSonnerToast, isStandalone: true, selector: "brn-sonner-toast", inputs: { toast: { classPropertyName: "toast", publicName: "toast", isSignal: true, isRequired: true, transformFunction: null }, index: { classPropertyName: "index", publicName: "index", isSignal: true, isRequired: true, transformFunction: null }, expanded: { classPropertyName: "expanded", publicName: "expanded", isSignal: true, isRequired: true, transformFunction: null }, invert: { classPropertyName: "invert", publicName: "invert", isSignal: true, isRequired: true, transformFunction: null }, position: { classPropertyName: "position", publicName: "position", isSignal: true, isRequired: true, transformFunction: null }, visibleToasts: { classPropertyName: "visibleToasts", publicName: "visibleToasts", isSignal: true, isRequired: true, transformFunction: null }, expandByDefault: { classPropertyName: "expandByDefault", publicName: "expandByDefault", isSignal: true, isRequired: true, transformFunction: null }, closeButton: { classPropertyName: "closeButton", publicName: "closeButton", isSignal: true, isRequired: true, transformFunction: null }, interacting: { classPropertyName: "interacting", publicName: "interacting", isSignal: true, isRequired: true, transformFunction: null }, cancelButtonStyle: { classPropertyName: "cancelButtonStyle", publicName: "cancelButtonStyle", isSignal: true, isRequired: false, transformFunction: null }, actionButtonStyle: { classPropertyName: "actionButtonStyle", publicName: "actionButtonStyle", isSignal: true, isRequired: false, transformFunction: null }, duration: { classPropertyName: "duration", publicName: "duration", isSignal: true, isRequired: false, transformFunction: null }, descriptionClass: { classPropertyName: "descriptionClass", publicName: "descriptionClass", isSignal: true, isRequired: false, transformFunction: null }, classes: { classPropertyName: "classes", publicName: "classes", isSignal: true, isRequired: false, transformFunction: null }, unstyled: { classPropertyName: "unstyled", publicName: "unstyled", isSignal: true, isRequired: false, transformFunction: null }, userClass: { classPropertyName: "userClass", publicName: "class", isSignal: true, isRequired: false, transformFunction: null }, style: { classPropertyName: "style", publicName: "style", isSignal: true, isRequired: false, transformFunction: null } }, viewQueries: [{ propertyName: "_toastRef", first: true, predicate: ["toastRef"], descendants: true, isSignal: true }], ngImport: i0, template: `
		<li
			#toastRef
			data-sonner-toast
			[attr.aria-live]="toast().important ? 'assertive' : 'polite'"
			aria-atomic="true"
			role="status"
			tabindex="0"
			[class]="_toastClasses()"
			[attr.data-styled]="!(toast().component || toast().unstyled || unstyled())"
			[attr.data-mounted]="_mounted()"
			[attr.data-promise]="!!toast().promise"
			[attr.data-removed]="_removed()"
			[attr.data-visible]="_isVisible()"
			[attr.data-y-position]="_coords()[0]"
			[attr.data-x-position]="_coords()[1]"
			[attr.data-index]="index()"
			[attr.data-front]="_isFront()"
			[attr.data-swiping]="_swiping()"
			[attr.data-dismissible]="toast().dismissible"
			[attr.data-type]="_toastType()"
			[attr.data-invert]="_invert()"
			[attr.data-swipe-out]="_swipeOut()"
			[attr.data-expanded]="expanded() || (expandByDefault() && _mounted())"
			[style]="_toastStyle()"
			(pointerdown)="onPointerDown($event)"
			(pointerup)="onPointerUp()"
			(pointermove)="onPointerMove($event)"
		>
			@if (_closeButton() && !toast().component) {
				<button
					aria-label="Close toast"
					[attr.data-disabled]="_disabled()"
					data-close-button
					(click)="onCloseButtonClick()"
					[class]="_closeButtonClasses()"
				>
					<svg
						xmlns="http://www.w3.org/2000/svg"
						width="12"
						height="12"
						viewBox="0 0 24 24"
						fill="none"
						stroke="currentColor"
						stroke-width="1.5"
						stroke-linecap="round"
						stroke-linejoin="round"
					>
						<line x1="18" y1="6" x2="6" y2="18" />
						<line x1="6" y1="6" x2="18" y2="18" />
					</svg>
				</button>
			}

			@if (toast().component) {
				<ng-container *ngComponentOutlet="toast().component | asComponent; inputs: toast().componentProps" />
			} @else {
				@if (_toastType() !== 'default' || toast().icon || toast().promise) {
					<div data-icon>
						@if (_toastType() === 'loading' && !toast().icon) {
							<ng-content select="[loading-icon]" />
						}
						@if (toast().icon) {
							<ng-container *ngComponentOutlet="toast().icon | asComponent; inputs: toast().componentProps" />
						} @else {
							@switch (_toastType()) {
								@case ('success') {
									<ng-content select="[success-icon]" />
								}
								@case ('error') {
									<ng-content select="[error-icon]" />
								}
								@case ('warning') {
									<ng-content select="[warning-icon]" />
								}
								@case ('info') {
									<ng-content select="[info-icon]" />
								}
							}
						}
					</div>
				}
				<div data-content>
					@if (toast().title; as title) {
						<div data-title [class]="_titleClasses()">
							@if (title | isString) {
								{{ toast().title }}
							} @else {
								<ng-container *ngComponentOutlet="title | asComponent; inputs: toast().componentProps" />
							}
						</div>
					}
					@if (toast().description; as description) {
						<div data-description [class]="_toastDescriptionClasses()">
							@if (description | isString) {
								{{ toast().description }}
							} @else {
								<ng-container *ngComponentOutlet="description | asComponent; inputs: toast().componentProps" />
							}
						</div>
					}
				</div>
				@if (toast().cancel; as cancel) {
					<button
						data-button
						data-cancel
						[style]="cancelButtonStyle() ?? toast().cancelButtonStyle"
						[class]="_cancelButtonClasses()"
						(click)="onCancelClick()"
					>
						{{ cancel.label }}
					</button>
				}
				@if (toast().action; as action) {
					<button
						data-button
						[style]="actionButtonStyle() ?? toast().actionButtonStyle"
						[class]="_actionButtonClasses()"
						(click)="onActionClick($event)"
					>
						{{ action.label }}
					</button>
				}
			}
		</li>
	`, isInline: true, dependencies: [{ kind: "directive", type: NgComponentOutlet, selector: "[ngComponentOutlet]", inputs: ["ngComponentOutlet", "ngComponentOutletInputs", "ngComponentOutletInjector", "ngComponentOutletEnvironmentInjector", "ngComponentOutletContent", "ngComponentOutletNgModule", "ngComponentOutletNgModuleFactory"], exportAs: ["ngComponentOutlet"] }, { kind: "pipe", type: IsStringPipe, name: "isString" }, { kind: "pipe", type: AsComponentPipe, name: "asComponent" }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSonnerToast, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-sonner-toast',
                    imports: [NgComponentOutlet, IsStringPipe, AsComponentPipe],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    template: `
		<li
			#toastRef
			data-sonner-toast
			[attr.aria-live]="toast().important ? 'assertive' : 'polite'"
			aria-atomic="true"
			role="status"
			tabindex="0"
			[class]="_toastClasses()"
			[attr.data-styled]="!(toast().component || toast().unstyled || unstyled())"
			[attr.data-mounted]="_mounted()"
			[attr.data-promise]="!!toast().promise"
			[attr.data-removed]="_removed()"
			[attr.data-visible]="_isVisible()"
			[attr.data-y-position]="_coords()[0]"
			[attr.data-x-position]="_coords()[1]"
			[attr.data-index]="index()"
			[attr.data-front]="_isFront()"
			[attr.data-swiping]="_swiping()"
			[attr.data-dismissible]="toast().dismissible"
			[attr.data-type]="_toastType()"
			[attr.data-invert]="_invert()"
			[attr.data-swipe-out]="_swipeOut()"
			[attr.data-expanded]="expanded() || (expandByDefault() && _mounted())"
			[style]="_toastStyle()"
			(pointerdown)="onPointerDown($event)"
			(pointerup)="onPointerUp()"
			(pointermove)="onPointerMove($event)"
		>
			@if (_closeButton() && !toast().component) {
				<button
					aria-label="Close toast"
					[attr.data-disabled]="_disabled()"
					data-close-button
					(click)="onCloseButtonClick()"
					[class]="_closeButtonClasses()"
				>
					<svg
						xmlns="http://www.w3.org/2000/svg"
						width="12"
						height="12"
						viewBox="0 0 24 24"
						fill="none"
						stroke="currentColor"
						stroke-width="1.5"
						stroke-linecap="round"
						stroke-linejoin="round"
					>
						<line x1="18" y1="6" x2="6" y2="18" />
						<line x1="6" y1="6" x2="18" y2="18" />
					</svg>
				</button>
			}

			@if (toast().component) {
				<ng-container *ngComponentOutlet="toast().component | asComponent; inputs: toast().componentProps" />
			} @else {
				@if (_toastType() !== 'default' || toast().icon || toast().promise) {
					<div data-icon>
						@if (_toastType() === 'loading' && !toast().icon) {
							<ng-content select="[loading-icon]" />
						}
						@if (toast().icon) {
							<ng-container *ngComponentOutlet="toast().icon | asComponent; inputs: toast().componentProps" />
						} @else {
							@switch (_toastType()) {
								@case ('success') {
									<ng-content select="[success-icon]" />
								}
								@case ('error') {
									<ng-content select="[error-icon]" />
								}
								@case ('warning') {
									<ng-content select="[warning-icon]" />
								}
								@case ('info') {
									<ng-content select="[info-icon]" />
								}
							}
						}
					</div>
				}
				<div data-content>
					@if (toast().title; as title) {
						<div data-title [class]="_titleClasses()">
							@if (title | isString) {
								{{ toast().title }}
							} @else {
								<ng-container *ngComponentOutlet="title | asComponent; inputs: toast().componentProps" />
							}
						</div>
					}
					@if (toast().description; as description) {
						<div data-description [class]="_toastDescriptionClasses()">
							@if (description | isString) {
								{{ toast().description }}
							} @else {
								<ng-container *ngComponentOutlet="description | asComponent; inputs: toast().componentProps" />
							}
						</div>
					}
				</div>
				@if (toast().cancel; as cancel) {
					<button
						data-button
						data-cancel
						[style]="cancelButtonStyle() ?? toast().cancelButtonStyle"
						[class]="_cancelButtonClasses()"
						(click)="onCancelClick()"
					>
						{{ cancel.label }}
					</button>
				}
				@if (toast().action; as action) {
					<button
						data-button
						[style]="actionButtonStyle() ?? toast().actionButtonStyle"
						[class]="_actionButtonClasses()"
						(click)="onActionClick($event)"
					>
						{{ action.label }}
					</button>
				}
			}
		</li>
	`,
                }]
        }], ctorParameters: () => [], propDecorators: { toast: [{ type: i0.Input, args: [{ isSignal: true, alias: "toast", required: true }] }], index: [{ type: i0.Input, args: [{ isSignal: true, alias: "index", required: true }] }], expanded: [{ type: i0.Input, args: [{ isSignal: true, alias: "expanded", required: true }] }], invert: [{ type: i0.Input, args: [{ isSignal: true, alias: "invert", required: true }] }], position: [{ type: i0.Input, args: [{ isSignal: true, alias: "position", required: true }] }], visibleToasts: [{ type: i0.Input, args: [{ isSignal: true, alias: "visibleToasts", required: true }] }], expandByDefault: [{ type: i0.Input, args: [{ isSignal: true, alias: "expandByDefault", required: true }] }], closeButton: [{ type: i0.Input, args: [{ isSignal: true, alias: "closeButton", required: true }] }], interacting: [{ type: i0.Input, args: [{ isSignal: true, alias: "interacting", required: true }] }], cancelButtonStyle: [{ type: i0.Input, args: [{ isSignal: true, alias: "cancelButtonStyle", required: false }] }], actionButtonStyle: [{ type: i0.Input, args: [{ isSignal: true, alias: "actionButtonStyle", required: false }] }], duration: [{ type: i0.Input, args: [{ isSignal: true, alias: "duration", required: false }] }], descriptionClass: [{ type: i0.Input, args: [{ isSignal: true, alias: "descriptionClass", required: false }] }], classes: [{ type: i0.Input, args: [{ isSignal: true, alias: "classes", required: false }] }], unstyled: [{ type: i0.Input, args: [{ isSignal: true, alias: "unstyled", required: false }] }], userClass: [{ type: i0.Input, args: [{ isSignal: true, alias: "class", required: false }] }], style: [{ type: i0.Input, args: [{ isSignal: true, alias: "style", required: false }] }], _toastRef: [{ type: i0.ViewChild, args: ['toastRef', { isSignal: true }] }] } });

class ToastFilterPipe {
    transform(toasts, index, position) {
        return toasts.filter((toast) => (!toast.position && index === 0) || toast.position === position);
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: ToastFilterPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
    /** @nocollapse */ static ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "14.0.0", version: "20.3.17", ngImport: i0, type: ToastFilterPipe, isStandalone: true, name: "toastFilter" });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: ToastFilterPipe, decorators: [{
            type: Pipe,
            args: [{ name: 'toastFilter' }]
        }] });

class BrnSonnerToaster {
    _platformId = inject(PLATFORM_ID);
    _config = injectBrnSonnerToasterConfig();
    _document = inject(DOCUMENT);
    _window = this._document.defaultView;
    _dir = inject(Directionality);
    _toasts = toastState.toasts;
    _heights = toastState.heights;
    _reset = toastState.reset;
    invert = input(false, ...(ngDevMode ? [{ debugName: "invert", transform: booleanAttribute }] : [{
            transform: booleanAttribute,
        }]));
    theme = input('light', ...(ngDevMode ? [{ debugName: "theme" }] : []));
    position = input('bottom-right', ...(ngDevMode ? [{ debugName: "position" }] : []));
    hotKey = input(['altKey', 'KeyT'], ...(ngDevMode ? [{ debugName: "hotKey" }] : []));
    richColors = input(false, ...(ngDevMode ? [{ debugName: "richColors", transform: booleanAttribute }] : [{
            transform: booleanAttribute,
        }]));
    expand = input(false, ...(ngDevMode ? [{ debugName: "expand", transform: booleanAttribute }] : [{
            transform: booleanAttribute,
        }]));
    duration = input(this._config.toastLifetime, ...(ngDevMode ? [{ debugName: "duration", transform: numberAttribute }] : [{
            transform: numberAttribute,
        }]));
    visibleToasts = input(this._config.visibleToastsAmount, ...(ngDevMode ? [{ debugName: "visibleToasts", transform: numberAttribute }] : [{
            transform: numberAttribute,
        }]));
    closeButton = input(false, ...(ngDevMode ? [{ debugName: "closeButton", transform: booleanAttribute }] : [{
            transform: booleanAttribute,
        }]));
    toastOptions = input({}, ...(ngDevMode ? [{ debugName: "toastOptions" }] : []));
    offset = input(null, ...(ngDevMode ? [{ debugName: "offset" }] : []));
    userClass = input('', ...(ngDevMode ? [{ debugName: "userClass", alias: 'class' }] : [{ alias: 'class' }]));
    style = input({}, ...(ngDevMode ? [{ debugName: "style" }] : []));
    _possiblePositions = computed(() => Array.from(new Set([
        this.position(),
        ...this._toasts()
            .filter((toast) => toast.position)
            .map((toast) => toast.position),
    ].filter(Boolean))), ...(ngDevMode ? [{ debugName: "_possiblePositions" }] : []));
    _expanded = linkedSignal(...(ngDevMode ? [{ debugName: "_expanded", source: this._toasts,
            computation: (toasts) => toasts.length < 1 }] : [{
            source: this._toasts,
            computation: (toasts) => toasts.length < 1,
        }]));
    _actualTheme = linkedSignal(...(ngDevMode ? [{ debugName: "_actualTheme", source: this.theme,
            computation: (newTheme) => this.getActualTheme(newTheme) }] : [{
            source: this.theme,
            computation: (newTheme) => this.getActualTheme(newTheme),
        }]));
    _interacting = signal(false, ...(ngDevMode ? [{ debugName: "_interacting" }] : []));
    /** internal **/
    direction = this._dir.valueSignal;
    _listRef = viewChild('listRef', ...(ngDevMode ? [{ debugName: "_listRef" }] : []));
    _lastFocusedElementRef = signal(null, ...(ngDevMode ? [{ debugName: "_lastFocusedElementRef" }] : []));
    _isFocusWithinRef = signal(false, ...(ngDevMode ? [{ debugName: "_isFocusWithinRef" }] : []));
    _hotKeyLabel = computed(() => this.hotKey().join('+').replace(/Key/g, '').replace(/Digit/g, ''), ...(ngDevMode ? [{ debugName: "_hotKeyLabel" }] : []));
    _toasterStyles = computed(() => ({
        '--front-toast-height': `${this._heights()[0]?.height}px`,
        '--offset': typeof this.offset() === 'number' ? `${this.offset()}px` : (this.offset() ?? `${this._config.viewPortOffset}`),
        '--width': `${this._config.toastWidth}px`,
        '--gap': `${this._config.gap}px`,
        ...this.style(),
    }), ...(ngDevMode ? [{ debugName: "_toasterStyles" }] : []));
    _loadingIcon = contentChild('loadingIcon', ...(ngDevMode ? [{ debugName: "_loadingIcon", read: TemplateRef }] : [{ read: TemplateRef }]));
    _successIcon = contentChild('successIcon', ...(ngDevMode ? [{ debugName: "_successIcon", read: TemplateRef }] : [{ read: TemplateRef }]));
    _errorIcon = contentChild('errorIcon', ...(ngDevMode ? [{ debugName: "_errorIcon", read: TemplateRef }] : [{ read: TemplateRef }]));
    _warningIcon = contentChild('warningIcon', ...(ngDevMode ? [{ debugName: "_warningIcon", read: TemplateRef }] : [{ read: TemplateRef }]));
    _infoIcon = contentChild('infoIcon', ...(ngDevMode ? [{ debugName: "_infoIcon", read: TemplateRef }] : [{ read: TemplateRef }]));
    constructor() {
        this._reset();
        const destroyRef = inject(DestroyRef);
        afterNextRender(() => {
            this._document.addEventListener('keydown', this.handleKeydown);
            const window = this._window;
            if (window) {
                this._window
                    .matchMedia('(prefers-color-scheme: dark)')
                    .addEventListener('change', this.handleThemePreferenceChange);
            }
            destroyRef.onDestroy(() => {
                this._document.removeEventListener('keydown', this.handleKeydown);
                if (window) {
                    window
                        .matchMedia('(prefers-color-scheme: dark)')
                        .removeEventListener('change', this.handleThemePreferenceChange);
                }
            });
        });
    }
    handleBlur(event) {
        if (this._isFocusWithinRef() && !event.target.contains(event.relatedTarget)) {
            this._isFocusWithinRef.set(false);
            if (this._lastFocusedElementRef()) {
                this._lastFocusedElementRef()?.focus({ preventScroll: true });
                this._lastFocusedElementRef.set(null);
            }
        }
    }
    handleFocus(event) {
        const isNotDismissible = event.target instanceof HTMLElement && event.target.dataset['dismissible'] === 'false';
        if (isNotDismissible)
            return;
        if (!this._isFocusWithinRef()) {
            this._isFocusWithinRef.set(true);
            this._lastFocusedElementRef.set(event.relatedTarget);
        }
    }
    handlePointerDown(event) {
        const isNotDismissible = event.target instanceof HTMLElement && event.target.dataset['dismissible'] === 'false';
        if (isNotDismissible)
            return;
        this._interacting.set(true);
    }
    handleMouseLeave() {
        if (!this._interacting()) {
            this._expanded.set(false);
        }
    }
    handleKeydown = (event) => {
        const listRef = this._listRef()?.nativeElement;
        if (!listRef)
            return;
        const isHotkeyPressed = this.hotKey().every((key) => event[key] || event.code === key);
        if (isHotkeyPressed) {
            this._expanded.set(true);
            listRef.focus();
        }
        if (event.code === 'Escape' &&
            (this._document.activeElement === listRef || listRef.contains(this._document.activeElement))) {
            this._expanded.set(false);
        }
    };
    handleThemePreferenceChange = ({ matches }) => {
        if (this.theme() === 'system') {
            this._actualTheme.set(matches ? 'dark' : 'light');
        }
    };
    getActualTheme(theme) {
        if (theme !== 'system') {
            return theme;
        }
        if (isPlatformBrowser(this._platformId) && this._window) {
            const prefersDark = this._window.matchMedia?.('(prefers-color-scheme: dark)').matches;
            return prefersDark ? 'dark' : 'light';
        }
        return 'light';
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSonnerToaster, deps: [], target: i0.ɵɵFactoryTarget.Component });
    /** @nocollapse */ static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.0.0", version: "20.3.17", type: BrnSonnerToaster, isStandalone: true, selector: "brn-sonner-toaster", inputs: { invert: { classPropertyName: "invert", publicName: "invert", isSignal: true, isRequired: false, transformFunction: null }, theme: { classPropertyName: "theme", publicName: "theme", isSignal: true, isRequired: false, transformFunction: null }, position: { classPropertyName: "position", publicName: "position", isSignal: true, isRequired: false, transformFunction: null }, hotKey: { classPropertyName: "hotKey", publicName: "hotKey", isSignal: true, isRequired: false, transformFunction: null }, richColors: { classPropertyName: "richColors", publicName: "richColors", isSignal: true, isRequired: false, transformFunction: null }, expand: { classPropertyName: "expand", publicName: "expand", isSignal: true, isRequired: false, transformFunction: null }, duration: { classPropertyName: "duration", publicName: "duration", isSignal: true, isRequired: false, transformFunction: null }, visibleToasts: { classPropertyName: "visibleToasts", publicName: "visibleToasts", isSignal: true, isRequired: false, transformFunction: null }, closeButton: { classPropertyName: "closeButton", publicName: "closeButton", isSignal: true, isRequired: false, transformFunction: null }, toastOptions: { classPropertyName: "toastOptions", publicName: "toastOptions", isSignal: true, isRequired: false, transformFunction: null }, offset: { classPropertyName: "offset", publicName: "offset", isSignal: true, isRequired: false, transformFunction: null }, userClass: { classPropertyName: "userClass", publicName: "class", isSignal: true, isRequired: false, transformFunction: null }, style: { classPropertyName: "style", publicName: "style", isSignal: true, isRequired: false, transformFunction: null } }, queries: [{ propertyName: "_loadingIcon", first: true, predicate: ["loadingIcon"], descendants: true, read: TemplateRef, isSignal: true }, { propertyName: "_successIcon", first: true, predicate: ["successIcon"], descendants: true, read: TemplateRef, isSignal: true }, { propertyName: "_errorIcon", first: true, predicate: ["errorIcon"], descendants: true, read: TemplateRef, isSignal: true }, { propertyName: "_warningIcon", first: true, predicate: ["warningIcon"], descendants: true, read: TemplateRef, isSignal: true }, { propertyName: "_infoIcon", first: true, predicate: ["infoIcon"], descendants: true, read: TemplateRef, isSignal: true }], viewQueries: [{ propertyName: "_listRef", first: true, predicate: ["listRef"], descendants: true, isSignal: true }], ngImport: i0, template: `
		@if (_toasts().length > 0) {
			<section [attr.aria-label]="'Notifications ' + _hotKeyLabel()" [tabIndex]="-1">
				@for (pos of _possiblePositions(); track pos) {
					<ol
						#listRef
						[tabIndex]="-1"
						[class]="userClass()"
						data-sonner-toaster
						[attr.data-theme]="_actualTheme()"
						[attr.data-rich-colors]="richColors()"
						[attr.dir]="direction()"
						[attr.data-y-position]="pos.split('-')[0]"
						[attr.data-x-position]="pos.split('-')[1]"
						[style]="_toasterStyles()"
						(blur)="handleBlur($event)"
						(focus)="handleFocus($event)"
						(mouseenter)="_expanded.set(true)"
						(mousemove)="_expanded.set(true)"
						(mouseleave)="handleMouseLeave()"
						(pointerdown)="handlePointerDown($event)"
						(pointerup)="_interacting.set(false)"
					>
						@for (toast of _toasts() | toastFilter: $index : pos; track toast.id) {
							<brn-sonner-toast
								[index]="$index"
								[toast]="toast"
								[invert]="invert()"
								[visibleToasts]="visibleToasts()"
								[closeButton]="closeButton()"
								[interacting]="_interacting()"
								[position]="pos"
								[expandByDefault]="expand()"
								[expanded]="_expanded()"
								[actionButtonStyle]="toastOptions().actionButtonStyle"
								[cancelButtonStyle]="toastOptions().cancelButtonStyle"
								[class]="toastOptions().class ?? ''"
								[descriptionClass]="toastOptions().descriptionClass ?? ''"
								[classes]="toastOptions().classes ?? {}"
								[duration]="toastOptions().duration ?? duration()"
								[unstyled]="toastOptions().unstyled ?? false"
							>
								@if (_loadingIcon(); as loadingIcon) {
									<ng-container *ngTemplateOutlet="loadingIcon" loading-icon />
								} @else {
									<brn-sonner-loader [isVisible]="toast.type === 'loading'" loading-icon />
								}

								@if (_successIcon(); as successIcon) {
									<ng-container *ngTemplateOutlet="successIcon" success-icon />
								} @else {
									<brn-sonner-icon type="success" success-icon />
								}

								@if (_errorIcon(); as errorIcon) {
									<ng-container *ngTemplateOutlet="errorIcon" error-icon />
								} @else {
									<brn-sonner-icon type="error" error-icon />
								}

								@if (_warningIcon(); as warningIcon) {
									<ng-container *ngTemplateOutlet="warningIcon" warning-icon />
								} @else {
									<brn-sonner-icon type="warning" warning-icon />
								}

								@if (_infoIcon(); as infoIcon) {
									<ng-container *ngTemplateOutlet="infoIcon" info-icon />
								} @else {
									<brn-sonner-icon type="info" info-icon />
								}
							</brn-sonner-toast>
						}
					</ol>
				}
			</section>
		}
	`, isInline: true, styles: ["html[dir=ltr],[data-sonner-toaster][dir=ltr]{--toast-icon-margin-start: var(--brn-sonner-toast-icon-margin-start, -3px);--toast-icon-margin-end: var(--brn-sonner-toast-icon-margin-end, 4px);--toast-svg-margin-start: var(--brn-sonner-toast-svg-margin-start, -1px);--toast-svg-margin-end: var(--brn-sonner-toast-svg-margin-end, 0px);--toast-button-margin-start: var(--brn-sonner-toast-button-margin-start, auto);--toast-button-margin-end: var(--brn-sonner-toast-button-margin-end, 0);--toast-close-button-start: var(--brn-sonner-toast-close-button-start, 0);--toast-close-button-end: var(--brn-sonner-toast-close-button-end, unset);--toast-close-button-transform: var(--brn-sonner-toast-close-button-transform, translate(-35%, -35%))}html[dir=rtl],[data-sonner-toaster][dir=rtl]{--toast-icon-margin-start: var(--brn-sonner-rtl-toast-icon-margin-start, 4px);--toast-icon-margin-end: var(--brn-sonner-rtl-toast-icon-margin-end, -3px);--toast-svg-margin-start: var(--brn-sonner-rtl-toast-svg-margin-start, 0px);--toast-svg-margin-end: var(--brn-sonner-rtl-toast-svg-margin-end, -1px);--toast-button-margin-start: var(--brn-sonner-rtl-toast-button-margin-start, 0);--toast-button-margin-end: var(--brn-sonner-rtl-toast-button-margin-end, auto);--toast-close-button-start: var(--brn-sonner-rtl-toast-close-button-start, unset);--toast-close-button-end: var(--brn-sonner-rtl-toast-close-button-end, 0);--toast-close-button-transform: var(--brn-sonner-rtl-toast-close-button-transform, translate(35%, -35%))}[data-sonner-toaster]{position:fixed;width:var(--width);font-family:var( --brn-sonner-font-family, ui-sans-serif, system-ui, -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Helvetica Neue, Arial, Noto Sans, sans-serif, Apple Color Emoji, Segoe UI Emoji, Segoe UI Symbol, Noto Color Emoji );--gray1: hsl(0, 0%, 99%);--gray2: hsl(0, 0%, 97.3%);--gray3: hsl(0, 0%, 95.1%);--gray4: hsl(0, 0%, 93%);--gray5: hsl(0, 0%, 90.9%);--gray6: hsl(0, 0%, 88.7%);--gray7: hsl(0, 0%, 85.8%);--gray8: hsl(0, 0%, 78%);--gray9: hsl(0, 0%, 56.1%);--gray10: hsl(0, 0%, 52.3%);--gray11: hsl(0, 0%, 43.5%);--gray12: hsl(0, 0%, 9%);--border-radius: var(--brn-sonner-border-radius, 8px);box-sizing:border-box;padding:0;margin:0;list-style:none;outline:none;z-index:999999999}[data-sonner-toaster][data-x-position=right]{right:max(var(--offset),env(safe-area-inset-right))}[data-sonner-toaster][data-x-position=left]{left:max(var(--offset),env(safe-area-inset-left))}[data-sonner-toaster][data-x-position=center]{left:50%;transform:translate(-50%)}[data-sonner-toaster][data-y-position=top]{top:max(var(--offset),env(safe-area-inset-top))}[data-sonner-toaster][data-y-position=bottom]{bottom:max(var(--offset),env(safe-area-inset-bottom))}[data-sonner-toast]{--y: translateY(100%);--lift-amount: calc(var(--lift) * var(--gap));z-index:var(--z-index);position:absolute;opacity:0;transform:var(--y);filter:blur(0);touch-action:none;transition:transform .4s,opacity .4s,height .4s,box-shadow .2s;box-sizing:border-box;outline:none;overflow-wrap:anywhere}[data-sonner-toast][data-styled=true]{padding:16px;background:var(--normal-bg);border:1px solid var(--normal-border);color:var(--normal-text);border-radius:var(--border-radius);box-shadow:0 4px 12px #0000001a;width:var(--width);font-size:13px;display:flex;align-items:center;gap:6px}[data-sonner-toast]:focus-visible{box-shadow:0 4px 12px #0000001a,0 0 0 2px #0003}[data-sonner-toast][data-y-position=top]{top:0;--y: translateY(-100%);--lift: 1;--lift-amount: calc(1 * var(--gap))}[data-sonner-toast][data-y-position=bottom]{bottom:0;--y: translateY(100%);--lift: -1;--lift-amount: calc(var(--lift) * var(--gap))}[data-sonner-toast] [data-description]{font-weight:400;line-height:1.4;color:inherit}[data-sonner-toast] [data-title]{font-weight:500;line-height:1.5;color:inherit}[data-sonner-toast] [data-icon]{display:flex;height:16px;width:16px;position:relative;justify-content:flex-start;align-items:center;flex-shrink:0;margin-left:var(--toast-icon-margin-start);margin-right:var(--toast-icon-margin-end)}[data-sonner-toast][data-promise=true] [data-icon]>svg{opacity:0;transform:scale(.8);transform-origin:center;animation:sonner-fade-in .3s ease forwards}[data-sonner-toast] [data-icon]>*{flex-shrink:0}[data-sonner-toast] [data-icon] svg{margin-left:var(--toast-svg-margin-start);margin-right:var(--toast-svg-margin-end)}[data-sonner-toast] [data-content]{display:flex;flex-direction:column;gap:2px}[data-sonner-toast] [data-button]{border-radius:4px;padding-left:8px;padding-right:8px;height:24px;font-size:12px;color:var(--normal-bg);background:var(--normal-text);margin-left:var(--toast-button-margin-start);margin-right:var(--toast-button-margin-end);border:none;cursor:pointer;outline:none;display:flex;align-items:center;flex-shrink:0;transition:opacity .4s,box-shadow .2s}[data-sonner-toast] [data-button]:focus-visible{box-shadow:var(--brn-sonner-toast-focus-box-shadow, 0 0 0 2px rgba(0, 0, 0, .4))}[data-sonner-toast] [data-button]:first-of-type{margin-left:var(--toast-button-margin-start);margin-right:var(--toast-button-margin-end)}[data-sonner-toast] [data-cancel]{color:var(--normal-text);background:#00000014}[data-sonner-toast][data-theme=dark] [data-cancel]{background:#ffffff4d}[data-sonner-toast] [data-close-button]{position:absolute;left:var(--toast-close-button-start);right:var(--toast-close-button-end);top:0;height:20px;width:20px;display:flex;justify-content:center;align-items:center;padding:0;background:var(--brn-sonner-toast-close-button-background, var(--gray1));color:var(--brn-sonner-toast-close-button-color, var(--gray12));border:var(--brn-sonner-toast-close-button-border, 1px solid var(--gray4));transform:var(--toast-close-button-transform);border-radius:50%;cursor:pointer;z-index:1;transition:opacity .1s,background .2s,border-color .2s}[data-sonner-toast] [data-close-button]:focus-visible{box-shadow:0 4px 12px #0000001a,0 0 0 2px #0003}[data-sonner-toast] [data-disabled=true]{cursor:not-allowed}[data-sonner-toast]:hover [data-close-button]:hover{background:var(--brn-sonner-toast-close-button-hover-background, var(--gray2));color:var(--brn-sonner-toast-close-button-hover-color, var(--gray12));border-color:var(--brn-sonner-toast-close-button-hover-border-color, var(--gray5))}[data-sonner-toast][data-swiping=true]:before{content:\"\";position:absolute;left:0;right:0;height:100%;z-index:-1}[data-sonner-toast][data-y-position=top][data-swiping=true]:before{bottom:50%;transform:scaleY(3) translateY(50%)}[data-sonner-toast][data-y-position=bottom][data-swiping=true]:before{top:50%;transform:scaleY(3) translateY(-50%)}[data-sonner-toast][data-swiping=false][data-removed=true]:before{content:\"\";position:absolute;inset:0;transform:scaleY(2)}[data-sonner-toast]:after{content:\"\";position:absolute;left:0;height:calc(var(--gap) + 1px);bottom:100%;width:100%}[data-sonner-toast][data-mounted=true]{--y: translateY(0);opacity:1}[data-sonner-toast][data-expanded=false][data-front=false]{--scale: var(--toasts-before) * .05 + 1;--y: translateY(calc(var(--lift-amount) * var(--toasts-before))) scale(calc(-1 * var(--scale)));height:var(--front-toast-height)}[data-sonner-toast]>*{transition:opacity .4s}[data-sonner-toast][data-expanded=false][data-front=false][data-styled=true]>*{opacity:0}[data-sonner-toast][data-visible=false]{opacity:0;pointer-events:none}[data-sonner-toast][data-mounted=true][data-expanded=true]{--y: translateY(calc(var(--lift) * var(--offset)));height:var(--initial-height)}[data-sonner-toast][data-removed=true][data-front=true][data-swipe-out=false]{--y: translateY(calc(var(--lift) * -100%));opacity:0}[data-sonner-toast][data-removed=true][data-front=false][data-swipe-out=false][data-expanded=true]{--y: translateY(calc(var(--lift) * var(--offset) + var(--lift) * -100%));opacity:0}[data-sonner-toast][data-removed=true][data-front=false][data-swipe-out=false][data-expanded=false]{--y: translateY(40%);opacity:0;transition:transform .5s,opacity .2s}[data-sonner-toast][data-removed=true][data-front=false]:before{height:calc(var(--initial-height) + 20%)}[data-sonner-toast][data-swiping=true]{transform:var(--y) translateY(var(--swipe-amount, 0px));transition:none}[data-sonner-toast][data-swipe-out=true][data-y-position=bottom],[data-sonner-toast][data-swipe-out=true][data-y-position=top]{animation:swipe-out .2s ease-out forwards}@keyframes swipe-out{0%{transform:translateY(calc(var(--lift) * var(--offset) + var(--swipe-amount)));opacity:1}to{transform:translateY(calc(var(--lift) * var(--offset) + var(--swipe-amount) + var(--lift) * -100%));opacity:0}}@media (max-width: 600px){[data-sonner-toaster]{position:fixed;--mobile-offset: 16px;right:var(--mobile-offset);left:var(--mobile-offset);width:100%}[data-sonner-toaster] [data-sonner-toast]{left:0;right:0;width:calc(100% - 32px)}[data-sonner-toaster][data-x-position=left]{left:var(--mobile-offset)}[data-sonner-toaster][data-y-position=bottom]{bottom:20px}[data-sonner-toaster][data-y-position=top]{top:20px}[data-sonner-toaster][data-x-position=center]{left:var(--mobile-offset);right:var(--mobile-offset);transform:none}}[data-sonner-toaster][data-theme=light]{--normal-bg: var(--brn-sonner-toast-normal-background, #fff);--normal-border: var(--brn-sonner-toast-normal-border-color, var(--gray4));--normal-text: var(--brn-sonner-toast-normal-color, var(--gray12));--success-bg: var(--brn-sonner-toast-success-background, hsl(143, 85%, 96%));--success-border: var(--brn-sonner-toast-success-border, hsl(145, 92%, 91%));--success-text: var(--brn-sonner-toast-success-color, hsl(140, 100%, 27%));--info-bg: var(--brn-sonner-toast-info-background, hsl(208, 100%, 97%));--info-border: var(--brn-sonner-toast-info-border, hsl(221, 91%, 91%));--info-text: var(--brn-sonner-toast-info-color, hsl(210, 92%, 45%));--warning-bg: var(--brn-sonner-toast-warning-background, hsl(49, 100%, 97%));--warning-border: var(--brn-sonner-toast-warning-border, hsl(49, 91%, 91%));--warning-text: var(--brn-sonner-toast-warning-color, hsl(31, 92%, 45%));--error-bg: var(--brn-sonner-toast-error-background, hsl(359, 100%, 97%));--error-border: var(--brn-sonner-toast-error-border, hsl(359, 100%, 94%));--error-text: var(--brn-sonner-toast-error-color, hsl(360, 100%, 45%))}[data-sonner-toaster][data-theme=light] [data-sonner-toast][data-invert=true]{--normal-bg: var(--brn-sonner-toast-inverse-normal-background, #000);--normal-border: var(--brn-sonner-toast-inverse-normal-border-color, hsl(0, 0%, 20%));--normal-text: var(--brn-sonner-toast-inverse-normal-color, var(--gray1))}[data-sonner-toaster][data-theme=dark] [data-sonner-toast][data-invert=true]{--normal-bg: var(--brn-sonner-toast-inverse-dark-normal-background, #fff);--normal-border: var(--brn-sonner-toast-inverse-dark-normal-border-color, var(--gray3));--normal-text: var(--brn-sonner-toast-inverse-dark-normal-color, var(--gray12))}[data-sonner-toaster][data-theme=dark]{--normal-bg: var(--brn-sonner-toast-dark-normal-background, #000);--normal-border: var(--brn-sonner-toast-dark-normal-border-color, hsl(0, 0%, 20%));--normal-text: var(--brn-sonner-toast-dark-normal-color, var(--gray1));--success-bg: var(--brn-sonner-toast-dark-success-background, hsl(150, 100%, 6%));--success-border: var(--brn-sonner-toast-dark-success-border, hsl(147, 100%, 12%));--success-text: var(--brn-sonner-toast-dark-success-color, hsl(150, 86%, 65%));--info-bg: var(--brn-sonner-toast-dark-info-background, hsl(215, 100%, 6%));--info-border: var(--brn-sonner-toast-dark-info-border, hsl(223, 100%, 12%));--info-text: var(--brn-sonner-toast-dark-info-color, hsl(216, 87%, 65%));--warning-bg: var(--brn-sonner-toast-dark-warning-background, hsl(64, 100%, 6%));--warning-border: var(--brn-sonner-toast-dark-warning-border, hsl(60, 100%, 12%));--warning-text: var(--brn-sonner-toast-dark-warning-color, hsl(46, 87%, 65%));--error-bg: var(--brn-sonner-toast-dark-error-background, hsl(358, 76%, 10%));--error-border: var(--brn-sonner-toast-dark-error-border, hsl(357, 89%, 16%));--error-text: var(--brn-sonner-toast-dark-error-color, hsl(358, 100%, 81%))}[data-rich-colors=true] [data-sonner-toast][data-type=success],[data-rich-colors=true] [data-sonner-toast][data-type=success] [data-close-button]{background:var(--success-bg);border-color:var(--success-border);color:var(--success-text)}[data-rich-colors=true] [data-sonner-toast][data-type=info],[data-rich-colors=true] [data-sonner-toast][data-type=info] [data-close-button]{background:var(--info-bg);border-color:var(--info-border);color:var(--info-text)}[data-rich-colors=true] [data-sonner-toast][data-type=warning],[data-rich-colors=true] [data-sonner-toast][data-type=warning] [data-close-button]{background:var(--warning-bg);border-color:var(--warning-border);color:var(--warning-text)}[data-rich-colors=true] [data-sonner-toast][data-type=error],[data-rich-colors=true] [data-sonner-toast][data-type=error] [data-close-button]{background:var(--error-bg);border-color:var(--error-border);color:var(--error-text)}.sonner-loading-wrapper{--size: 16px;height:var(--size);width:var(--size);position:absolute;inset:0;z-index:10}.sonner-loading-wrapper[data-visible=false]{transform-origin:center;animation:sonner-fade-out .2s ease forwards}.sonner-spinner{position:relative;top:50%;left:50%;height:var(--size);width:var(--size)}.sonner-loading-bar{animation:sonner-spin 1.2s linear infinite;background:var(--gray11);border-radius:6px;height:8%;left:-10%;position:absolute;top:-3.9%;width:24%}.sonner-loading-bar:nth-child(1){animation-delay:-1.2s;transform:rotate(.0001deg) translate(146%)}.sonner-loading-bar:nth-child(2){animation-delay:-1.1s;transform:rotate(30deg) translate(146%)}.sonner-loading-bar:nth-child(3){animation-delay:-1s;transform:rotate(60deg) translate(146%)}.sonner-loading-bar:nth-child(4){animation-delay:-.9s;transform:rotate(90deg) translate(146%)}.sonner-loading-bar:nth-child(5){animation-delay:-.8s;transform:rotate(120deg) translate(146%)}.sonner-loading-bar:nth-child(6){animation-delay:-.7s;transform:rotate(150deg) translate(146%)}.sonner-loading-bar:nth-child(7){animation-delay:-.6s;transform:rotate(180deg) translate(146%)}.sonner-loading-bar:nth-child(8){animation-delay:-.5s;transform:rotate(210deg) translate(146%)}.sonner-loading-bar:nth-child(9){animation-delay:-.4s;transform:rotate(240deg) translate(146%)}.sonner-loading-bar:nth-child(10){animation-delay:-.3s;transform:rotate(270deg) translate(146%)}.sonner-loading-bar:nth-child(11){animation-delay:-.2s;transform:rotate(300deg) translate(146%)}.sonner-loading-bar:nth-child(12){animation-delay:-.1s;transform:rotate(330deg) translate(146%)}@keyframes sonner-fade-in{0%{opacity:0;transform:scale(.8)}to{opacity:1;transform:scale(1)}}@keyframes sonner-fade-out{0%{opacity:1;transform:scale(1)}to{opacity:0;transform:scale(.8)}}@keyframes sonner-spin{0%{opacity:1}to{opacity:.15}}@media (prefers-reduced-motion){[data-sonner-toast],[data-sonner-toast]>*,.sonner-loading-bar{transition:none!important;animation:none!important}}.sonner-loader{position:absolute;top:50%;left:50%;transform:translate(-50%,-50%);transform-origin:center;transition:opacity .2s,transform .2s}.sonner-loader[data-visible=false]{opacity:0;transform:scale(.8) translate(-50%,-50%)}\n"], dependencies: [{ kind: "component", type: BrnSonnerToast, selector: "brn-sonner-toast", inputs: ["toast", "index", "expanded", "invert", "position", "visibleToasts", "expandByDefault", "closeButton", "interacting", "cancelButtonStyle", "actionButtonStyle", "duration", "descriptionClass", "classes", "unstyled", "class", "style"] }, { kind: "component", type: BrnSonnerIcon, selector: "brn-sonner-icon", inputs: ["type"] }, { kind: "component", type: BrnSonnerLoader, selector: "brn-sonner-loader", inputs: ["isVisible"] }, { kind: "directive", type: NgTemplateOutlet, selector: "[ngTemplateOutlet]", inputs: ["ngTemplateOutletContext", "ngTemplateOutlet", "ngTemplateOutletInjector"] }, { kind: "pipe", type: ToastFilterPipe, name: "toastFilter" }], changeDetection: i0.ChangeDetectionStrategy.OnPush, encapsulation: i0.ViewEncapsulation.None });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSonnerToaster, decorators: [{
            type: Component,
            args: [{ selector: 'brn-sonner-toaster', imports: [BrnSonnerToast, ToastFilterPipe, BrnSonnerIcon, BrnSonnerLoader, NgTemplateOutlet], encapsulation: ViewEncapsulation.None, changeDetection: ChangeDetectionStrategy.OnPush, template: `
		@if (_toasts().length > 0) {
			<section [attr.aria-label]="'Notifications ' + _hotKeyLabel()" [tabIndex]="-1">
				@for (pos of _possiblePositions(); track pos) {
					<ol
						#listRef
						[tabIndex]="-1"
						[class]="userClass()"
						data-sonner-toaster
						[attr.data-theme]="_actualTheme()"
						[attr.data-rich-colors]="richColors()"
						[attr.dir]="direction()"
						[attr.data-y-position]="pos.split('-')[0]"
						[attr.data-x-position]="pos.split('-')[1]"
						[style]="_toasterStyles()"
						(blur)="handleBlur($event)"
						(focus)="handleFocus($event)"
						(mouseenter)="_expanded.set(true)"
						(mousemove)="_expanded.set(true)"
						(mouseleave)="handleMouseLeave()"
						(pointerdown)="handlePointerDown($event)"
						(pointerup)="_interacting.set(false)"
					>
						@for (toast of _toasts() | toastFilter: $index : pos; track toast.id) {
							<brn-sonner-toast
								[index]="$index"
								[toast]="toast"
								[invert]="invert()"
								[visibleToasts]="visibleToasts()"
								[closeButton]="closeButton()"
								[interacting]="_interacting()"
								[position]="pos"
								[expandByDefault]="expand()"
								[expanded]="_expanded()"
								[actionButtonStyle]="toastOptions().actionButtonStyle"
								[cancelButtonStyle]="toastOptions().cancelButtonStyle"
								[class]="toastOptions().class ?? ''"
								[descriptionClass]="toastOptions().descriptionClass ?? ''"
								[classes]="toastOptions().classes ?? {}"
								[duration]="toastOptions().duration ?? duration()"
								[unstyled]="toastOptions().unstyled ?? false"
							>
								@if (_loadingIcon(); as loadingIcon) {
									<ng-container *ngTemplateOutlet="loadingIcon" loading-icon />
								} @else {
									<brn-sonner-loader [isVisible]="toast.type === 'loading'" loading-icon />
								}

								@if (_successIcon(); as successIcon) {
									<ng-container *ngTemplateOutlet="successIcon" success-icon />
								} @else {
									<brn-sonner-icon type="success" success-icon />
								}

								@if (_errorIcon(); as errorIcon) {
									<ng-container *ngTemplateOutlet="errorIcon" error-icon />
								} @else {
									<brn-sonner-icon type="error" error-icon />
								}

								@if (_warningIcon(); as warningIcon) {
									<ng-container *ngTemplateOutlet="warningIcon" warning-icon />
								} @else {
									<brn-sonner-icon type="warning" warning-icon />
								}

								@if (_infoIcon(); as infoIcon) {
									<ng-container *ngTemplateOutlet="infoIcon" info-icon />
								} @else {
									<brn-sonner-icon type="info" info-icon />
								}
							</brn-sonner-toast>
						}
					</ol>
				}
			</section>
		}
	`, styles: ["html[dir=ltr],[data-sonner-toaster][dir=ltr]{--toast-icon-margin-start: var(--brn-sonner-toast-icon-margin-start, -3px);--toast-icon-margin-end: var(--brn-sonner-toast-icon-margin-end, 4px);--toast-svg-margin-start: var(--brn-sonner-toast-svg-margin-start, -1px);--toast-svg-margin-end: var(--brn-sonner-toast-svg-margin-end, 0px);--toast-button-margin-start: var(--brn-sonner-toast-button-margin-start, auto);--toast-button-margin-end: var(--brn-sonner-toast-button-margin-end, 0);--toast-close-button-start: var(--brn-sonner-toast-close-button-start, 0);--toast-close-button-end: var(--brn-sonner-toast-close-button-end, unset);--toast-close-button-transform: var(--brn-sonner-toast-close-button-transform, translate(-35%, -35%))}html[dir=rtl],[data-sonner-toaster][dir=rtl]{--toast-icon-margin-start: var(--brn-sonner-rtl-toast-icon-margin-start, 4px);--toast-icon-margin-end: var(--brn-sonner-rtl-toast-icon-margin-end, -3px);--toast-svg-margin-start: var(--brn-sonner-rtl-toast-svg-margin-start, 0px);--toast-svg-margin-end: var(--brn-sonner-rtl-toast-svg-margin-end, -1px);--toast-button-margin-start: var(--brn-sonner-rtl-toast-button-margin-start, 0);--toast-button-margin-end: var(--brn-sonner-rtl-toast-button-margin-end, auto);--toast-close-button-start: var(--brn-sonner-rtl-toast-close-button-start, unset);--toast-close-button-end: var(--brn-sonner-rtl-toast-close-button-end, 0);--toast-close-button-transform: var(--brn-sonner-rtl-toast-close-button-transform, translate(35%, -35%))}[data-sonner-toaster]{position:fixed;width:var(--width);font-family:var( --brn-sonner-font-family, ui-sans-serif, system-ui, -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Helvetica Neue, Arial, Noto Sans, sans-serif, Apple Color Emoji, Segoe UI Emoji, Segoe UI Symbol, Noto Color Emoji );--gray1: hsl(0, 0%, 99%);--gray2: hsl(0, 0%, 97.3%);--gray3: hsl(0, 0%, 95.1%);--gray4: hsl(0, 0%, 93%);--gray5: hsl(0, 0%, 90.9%);--gray6: hsl(0, 0%, 88.7%);--gray7: hsl(0, 0%, 85.8%);--gray8: hsl(0, 0%, 78%);--gray9: hsl(0, 0%, 56.1%);--gray10: hsl(0, 0%, 52.3%);--gray11: hsl(0, 0%, 43.5%);--gray12: hsl(0, 0%, 9%);--border-radius: var(--brn-sonner-border-radius, 8px);box-sizing:border-box;padding:0;margin:0;list-style:none;outline:none;z-index:999999999}[data-sonner-toaster][data-x-position=right]{right:max(var(--offset),env(safe-area-inset-right))}[data-sonner-toaster][data-x-position=left]{left:max(var(--offset),env(safe-area-inset-left))}[data-sonner-toaster][data-x-position=center]{left:50%;transform:translate(-50%)}[data-sonner-toaster][data-y-position=top]{top:max(var(--offset),env(safe-area-inset-top))}[data-sonner-toaster][data-y-position=bottom]{bottom:max(var(--offset),env(safe-area-inset-bottom))}[data-sonner-toast]{--y: translateY(100%);--lift-amount: calc(var(--lift) * var(--gap));z-index:var(--z-index);position:absolute;opacity:0;transform:var(--y);filter:blur(0);touch-action:none;transition:transform .4s,opacity .4s,height .4s,box-shadow .2s;box-sizing:border-box;outline:none;overflow-wrap:anywhere}[data-sonner-toast][data-styled=true]{padding:16px;background:var(--normal-bg);border:1px solid var(--normal-border);color:var(--normal-text);border-radius:var(--border-radius);box-shadow:0 4px 12px #0000001a;width:var(--width);font-size:13px;display:flex;align-items:center;gap:6px}[data-sonner-toast]:focus-visible{box-shadow:0 4px 12px #0000001a,0 0 0 2px #0003}[data-sonner-toast][data-y-position=top]{top:0;--y: translateY(-100%);--lift: 1;--lift-amount: calc(1 * var(--gap))}[data-sonner-toast][data-y-position=bottom]{bottom:0;--y: translateY(100%);--lift: -1;--lift-amount: calc(var(--lift) * var(--gap))}[data-sonner-toast] [data-description]{font-weight:400;line-height:1.4;color:inherit}[data-sonner-toast] [data-title]{font-weight:500;line-height:1.5;color:inherit}[data-sonner-toast] [data-icon]{display:flex;height:16px;width:16px;position:relative;justify-content:flex-start;align-items:center;flex-shrink:0;margin-left:var(--toast-icon-margin-start);margin-right:var(--toast-icon-margin-end)}[data-sonner-toast][data-promise=true] [data-icon]>svg{opacity:0;transform:scale(.8);transform-origin:center;animation:sonner-fade-in .3s ease forwards}[data-sonner-toast] [data-icon]>*{flex-shrink:0}[data-sonner-toast] [data-icon] svg{margin-left:var(--toast-svg-margin-start);margin-right:var(--toast-svg-margin-end)}[data-sonner-toast] [data-content]{display:flex;flex-direction:column;gap:2px}[data-sonner-toast] [data-button]{border-radius:4px;padding-left:8px;padding-right:8px;height:24px;font-size:12px;color:var(--normal-bg);background:var(--normal-text);margin-left:var(--toast-button-margin-start);margin-right:var(--toast-button-margin-end);border:none;cursor:pointer;outline:none;display:flex;align-items:center;flex-shrink:0;transition:opacity .4s,box-shadow .2s}[data-sonner-toast] [data-button]:focus-visible{box-shadow:var(--brn-sonner-toast-focus-box-shadow, 0 0 0 2px rgba(0, 0, 0, .4))}[data-sonner-toast] [data-button]:first-of-type{margin-left:var(--toast-button-margin-start);margin-right:var(--toast-button-margin-end)}[data-sonner-toast] [data-cancel]{color:var(--normal-text);background:#00000014}[data-sonner-toast][data-theme=dark] [data-cancel]{background:#ffffff4d}[data-sonner-toast] [data-close-button]{position:absolute;left:var(--toast-close-button-start);right:var(--toast-close-button-end);top:0;height:20px;width:20px;display:flex;justify-content:center;align-items:center;padding:0;background:var(--brn-sonner-toast-close-button-background, var(--gray1));color:var(--brn-sonner-toast-close-button-color, var(--gray12));border:var(--brn-sonner-toast-close-button-border, 1px solid var(--gray4));transform:var(--toast-close-button-transform);border-radius:50%;cursor:pointer;z-index:1;transition:opacity .1s,background .2s,border-color .2s}[data-sonner-toast] [data-close-button]:focus-visible{box-shadow:0 4px 12px #0000001a,0 0 0 2px #0003}[data-sonner-toast] [data-disabled=true]{cursor:not-allowed}[data-sonner-toast]:hover [data-close-button]:hover{background:var(--brn-sonner-toast-close-button-hover-background, var(--gray2));color:var(--brn-sonner-toast-close-button-hover-color, var(--gray12));border-color:var(--brn-sonner-toast-close-button-hover-border-color, var(--gray5))}[data-sonner-toast][data-swiping=true]:before{content:\"\";position:absolute;left:0;right:0;height:100%;z-index:-1}[data-sonner-toast][data-y-position=top][data-swiping=true]:before{bottom:50%;transform:scaleY(3) translateY(50%)}[data-sonner-toast][data-y-position=bottom][data-swiping=true]:before{top:50%;transform:scaleY(3) translateY(-50%)}[data-sonner-toast][data-swiping=false][data-removed=true]:before{content:\"\";position:absolute;inset:0;transform:scaleY(2)}[data-sonner-toast]:after{content:\"\";position:absolute;left:0;height:calc(var(--gap) + 1px);bottom:100%;width:100%}[data-sonner-toast][data-mounted=true]{--y: translateY(0);opacity:1}[data-sonner-toast][data-expanded=false][data-front=false]{--scale: var(--toasts-before) * .05 + 1;--y: translateY(calc(var(--lift-amount) * var(--toasts-before))) scale(calc(-1 * var(--scale)));height:var(--front-toast-height)}[data-sonner-toast]>*{transition:opacity .4s}[data-sonner-toast][data-expanded=false][data-front=false][data-styled=true]>*{opacity:0}[data-sonner-toast][data-visible=false]{opacity:0;pointer-events:none}[data-sonner-toast][data-mounted=true][data-expanded=true]{--y: translateY(calc(var(--lift) * var(--offset)));height:var(--initial-height)}[data-sonner-toast][data-removed=true][data-front=true][data-swipe-out=false]{--y: translateY(calc(var(--lift) * -100%));opacity:0}[data-sonner-toast][data-removed=true][data-front=false][data-swipe-out=false][data-expanded=true]{--y: translateY(calc(var(--lift) * var(--offset) + var(--lift) * -100%));opacity:0}[data-sonner-toast][data-removed=true][data-front=false][data-swipe-out=false][data-expanded=false]{--y: translateY(40%);opacity:0;transition:transform .5s,opacity .2s}[data-sonner-toast][data-removed=true][data-front=false]:before{height:calc(var(--initial-height) + 20%)}[data-sonner-toast][data-swiping=true]{transform:var(--y) translateY(var(--swipe-amount, 0px));transition:none}[data-sonner-toast][data-swipe-out=true][data-y-position=bottom],[data-sonner-toast][data-swipe-out=true][data-y-position=top]{animation:swipe-out .2s ease-out forwards}@keyframes swipe-out{0%{transform:translateY(calc(var(--lift) * var(--offset) + var(--swipe-amount)));opacity:1}to{transform:translateY(calc(var(--lift) * var(--offset) + var(--swipe-amount) + var(--lift) * -100%));opacity:0}}@media (max-width: 600px){[data-sonner-toaster]{position:fixed;--mobile-offset: 16px;right:var(--mobile-offset);left:var(--mobile-offset);width:100%}[data-sonner-toaster] [data-sonner-toast]{left:0;right:0;width:calc(100% - 32px)}[data-sonner-toaster][data-x-position=left]{left:var(--mobile-offset)}[data-sonner-toaster][data-y-position=bottom]{bottom:20px}[data-sonner-toaster][data-y-position=top]{top:20px}[data-sonner-toaster][data-x-position=center]{left:var(--mobile-offset);right:var(--mobile-offset);transform:none}}[data-sonner-toaster][data-theme=light]{--normal-bg: var(--brn-sonner-toast-normal-background, #fff);--normal-border: var(--brn-sonner-toast-normal-border-color, var(--gray4));--normal-text: var(--brn-sonner-toast-normal-color, var(--gray12));--success-bg: var(--brn-sonner-toast-success-background, hsl(143, 85%, 96%));--success-border: var(--brn-sonner-toast-success-border, hsl(145, 92%, 91%));--success-text: var(--brn-sonner-toast-success-color, hsl(140, 100%, 27%));--info-bg: var(--brn-sonner-toast-info-background, hsl(208, 100%, 97%));--info-border: var(--brn-sonner-toast-info-border, hsl(221, 91%, 91%));--info-text: var(--brn-sonner-toast-info-color, hsl(210, 92%, 45%));--warning-bg: var(--brn-sonner-toast-warning-background, hsl(49, 100%, 97%));--warning-border: var(--brn-sonner-toast-warning-border, hsl(49, 91%, 91%));--warning-text: var(--brn-sonner-toast-warning-color, hsl(31, 92%, 45%));--error-bg: var(--brn-sonner-toast-error-background, hsl(359, 100%, 97%));--error-border: var(--brn-sonner-toast-error-border, hsl(359, 100%, 94%));--error-text: var(--brn-sonner-toast-error-color, hsl(360, 100%, 45%))}[data-sonner-toaster][data-theme=light] [data-sonner-toast][data-invert=true]{--normal-bg: var(--brn-sonner-toast-inverse-normal-background, #000);--normal-border: var(--brn-sonner-toast-inverse-normal-border-color, hsl(0, 0%, 20%));--normal-text: var(--brn-sonner-toast-inverse-normal-color, var(--gray1))}[data-sonner-toaster][data-theme=dark] [data-sonner-toast][data-invert=true]{--normal-bg: var(--brn-sonner-toast-inverse-dark-normal-background, #fff);--normal-border: var(--brn-sonner-toast-inverse-dark-normal-border-color, var(--gray3));--normal-text: var(--brn-sonner-toast-inverse-dark-normal-color, var(--gray12))}[data-sonner-toaster][data-theme=dark]{--normal-bg: var(--brn-sonner-toast-dark-normal-background, #000);--normal-border: var(--brn-sonner-toast-dark-normal-border-color, hsl(0, 0%, 20%));--normal-text: var(--brn-sonner-toast-dark-normal-color, var(--gray1));--success-bg: var(--brn-sonner-toast-dark-success-background, hsl(150, 100%, 6%));--success-border: var(--brn-sonner-toast-dark-success-border, hsl(147, 100%, 12%));--success-text: var(--brn-sonner-toast-dark-success-color, hsl(150, 86%, 65%));--info-bg: var(--brn-sonner-toast-dark-info-background, hsl(215, 100%, 6%));--info-border: var(--brn-sonner-toast-dark-info-border, hsl(223, 100%, 12%));--info-text: var(--brn-sonner-toast-dark-info-color, hsl(216, 87%, 65%));--warning-bg: var(--brn-sonner-toast-dark-warning-background, hsl(64, 100%, 6%));--warning-border: var(--brn-sonner-toast-dark-warning-border, hsl(60, 100%, 12%));--warning-text: var(--brn-sonner-toast-dark-warning-color, hsl(46, 87%, 65%));--error-bg: var(--brn-sonner-toast-dark-error-background, hsl(358, 76%, 10%));--error-border: var(--brn-sonner-toast-dark-error-border, hsl(357, 89%, 16%));--error-text: var(--brn-sonner-toast-dark-error-color, hsl(358, 100%, 81%))}[data-rich-colors=true] [data-sonner-toast][data-type=success],[data-rich-colors=true] [data-sonner-toast][data-type=success] [data-close-button]{background:var(--success-bg);border-color:var(--success-border);color:var(--success-text)}[data-rich-colors=true] [data-sonner-toast][data-type=info],[data-rich-colors=true] [data-sonner-toast][data-type=info] [data-close-button]{background:var(--info-bg);border-color:var(--info-border);color:var(--info-text)}[data-rich-colors=true] [data-sonner-toast][data-type=warning],[data-rich-colors=true] [data-sonner-toast][data-type=warning] [data-close-button]{background:var(--warning-bg);border-color:var(--warning-border);color:var(--warning-text)}[data-rich-colors=true] [data-sonner-toast][data-type=error],[data-rich-colors=true] [data-sonner-toast][data-type=error] [data-close-button]{background:var(--error-bg);border-color:var(--error-border);color:var(--error-text)}.sonner-loading-wrapper{--size: 16px;height:var(--size);width:var(--size);position:absolute;inset:0;z-index:10}.sonner-loading-wrapper[data-visible=false]{transform-origin:center;animation:sonner-fade-out .2s ease forwards}.sonner-spinner{position:relative;top:50%;left:50%;height:var(--size);width:var(--size)}.sonner-loading-bar{animation:sonner-spin 1.2s linear infinite;background:var(--gray11);border-radius:6px;height:8%;left:-10%;position:absolute;top:-3.9%;width:24%}.sonner-loading-bar:nth-child(1){animation-delay:-1.2s;transform:rotate(.0001deg) translate(146%)}.sonner-loading-bar:nth-child(2){animation-delay:-1.1s;transform:rotate(30deg) translate(146%)}.sonner-loading-bar:nth-child(3){animation-delay:-1s;transform:rotate(60deg) translate(146%)}.sonner-loading-bar:nth-child(4){animation-delay:-.9s;transform:rotate(90deg) translate(146%)}.sonner-loading-bar:nth-child(5){animation-delay:-.8s;transform:rotate(120deg) translate(146%)}.sonner-loading-bar:nth-child(6){animation-delay:-.7s;transform:rotate(150deg) translate(146%)}.sonner-loading-bar:nth-child(7){animation-delay:-.6s;transform:rotate(180deg) translate(146%)}.sonner-loading-bar:nth-child(8){animation-delay:-.5s;transform:rotate(210deg) translate(146%)}.sonner-loading-bar:nth-child(9){animation-delay:-.4s;transform:rotate(240deg) translate(146%)}.sonner-loading-bar:nth-child(10){animation-delay:-.3s;transform:rotate(270deg) translate(146%)}.sonner-loading-bar:nth-child(11){animation-delay:-.2s;transform:rotate(300deg) translate(146%)}.sonner-loading-bar:nth-child(12){animation-delay:-.1s;transform:rotate(330deg) translate(146%)}@keyframes sonner-fade-in{0%{opacity:0;transform:scale(.8)}to{opacity:1;transform:scale(1)}}@keyframes sonner-fade-out{0%{opacity:1;transform:scale(1)}to{opacity:0;transform:scale(.8)}}@keyframes sonner-spin{0%{opacity:1}to{opacity:.15}}@media (prefers-reduced-motion){[data-sonner-toast],[data-sonner-toast]>*,.sonner-loading-bar{transition:none!important;animation:none!important}}.sonner-loader{position:absolute;top:50%;left:50%;transform:translate(-50%,-50%);transform-origin:center;transition:opacity .2s,transform .2s}.sonner-loader[data-visible=false]{opacity:0;transform:scale(.8) translate(-50%,-50%)}\n"] }]
        }], ctorParameters: () => [], propDecorators: { invert: [{ type: i0.Input, args: [{ isSignal: true, alias: "invert", required: false }] }], theme: [{ type: i0.Input, args: [{ isSignal: true, alias: "theme", required: false }] }], position: [{ type: i0.Input, args: [{ isSignal: true, alias: "position", required: false }] }], hotKey: [{ type: i0.Input, args: [{ isSignal: true, alias: "hotKey", required: false }] }], richColors: [{ type: i0.Input, args: [{ isSignal: true, alias: "richColors", required: false }] }], expand: [{ type: i0.Input, args: [{ isSignal: true, alias: "expand", required: false }] }], duration: [{ type: i0.Input, args: [{ isSignal: true, alias: "duration", required: false }] }], visibleToasts: [{ type: i0.Input, args: [{ isSignal: true, alias: "visibleToasts", required: false }] }], closeButton: [{ type: i0.Input, args: [{ isSignal: true, alias: "closeButton", required: false }] }], toastOptions: [{ type: i0.Input, args: [{ isSignal: true, alias: "toastOptions", required: false }] }], offset: [{ type: i0.Input, args: [{ isSignal: true, alias: "offset", required: false }] }], userClass: [{ type: i0.Input, args: [{ isSignal: true, alias: "class", required: false }] }], style: [{ type: i0.Input, args: [{ isSignal: true, alias: "style", required: false }] }], _listRef: [{ type: i0.ViewChild, args: ['listRef', { isSignal: true }] }], _loadingIcon: [{ type: i0.ContentChild, args: ['loadingIcon', { ...{ read: TemplateRef }, isSignal: true }] }], _successIcon: [{ type: i0.ContentChild, args: ['successIcon', { ...{ read: TemplateRef }, isSignal: true }] }], _errorIcon: [{ type: i0.ContentChild, args: ['errorIcon', { ...{ read: TemplateRef }, isSignal: true }] }], _warningIcon: [{ type: i0.ContentChild, args: ['warningIcon', { ...{ read: TemplateRef }, isSignal: true }] }], _infoIcon: [{ type: i0.ContentChild, args: ['infoIcon', { ...{ read: TemplateRef }, isSignal: true }] }] } });

const BrnSonnerImports = [BrnSonnerToaster];

/**
 * Generated bundle index. Do not edit.
 */

export { BrnSonnerImports, BrnSonnerToaster, defaultClasses, injectBrnSonnerToasterConfig, provideBrnSonnerToasterConfig, toast, toastState };
