// @spartan-ng/brain/fesm2022/spartan-ng-brain-core.mjs
import { untracked, computed, InjectionToken, forwardRef, inject, runInInjectionContext, Injector, ElementRef, PLATFORM_ID, DestroyRef, signal, afterNextRender } from '@angular/core';
import { Observable, pipe, merge, fromEvent } from 'rxjs';
import { map, filter, distinctUntilChanged, takeUntil, debounceTime } from 'rxjs/operators';
import { toObservable, toSignal } from '@angular/core/rxjs-interop';
import { isPlatformServer } from '@angular/common';

/**
 * Returns a signal that emits the previous value of the given signal.
 * The first time the signal is emitted, the previous value will be the same as the current value.
 *
 * @example
 * ```ts
 * const value = signal(0);
 * const previous = computedPrevious(value);
 *
 * effect(() => {
 *  console.log('Current value:', value());
 *  console.log('Previous value:', previous());
 * });
 *
 * Logs:
 * // Current value: 0
 * // Previous value: 0
 *
 * value.set(1);
 *
 * Logs:
 * // Current value: 1
 * // Previous value: 0
 *
 * value.set(2);
 *
 * Logs:
 * // Current value: 2
 * // Previous value: 1
 *```
 *
 * @param computation Signal to compute previous value for
 * @returns Signal that emits previous value of `s`
 */
function computedPrevious(computation) {
    let current = null;
    let previous = untracked(() => computation()); // initial value is the current value
    return computed(() => {
        current = computation();
        const result = previous;
        previous = current;
        return result;
    });
}

function brnZoneFull(zone) {
    return (source) => new Observable((subscriber) => source.subscribe({
        next: (value) => zone.run(() => subscriber.next(value)),
        error: (error) => zone.run(() => subscriber.error(error)),
        complete: () => zone.run(() => subscriber.complete()),
    }));
}
function brnZoneFree(zone) {
    return (source) => new Observable((subscriber) => zone.runOutsideAngular(() => source.subscribe(subscriber)));
}
function brnZoneOptimized(zone) {
    return pipe(brnZoneFree(zone), brnZoneFull(zone));
}

function movedOut({ currentTarget, relatedTarget }) {
    return !isElement(relatedTarget) || !isElement(currentTarget) || !currentTarget.contains(relatedTarget);
}
function isElement(node) {
    return !!node && 'nodeType' in node && node.nodeType === Node.ELEMENT_NODE;
}
const createHoverObservable = (nativeElement, zone, destroyed$) => {
    return merge(fromEvent(nativeElement, 'mouseenter').pipe(map(() => ({ hover: true }))), fromEvent(nativeElement, 'mouseleave').pipe(map((e) => ({ hover: false, relatedTarget: e.relatedTarget }))), 
    // Hello, Safari
    fromEvent(nativeElement, 'mouseout').pipe(filter(movedOut), map((e) => ({ hover: false, relatedTarget: e.relatedTarget })))).pipe(distinctUntilChanged(), brnZoneOptimized(zone), takeUntil(destroyed$));
};

function createInjectionToken(description) {
    const token = new InjectionToken(description);
    const provideFn = (value) => {
        return { provide: token, useValue: value };
    };
    const provideExistingFn = (value) => {
        return { provide: token, useExisting: forwardRef(value) };
    };
    const injectFn = (options = {}) => {
        return inject(token, options);
    };
    return [injectFn, provideFn, provideExistingFn, token];
}

const [injectCustomClassSettable, provideCustomClassSettable, provideCustomClassSettableExisting, SET_CLASS_TO_CUSTOM_ELEMENT_TOKEN,] = createInjectionToken('@spartan-ng SET_CLASS_TO_CUSTOM_ELEMENT_TOKEN');

/**
 * Creates a debounced version of a source signal.
 *
 * @param source - The input signal to debounce.
 * @param delay - Debounce time in milliseconds.
 * @returns A new signal that updates only after the source has stopped changing for `delay` ms.
 */
function debouncedSignal(source, delay) {
    const source$ = toObservable(source);
    const debounced$ = source$.pipe(debounceTime(delay), distinctUntilChanged());
    return toSignal(debounced$, { initialValue: source() });
}

/**
 * Set by Angular to true when in development mode.
 * Allows for tree-shaking code that is only used in development.
 */
const brnDevMode = ngDevMode;

const [injectExposedSideProvider, provideExposedSideProvider, provideExposedSideProviderExisting, EXPOSES_SIDE_TOKEN,] = createInjectionToken('@spartan-ng EXPOSES_SIDE_TOKEN');

const [injectExposesStateProvider, provideExposesStateProvider, provideExposesStateProviderExisting, EXPOSES_STATE_TOKEN,] = createInjectionToken('@spartan-ng EXPOSES_STATE_TOKEN');

/**
 * Returns a reactive signal that tracks the size of a DOM element.
 *
 * This function uses a shared {@link ResizeObserver} internally to monitor multiple elements efficiently.
 * The returned signal updates whenever the element's width or height changes. It is fully SSR-safe.
 *
 * @param options Optional {@link ElementSizeOptions} configuration object.
 * @param options.elementRef The {@link ElementRef} of the element to observe. Defaults to the {@link ElementRef} injected in the current context.
 * @param options.injector Optional custom {@link Injector} to run the function in. Defaults to the current injection context.
 *
 * @returns A readonly {@link Signal} containing `{ width, height }` of the element, or `undefined` if the element has not yet been measured or on the server.
 *
 * Based on https://github.com/radix-ui/primitives/blob/main/packages/react/use-size/src/use-size.tsx
 */
function injectElementSize(options = {}) {
    return runInInjectionContext(options.injector ?? inject(Injector), () => {
        const elementRef = options.elementRef ?? inject(ElementRef);
        const platformId = inject(PLATFORM_ID);
        const destroyRef = inject(DestroyRef);
        const element = elementRef.nativeElement;
        const size = signal(undefined, ...(ngDevMode ? [{ debugName: "size" }] : []));
        if (isPlatformServer(platformId)) {
            return size;
        }
        afterNextRender({
            read: () => {
                // Initial read
                size.set({
                    width: element.offsetWidth,
                    height: element.offsetHeight,
                });
                // Register element in shared observer
                observerMap.set(element, { element, size });
                getSharedObserver().observe(element, { box: 'border-box' });
                destroyRef.onDestroy(() => {
                    getSharedObserver().unobserve(element);
                    observerMap.delete(element);
                    // Disconnect global observer if no elements left
                    if (observerMap.size === 0 && sharedObserver) {
                        sharedObserver.disconnect();
                        sharedObserver = undefined;
                    }
                });
            },
        });
        return size.asReadonly();
    });
}
const observerMap = new Map();
let sharedObserver;
function getSharedObserver() {
    if (!sharedObserver) {
        sharedObserver = new ResizeObserver((entries) => {
            for (const entry of entries) {
                const el = entry.target;
                const entryData = observerMap.get(el);
                if (!entryData)
                    continue;
                let width;
                let height;
                if ('borderBoxSize' in entry) {
                    // Iron out differences between browsers
                    const borderSize = Array.isArray(entry.borderBoxSize) ? entry.borderBoxSize[0] : entry.borderBoxSize;
                    width = borderSize.inlineSize;
                    height = borderSize.blockSize;
                }
                else {
                    width = el.offsetWidth;
                    height = el.offsetHeight;
                }
                entryData.size.set({ width, height });
            }
        });
    }
    return sharedObserver;
}

const measureDimensions = (elementToMeasure, measurementDisplay) => {
    const previousHeight = elementToMeasure.style.height;
    const previousDisplay = elementToMeasure.style.display;
    const previousHidden = elementToMeasure.hidden;
    elementToMeasure.hidden = false;
    elementToMeasure.style.height = 'auto';
    elementToMeasure.style.display =
        !previousDisplay || previousDisplay === 'none' ? measurementDisplay : previousDisplay;
    const { width, height } = elementToMeasure.getBoundingClientRect();
    elementToMeasure.hidden = previousHidden;
    elementToMeasure.style.display = previousDisplay;
    elementToMeasure.style.height = previousHeight;
    return { width, height };
};

const createMenuPosition = (align, side) => {
    const verticalAlign = align === 'start' ? 'top' : align === 'end' ? 'bottom' : 'center';
    const createPositions = (originX, originY, overlayX, overlayY) => [
        { originX, originY, overlayX, overlayY },
        { originX: overlayX, originY: overlayY, overlayX: originX, overlayY: originY },
    ];
    switch (side) {
        case 'top':
            return createPositions(align, 'top', align, 'bottom');
        case 'bottom':
            return createPositions(align, 'bottom', align, 'top');
        case 'left':
            return createPositions('start', verticalAlign, 'end', verticalAlign);
        case 'right':
            return createPositions('end', verticalAlign, 'start', verticalAlign);
    }
};

function stringifyAsLabel(item, itemToStringLabel) {
    if (itemToStringLabel && item != null) {
        return itemToStringLabel(item) ?? '';
    }
    if (item && typeof item === 'object') {
        if ('label' in item && item.label != null) {
            return String(item.label);
        }
        if ('value' in item) {
            return String(item.value);
        }
    }
    return serializeValue(item);
}
function serializeValue(value) {
    if (value == null) {
        return '';
    }
    if (typeof value === 'string') {
        return value;
    }
    try {
        return JSON.stringify(value);
    }
    catch {
        return String(value);
    }
}

const [injectTableClassesSettable, provideTableClassesSettable, provideTableClassesSettableExisting, SET_TABLE_CLASSES_TOKEN,] = createInjectionToken('@spartan-ng SET_TABLE_CLASSES_TOKEN');

/**
 * Waits for all animations (including subtree) within the given element to finish.
 * Ignores animations canceled with an AbortError.
 */
async function waitForElementAnimations(el) {
    // Wait a tick to allow newly triggered animations to start
    await new Promise((resolve) => setTimeout(resolve, 0));
    const animationFillMode = el.style.animationFillMode;
    const animations = el.getAnimations({ subtree: true });
    el.style.animationFillMode = 'forwards';
    await Promise.all(animations.map((animation) => animation.finished.catch((err) => {
        // Ignore AbortError from canceled animations (treated as "finished")
        if (!(err instanceof Error && err.name === 'AbortError')) {
            throw err;
        }
        return animation;
    })));
    setTimeout(() => {
        if (el.style.animationFillMode === 'forwards')
            el.style.animationFillMode = animationFillMode;
    });
}

/**
 * Generated bundle index. Do not edit.
 */

export { EXPOSES_SIDE_TOKEN, EXPOSES_STATE_TOKEN, SET_CLASS_TO_CUSTOM_ELEMENT_TOKEN, SET_TABLE_CLASSES_TOKEN, brnDevMode, brnZoneFree, brnZoneFull, brnZoneOptimized, computedPrevious, createHoverObservable, createMenuPosition, debouncedSignal, injectCustomClassSettable, injectElementSize, injectExposedSideProvider, injectExposesStateProvider, injectTableClassesSettable, isElement, measureDimensions, provideCustomClassSettable, provideCustomClassSettableExisting, provideExposedSideProvider, provideExposedSideProviderExisting, provideExposesStateProvider, provideExposesStateProviderExisting, provideTableClassesSettable, provideTableClassesSettableExisting, serializeValue, stringifyAsLabel, waitForElementAnimations };
