// @spartan-ng/brain/fesm2022/spartan-ng-brain-switch.mjs
import { FocusMonitor } from '@angular/cdk/a11y';
import { isPlatformBrowser } from '@angular/common';
import * as i0 from '@angular/core';
import { forwardRef, inject, DestroyRef, Renderer2, PLATFORM_ID, ElementRef, ChangeDetectorRef, DOCUMENT, signal, model, output, input, linkedSignal, booleanAttribute, numberAttribute, viewChild, computed, afterRenderEffect, ChangeDetectionStrategy, Component } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import * as i1 from '@spartan-ng/brain/field';
import { BrnFieldControl, provideBrnLabelable } from '@spartan-ng/brain/field';

const BRN_SWITCH_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(() => BrnSwitch),
    multi: true,
};
const CONTAINER_POST_FIX = '-switch';
let uniqueIdCounter = 0;
class BrnSwitch {
    _destroyRef = inject(DestroyRef);
    _renderer = inject(Renderer2);
    _isBrowser = isPlatformBrowser(inject(PLATFORM_ID));
    _elementRef = inject(ElementRef);
    _focusMonitor = inject(FocusMonitor);
    _cdr = inject(ChangeDetectorRef);
    _document = inject(DOCUMENT);
    _fieldControl = inject(BrnFieldControl, { optional: true });
    _focusVisible = signal(false, ...(ngDevMode ? [{ debugName: "_focusVisible" }] : []));
    _focused = signal(false, ...(ngDevMode ? [{ debugName: "_focused" }] : []));
    /**
     * Whether switch is checked/toggled on.
     * Can be bound with [(checked)] for two-way binding.
     */
    checked = model(false, ...(ngDevMode ? [{ debugName: "checked" }] : []));
    /** Emits when checked state changes. */
    checkedChange = output();
    /**
     * Unique identifier for switch component.
     * When provided, inner button gets ID without '-switch' suffix.
     * Auto-generates ID if not provided.
     */
    id = input(++uniqueIdCounter + '', ...(ngDevMode ? [{ debugName: "id" }] : []));
    /**
     * Form control name for switch.
     * When provided, inner button gets name without '-switch' suffix.
     */
    name = input(null, ...(ngDevMode ? [{ debugName: "name" }] : []));
    /**
     * CSS classes applied to inner button element.
     */
    class = input(null, ...(ngDevMode ? [{ debugName: "class" }] : []));
    /**
     * Accessibility label for screen readers.
     * Use when no visible label exists.
     */
    ariaLabel = input(null, ...(ngDevMode ? [{ debugName: "ariaLabel", alias: 'aria-label' }] : [{ alias: 'aria-label' }]));
    /**
     * ID of element that labels this switch for accessibility.
     * Auto-set when switch is inside label element.
     */
    ariaLabelledby = input(null, ...(ngDevMode ? [{ debugName: "ariaLabelledby", alias: 'aria-labelledby' }] : [{ alias: 'aria-labelledby' }]));
    mutableAriaLabelledby = linkedSignal(() => this.ariaLabelledby(), ...(ngDevMode ? [{ debugName: "mutableAriaLabelledby" }] : []));
    /**
     * ID of element that describes this switch for accessibility.
     */
    ariaDescribedby = input(null, ...(ngDevMode ? [{ debugName: "ariaDescribedby", alias: 'aria-describedby' }] : [{ alias: 'aria-describedby' }]));
    /**
     * Whether switch is required in a form.
     */
    required = input(false, ...(ngDevMode ? [{ debugName: "required", transform: booleanAttribute }] : [{ transform: booleanAttribute }]));
    /**
     * Whether switch is disabled.
     * Disabled switches cannot be toggled and indicate disabled state with data attribute.
     */
    disabled = input(false, ...(ngDevMode ? [{ debugName: "disabled", transform: booleanAttribute }] : [{
            transform: booleanAttribute,
        }]));
    /**
     * Keyboard tab order for switch.
     * @default 0
     */
    tabIndex = input(0, ...(ngDevMode ? [{ debugName: "tabIndex", transform: numberAttribute }] : [{ transform: numberAttribute }]));
    /**
     * Event emitted when switch is blurred (loses focus).
     * Used for form validation.
     */
    touched = output();
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    _onChange = () => { };
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    _onTouched = () => { };
    switch = viewChild.required('switch');
    _state = computed(() => {
        const name = this.name();
        const id = this.id();
        return {
            disabled: signal(this.disabled()),
            name: name ? name + CONTAINER_POST_FIX : null,
            id: id ? id + CONTAINER_POST_FIX : null,
        };
    }, ...(ngDevMode ? [{ debugName: "_state" }] : []));
    controlState = this._fieldControl?.controlState;
    _invalid = this._fieldControl?.invalid;
    _touched = this._fieldControl?.touched;
    _dirty = this._fieldControl?.dirty;
    _spartanInvalid = this._fieldControl?.spartanInvalid;
    labelableId = computed(() => this.getSwitchButtonId(this._state().id), ...(ngDevMode ? [{ debugName: "labelableId" }] : []));
    constructor() {
        afterRenderEffect(() => {
            const state = this._state();
            const isDisabled = state.disabled();
            if (!this._elementRef.nativeElement || !this._isBrowser)
                return;
            const newLabelId = state.id + '-label';
            const switchButtonId = this.getSwitchButtonId(state.id);
            const labelElement = this._elementRef.nativeElement.closest('label') ??
                this._document.querySelector(`label[for="${switchButtonId}"]`);
            if (!labelElement)
                return;
            const existingLabelId = labelElement.id;
            this._renderer.setAttribute(labelElement, 'data-disabled', isDisabled ? 'true' : 'false');
            this.mutableAriaLabelledby.set(existingLabelId || newLabelId);
            if (!existingLabelId || existingLabelId.length === 0) {
                this._renderer.setAttribute(labelElement, 'id', newLabelId);
            }
        });
    }
    /**
     * Toggles switch between checked/unchecked states.
     * Does nothing if switch is disabled.
     */
    toggle() {
        if (this._state().disabled())
            return;
        this._onTouched();
        this.touched.emit();
        this.checked.update((checked) => !checked);
        this._onChange(this.checked());
        this.checkedChange.emit(this.checked());
    }
    ngAfterContentInit() {
        this._focusMonitor
            .monitor(this._elementRef, true)
            .pipe(takeUntilDestroyed(this._destroyRef))
            .subscribe((focusOrigin) => {
            if (focusOrigin)
                this._focused.set(true);
            if (focusOrigin === 'keyboard' || focusOrigin === 'program') {
                this._focusVisible.set(true);
                this._cdr.markForCheck();
            }
            if (!focusOrigin) {
                // When a focused element becomes disabled, the browser *immediately* fires a blur event.
                // Angular does not expect events to be raised during change detection, so any state
                // change (such as a form control's ng-touched) will cause a changed-after-checked error.
                // See https://github.com/angular/angular/issues/17793. To work around this, we defer
                // telling the form control it has been touched until the next tick.
                Promise.resolve().then(() => {
                    this._focusVisible.set(false);
                    this._focused.set(false);
                    this._onTouched();
                    this.touched.emit();
                    this._cdr.markForCheck();
                });
            }
        });
        if (!this.switch())
            return;
        this.switch().nativeElement.value = this.checked() ? 'on' : 'off';
        this.switch().nativeElement.dispatchEvent(new Event('change'));
    }
    ngOnDestroy() {
        this._focusMonitor.stopMonitoring(this._elementRef);
    }
    /**
     * Gets proper ID for inner button element.
     * Removes '-switch' suffix if present in container ID.
     *
     * @param idPassedToContainer - ID applied to container element
     * @returns ID to use for inner button or null
     */
    getSwitchButtonId(idPassedToContainer) {
        return idPassedToContainer ? idPassedToContainer.replace(new RegExp(CONTAINER_POST_FIX + '$'), '') : null;
    }
    /**
     * Updates internal state when control value changes from outside.
     * Part of ControlValueAccessor interface.
     *
     * @param value - New checked state
     */
    writeValue(value) {
        this.checked.set(Boolean(value));
    }
    /**
     * Registers callback for value changes.
     * Part of ControlValueAccessor interface.
     *
     * @param fn - Function to call when value changes
     */
    registerOnChange(fn) {
        this._onChange = fn;
    }
    /**
     * Registers callback for touched events.
     * Part of ControlValueAccessor interface.
     *
     * @param fn - Function to call when control is touched
     */
    registerOnTouched(fn) {
        this._onTouched = fn;
    }
    /**
     * Updates disabled state from form control.
     * Part of ControlValueAccessor interface.
     *
     * @param isDisabled - Whether switch should be disabled
     */
    setDisabledState(isDisabled) {
        this._state().disabled.set(isDisabled);
        this._cdr.markForCheck();
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSwitch, deps: [], target: i0.ɵɵFactoryTarget.Component });
    /** @nocollapse */ static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.2.0", version: "20.3.17", type: BrnSwitch, isStandalone: true, selector: "brn-switch", inputs: { checked: { classPropertyName: "checked", publicName: "checked", isSignal: true, isRequired: false, transformFunction: null }, id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null }, name: { classPropertyName: "name", publicName: "name", isSignal: true, isRequired: false, transformFunction: null }, class: { classPropertyName: "class", publicName: "class", isSignal: true, isRequired: false, transformFunction: null }, ariaLabel: { classPropertyName: "ariaLabel", publicName: "aria-label", isSignal: true, isRequired: false, transformFunction: null }, ariaLabelledby: { classPropertyName: "ariaLabelledby", publicName: "aria-labelledby", isSignal: true, isRequired: false, transformFunction: null }, ariaDescribedby: { classPropertyName: "ariaDescribedby", publicName: "aria-describedby", isSignal: true, isRequired: false, transformFunction: null }, required: { classPropertyName: "required", publicName: "required", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "disabled", isSignal: true, isRequired: false, transformFunction: null }, tabIndex: { classPropertyName: "tabIndex", publicName: "tabIndex", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { checked: "checkedChange", checkedChange: "checkedChange", touched: "touched" }, host: { properties: { "style": "{display: \"contents\"}", "attr.id": "_state().id", "attr.name": "_state().name", "attr.aria-labelledby": "null", "attr.aria-label": "null", "attr.aria-describedby": "null", "attr.aria-invalid": "_invalid?.() ? \"true\" : null", "attr.data-dirty": "_dirty?.() ? \"true\": null", "attr.data-touched": "_touched?.() ? \"true\" : null", "attr.data-matches-spartan-invalid": "_spartanInvalid?.() ? \"true\" : null", "attr.data-state": "checked() ? \"checked\" : \"unchecked\"", "attr.data-focus-visible": "_focusVisible()", "attr.data-focus": "_focused()", "attr.data-disabled": "_state().disabled()" } }, providers: [BRN_SWITCH_VALUE_ACCESSOR, provideBrnLabelable(BrnSwitch)], viewQueries: [{ propertyName: "switch", first: true, predicate: ["switch"], descendants: true, isSignal: true }], hostDirectives: [{ directive: i1.BrnFieldControl }], ngImport: i0, template: `
		<button
			#switch
			role="switch"
			type="button"
			[class]="class()"
			[id]="getSwitchButtonId(_state().id) ?? ''"
			[name]="getSwitchButtonId(_state().name) ?? ''"
			[value]="checked() ? 'on' : 'off'"
			[attr.aria-checked]="checked()"
			[attr.aria-label]="ariaLabel() || null"
			[attr.aria-labelledby]="mutableAriaLabelledby() || null"
			[attr.aria-describedby]="ariaDescribedby() || null"
			[attr.aria-invalid]="_invalid?.() ? 'true' : null"
			[attr.data-dirty]="_dirty?.() ? 'true' : null"
			[attr.data-touched]="_touched?.() ? 'true' : null"
			[attr.data-matches-spartan-invalid]="_spartanInvalid?.() ? 'true' : null"
			[attr.data-state]="checked() ? 'checked' : 'unchecked'"
			[attr.data-focus-visible]="_focusVisible()"
			[attr.data-focus]="_focused()"
			[attr.data-disabled]="_state().disabled()"
			[disabled]="_state().disabled()"
			[tabIndex]="tabIndex()"
			(click)="$event.preventDefault(); toggle()"
		>
			<ng-content select="brn-switch-thumb" />
		</button>
	`, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSwitch, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-switch',
                    providers: [BRN_SWITCH_VALUE_ACCESSOR, provideBrnLabelable(BrnSwitch)],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    hostDirectives: [BrnFieldControl],
                    host: {
                        '[style]': '{display: "contents"}',
                        '[attr.id]': '_state().id',
                        '[attr.name]': '_state().name',
                        '[attr.aria-labelledby]': 'null',
                        '[attr.aria-label]': 'null',
                        '[attr.aria-describedby]': 'null',
                        '[attr.aria-invalid]': '_invalid?.() ? "true" : null',
                        '[attr.data-dirty]': '_dirty?.() ? "true": null',
                        '[attr.data-touched]': '_touched?.() ? "true" : null',
                        '[attr.data-matches-spartan-invalid]': '_spartanInvalid?.() ? "true" : null',
                        '[attr.data-state]': 'checked() ? "checked" : "unchecked"',
                        '[attr.data-focus-visible]': '_focusVisible()',
                        '[attr.data-focus]': '_focused()',
                        '[attr.data-disabled]': '_state().disabled()',
                    },
                    template: `
		<button
			#switch
			role="switch"
			type="button"
			[class]="class()"
			[id]="getSwitchButtonId(_state().id) ?? ''"
			[name]="getSwitchButtonId(_state().name) ?? ''"
			[value]="checked() ? 'on' : 'off'"
			[attr.aria-checked]="checked()"
			[attr.aria-label]="ariaLabel() || null"
			[attr.aria-labelledby]="mutableAriaLabelledby() || null"
			[attr.aria-describedby]="ariaDescribedby() || null"
			[attr.aria-invalid]="_invalid?.() ? 'true' : null"
			[attr.data-dirty]="_dirty?.() ? 'true' : null"
			[attr.data-touched]="_touched?.() ? 'true' : null"
			[attr.data-matches-spartan-invalid]="_spartanInvalid?.() ? 'true' : null"
			[attr.data-state]="checked() ? 'checked' : 'unchecked'"
			[attr.data-focus-visible]="_focusVisible()"
			[attr.data-focus]="_focused()"
			[attr.data-disabled]="_state().disabled()"
			[disabled]="_state().disabled()"
			[tabIndex]="tabIndex()"
			(click)="$event.preventDefault(); toggle()"
		>
			<ng-content select="brn-switch-thumb" />
		</button>
	`,
                }]
        }], ctorParameters: () => [], propDecorators: { checked: [{ type: i0.Input, args: [{ isSignal: true, alias: "checked", required: false }] }, { type: i0.Output, args: ["checkedChange"] }], checkedChange: [{ type: i0.Output, args: ["checkedChange"] }], id: [{ type: i0.Input, args: [{ isSignal: true, alias: "id", required: false }] }], name: [{ type: i0.Input, args: [{ isSignal: true, alias: "name", required: false }] }], class: [{ type: i0.Input, args: [{ isSignal: true, alias: "class", required: false }] }], ariaLabel: [{ type: i0.Input, args: [{ isSignal: true, alias: "aria-label", required: false }] }], ariaLabelledby: [{ type: i0.Input, args: [{ isSignal: true, alias: "aria-labelledby", required: false }] }], ariaDescribedby: [{ type: i0.Input, args: [{ isSignal: true, alias: "aria-describedby", required: false }] }], required: [{ type: i0.Input, args: [{ isSignal: true, alias: "required", required: false }] }], disabled: [{ type: i0.Input, args: [{ isSignal: true, alias: "disabled", required: false }] }], tabIndex: [{ type: i0.Input, args: [{ isSignal: true, alias: "tabIndex", required: false }] }], touched: [{ type: i0.Output, args: ["touched"] }], switch: [{ type: i0.ViewChild, args: ['switch', { isSignal: true }] }] } });

class BrnSwitchThumb {
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSwitchThumb, deps: [], target: i0.ɵɵFactoryTarget.Component });
    /** @nocollapse */ static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "20.3.17", type: BrnSwitchThumb, isStandalone: true, selector: "brn-switch-thumb", host: { attributes: { "role": "presentation" }, listeners: { "click": "$event.preventDefault()" } }, ngImport: i0, template: '', isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnSwitchThumb, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-switch-thumb',
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        role: 'presentation',
                        '(click)': '$event.preventDefault()',
                    },
                    template: '',
                }]
        }] });

const BrnSwitchImports = [BrnSwitch, BrnSwitchThumb];

/**
 * Generated bundle index. Do not edit.
 */

export { BRN_SWITCH_VALUE_ACCESSOR, BrnSwitch, BrnSwitchImports, BrnSwitchThumb };
