// @spartan-ng/brain/fesm2022/spartan-ng-brain-toggle.mjs
import * as i0 from '@angular/core';
import { input, booleanAttribute, model, computed, Directive } from '@angular/core';

class BrnToggle {
    static _uniqueId = 0;
    /** The id of the toggle. */
    id = input(`brn-toggle-${++BrnToggle._uniqueId}`, ...(ngDevMode ? [{ debugName: "id" }] : []));
    /** The value this toggle represents. */
    value = input(...(ngDevMode ? [undefined, { debugName: "value" }] : []));
    /** Whether the toggle is disabled. */
    disabled = input(false, ...(ngDevMode ? [{ debugName: "disabled", transform: booleanAttribute }] : [{
            transform: booleanAttribute,
        }]));
    /** The current state of the toggle when not used in a group. */
    state = model('off', ...(ngDevMode ? [{ debugName: "state" }] : []));
    /** The type of the button. */
    type = input('button', ...(ngDevMode ? [{ debugName: "type" }] : []));
    /**
     * Accessibility label for screen readers.
     * Use when no visible label exists.
     */
    ariaLabel = input(null, ...(ngDevMode ? [{ debugName: "ariaLabel", alias: 'aria-label' }] : [{ alias: 'aria-label' }]));
    /** Whether the toggle is in the on state. */
    _isOn = computed(() => this.state() === 'on', ...(ngDevMode ? [{ debugName: "_isOn" }] : []));
    toggle() {
        if (this.disabled())
            return;
        this.state.set(this._isOn() ? 'off' : 'on');
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnToggle, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.3.17", type: BrnToggle, isStandalone: true, selector: "button[brnToggle]", inputs: { id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null }, value: { classPropertyName: "value", publicName: "value", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "disabled", isSignal: true, isRequired: false, transformFunction: null }, state: { classPropertyName: "state", publicName: "state", isSignal: true, isRequired: false, transformFunction: null }, type: { classPropertyName: "type", publicName: "type", isSignal: true, isRequired: false, transformFunction: null }, ariaLabel: { classPropertyName: "ariaLabel", publicName: "aria-label", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { state: "stateChange" }, host: { listeners: { "click": "toggle()" }, properties: { "id": "id()", "attr.disabled": "disabled() ? true : null", "attr.data-disabled": "disabled() ? true : null", "attr.data-state": "state()", "attr.aria-pressed": "_isOn()", "attr.aria-label": "ariaLabel() || null", "type": "type()" } }, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnToggle, decorators: [{
            type: Directive,
            args: [{
                    selector: 'button[brnToggle]',
                    host: {
                        '[id]': 'id()',
                        '[attr.disabled]': 'disabled() ? true : null',
                        '[attr.data-disabled]': 'disabled() ? true : null',
                        '[attr.data-state]': 'state()',
                        '[attr.aria-pressed]': '_isOn()',
                        '[attr.aria-label]': 'ariaLabel() || null',
                        '[type]': 'type()',
                        '(click)': 'toggle()',
                    },
                }]
        }], propDecorators: { id: [{ type: i0.Input, args: [{ isSignal: true, alias: "id", required: false }] }], value: [{ type: i0.Input, args: [{ isSignal: true, alias: "value", required: false }] }], disabled: [{ type: i0.Input, args: [{ isSignal: true, alias: "disabled", required: false }] }], state: [{ type: i0.Input, args: [{ isSignal: true, alias: "state", required: false }] }, { type: i0.Output, args: ["stateChange"] }], type: [{ type: i0.Input, args: [{ isSignal: true, alias: "type", required: false }] }], ariaLabel: [{ type: i0.Input, args: [{ isSignal: true, alias: "aria-label", required: false }] }] } });

const BrnToggleImports = [BrnToggle];

/**
 * Generated bundle index. Do not edit.
 */

export { BrnToggle, BrnToggleImports };
