// @spartan-ng/brain/fesm2022/spartan-ng-brain-tooltip.mjs
import { Overlay, OverlayPositionBuilder } from '@angular/cdk/overlay';
import { ComponentPortal } from '@angular/cdk/portal';
import { DOCUMENT } from '@angular/common';
import * as i0 from '@angular/core';
import { TemplateRef, inject, ViewContainerRef, input, effect, Directive, signal, ChangeDetectionStrategy, Component, InjectionToken, DestroyRef, ElementRef, Injector, Renderer2, booleanAttribute, linkedSignal, numberAttribute, output, computed, afterNextRender, runInInjectionContext } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { Directionality } from '@angular/cdk/bidi';
import { Subject, of, timer } from 'rxjs';
import { switchMap, map } from 'rxjs/operators';

function isTemplateRef(value) {
    return value instanceof TemplateRef;
}
class BrnTooltipStringTemplateOutlet {
    _viewContainer = inject(ViewContainerRef);
    _templateRef = inject((TemplateRef));
    _embeddedViewRef = null;
    _context = {};
    _isFirstChange = true;
    _lastOutletWasTemplate = false;
    _lastTemplateRef = null;
    _lastContext;
    brnTooltipStringTemplateOutletContext = input(undefined, ...(ngDevMode ? [{ debugName: "brnTooltipStringTemplateOutletContext" }] : []));
    brnTooltipStringTemplateOutlet = input.required(...(ngDevMode ? [{ debugName: "brnTooltipStringTemplateOutlet" }] : []));
    _hasContextShapeChanged(context) {
        if (!context) {
            return false;
        }
        const prevCtxKeys = Object.keys(this._lastContext || {});
        const currCtxKeys = Object.keys(context || {});
        if (prevCtxKeys.length === currCtxKeys.length) {
            for (const propName of currCtxKeys) {
                if (!prevCtxKeys.includes(propName)) {
                    return true;
                }
            }
            return false;
        }
        else {
            return true;
        }
    }
    _shouldViewBeRecreated(stringTemplateOutlet, stringTemplateOutletContext) {
        const isTemplate = isTemplateRef(stringTemplateOutlet);
        const shouldOutletRecreate = this._isFirstChange ||
            isTemplate !== this._lastOutletWasTemplate ||
            (isTemplate && stringTemplateOutlet !== this._lastTemplateRef);
        const shouldContextRecreate = this._hasContextShapeChanged(stringTemplateOutletContext);
        return shouldContextRecreate || shouldOutletRecreate;
    }
    _updateTrackingState(stringTemplateOutlet, stringTemplateOutletContext) {
        const isTemplate = isTemplateRef(stringTemplateOutlet);
        if (this._isFirstChange && !isTemplate) {
            this._isFirstChange = false;
        }
        if (stringTemplateOutletContext !== undefined) {
            this._lastContext = stringTemplateOutletContext;
        }
        this._lastOutletWasTemplate = isTemplate;
        this._lastTemplateRef = isTemplate ? stringTemplateOutlet : null;
    }
    _viewEffect = effect(() => {
        const stringTemplateOutlet = this.brnTooltipStringTemplateOutlet();
        const stringTemplateOutletContext = this.brnTooltipStringTemplateOutletContext();
        if (!this._isFirstChange && isTemplateRef(stringTemplateOutlet)) {
            this._isFirstChange = true;
        }
        if (!isTemplateRef(stringTemplateOutlet)) {
            this._context['$implicit'] = stringTemplateOutlet;
        }
        const recreateView = this._shouldViewBeRecreated(stringTemplateOutlet, stringTemplateOutletContext);
        this._updateTrackingState(stringTemplateOutlet, stringTemplateOutletContext);
        if (recreateView) {
            this._recreateView(stringTemplateOutlet, stringTemplateOutletContext);
        }
        else {
            this._updateContext(stringTemplateOutlet, stringTemplateOutletContext);
        }
    }, ...(ngDevMode ? [{ debugName: "_viewEffect" }] : []));
    _recreateView(outlet, context) {
        this._viewContainer.clear();
        if (isTemplateRef(outlet)) {
            this._embeddedViewRef = this._viewContainer.createEmbeddedView(outlet, context);
        }
        else {
            this._embeddedViewRef = this._viewContainer.createEmbeddedView(this._templateRef, this._context);
        }
    }
    _updateContext(outlet, context) {
        const newCtx = isTemplateRef(outlet) ? context : this._context;
        let oldCtx = this._embeddedViewRef?.context;
        if (!oldCtx) {
            oldCtx = newCtx;
        }
        else if (newCtx && typeof newCtx === 'object') {
            for (const propName of Object.keys(newCtx)) {
                oldCtx[propName] = newCtx[propName];
            }
        }
        this._lastContext = oldCtx;
    }
    static ngTemplateContextGuard(_dir, _ctx) {
        return true;
    }
    ngOnDestroy() {
        this._viewEffect.destroy();
        this._viewContainer.clear();
        this._embeddedViewRef = null;
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnTooltipStringTemplateOutlet, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.3.17", type: BrnTooltipStringTemplateOutlet, isStandalone: true, selector: "[brnTooltipStringTemplateOutlet]", inputs: { brnTooltipStringTemplateOutletContext: { classPropertyName: "brnTooltipStringTemplateOutletContext", publicName: "brnTooltipStringTemplateOutletContext", isSignal: true, isRequired: false, transformFunction: null }, brnTooltipStringTemplateOutlet: { classPropertyName: "brnTooltipStringTemplateOutlet", publicName: "brnTooltipStringTemplateOutlet", isSignal: true, isRequired: true, transformFunction: null } }, exportAs: ["brnTooltipStringTemplateOutlet"], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnTooltipStringTemplateOutlet, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnTooltipStringTemplateOutlet]',
                    exportAs: 'brnTooltipStringTemplateOutlet',
                }]
        }], propDecorators: { brnTooltipStringTemplateOutletContext: [{ type: i0.Input, args: [{ isSignal: true, alias: "brnTooltipStringTemplateOutletContext", required: false }] }], brnTooltipStringTemplateOutlet: [{ type: i0.Input, args: [{ isSignal: true, alias: "brnTooltipStringTemplateOutlet", required: true }] }] } });

let uniqueId = 0;
class BrnTooltipContent {
    id = input(`brn-tooltip-${++uniqueId}`, ...(ngDevMode ? [{ debugName: "id" }] : []));
    state = signal('closed', ...(ngDevMode ? [{ debugName: "state" }] : []));
    _tooltipClass = signal('', ...(ngDevMode ? [{ debugName: "_tooltipClass" }] : []));
    _arrowClass = signal('', ...(ngDevMode ? [{ debugName: "_arrowClass" }] : []));
    _svgClass = signal('', ...(ngDevMode ? [{ debugName: "_svgClass" }] : []));
    _position = signal('top', ...(ngDevMode ? [{ debugName: "_position" }] : []));
    _tooltipText = signal(null, ...(ngDevMode ? [{ debugName: "_tooltipText" }] : []));
    setProps(tooltipText, position, tooltipClasses, arrowClasses, svgClasses) {
        if (tooltipText) {
            this._tooltipText.set(tooltipText);
        }
        this._position.set(position);
        this._tooltipClass.set(tooltipClasses);
        this._arrowClass.set(arrowClasses);
        this._svgClass.set(svgClasses);
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnTooltipContent, deps: [], target: i0.ɵɵFactoryTarget.Component });
    /** @nocollapse */ static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.1.0", version: "20.3.17", type: BrnTooltipContent, isStandalone: true, selector: "ng-component", inputs: { id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null } }, host: { attributes: { "role": "tooltip" }, properties: { "class": "_tooltipClass()", "attr.id": "id()", "attr.data-side": "_position()", "attr.data-state": "state()" } }, ngImport: i0, template: `
		<ng-container *brnTooltipStringTemplateOutlet="_tooltipText()">{{ _tooltipText() }}</ng-container>

		<span [class]="_arrowClass()">
			<svg [class]="_svgClass()" width="10" height="5" viewBox="0 0 30 10" preserveAspectRatio="none">
				<polygon points="0,0 30,0 15,10" />
			</svg>
		</span>
	`, isInline: true, dependencies: [{ kind: "directive", type: BrnTooltipStringTemplateOutlet, selector: "[brnTooltipStringTemplateOutlet]", inputs: ["brnTooltipStringTemplateOutletContext", "brnTooltipStringTemplateOutlet"], exportAs: ["brnTooltipStringTemplateOutlet"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnTooltipContent, decorators: [{
            type: Component,
            args: [{
                    imports: [BrnTooltipStringTemplateOutlet],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        '[class]': '_tooltipClass()',
                        '[attr.id]': 'id()',
                        '[attr.data-side]': '_position()',
                        '[attr.data-state]': 'state()',
                        role: 'tooltip',
                    },
                    template: `
		<ng-container *brnTooltipStringTemplateOutlet="_tooltipText()">{{ _tooltipText() }}</ng-container>

		<span [class]="_arrowClass()">
			<svg [class]="_svgClass()" width="10" height="5" viewBox="0 0 30 10" preserveAspectRatio="none">
				<polygon points="0,0 30,0 15,10" />
			</svg>
		</span>
	`,
                }]
        }], propDecorators: { id: [{ type: i0.Input, args: [{ isSignal: true, alias: "id", required: false }] }] } });

const BRN_TOOLTIP_POSITIONS_MAP = {
    top: {
        originX: 'center',
        originY: 'top',
        overlayX: 'center',
        overlayY: 'bottom',
        offsetY: -8,
    },
    bottom: {
        originX: 'center',
        originY: 'bottom',
        overlayX: 'center',
        overlayY: 'top',
        offsetY: 8,
    },
    left: {
        originX: 'start',
        originY: 'center',
        overlayX: 'end',
        overlayY: 'center',
        offsetX: -8,
    },
    right: {
        originX: 'end',
        originY: 'center',
        overlayX: 'start',
        overlayY: 'center',
        offsetX: 8,
    },
};

const defaultOptions = {
    showDelay: 150,
    hideDelay: 100,
    svgClasses: '',
    arrowClasses: () => '',
    tooltipContentClasses: '',
};
const BRN_TOOLTIP_DEFAULT_OPTIONS = new InjectionToken('brn-tooltip-default-options', {
    providedIn: 'root',
    factory: () => defaultOptions,
});
function provideBrnTooltipDefaultOptions(options) {
    return { provide: BRN_TOOLTIP_DEFAULT_OPTIONS, useValue: { ...defaultOptions, ...options } };
}
function injectBrnTooltipDefaultOptions() {
    return inject(BRN_TOOLTIP_DEFAULT_OPTIONS, { optional: true }) ?? defaultOptions;
}

class BrnTooltip {
    _config = injectBrnTooltipDefaultOptions();
    _destroyRef = inject(DestroyRef);
    _document = inject(DOCUMENT);
    _elementRef = inject((ElementRef));
    _injector = inject(Injector);
    _overlay = inject(Overlay);
    _overlayPositionBuilder = inject(OverlayPositionBuilder);
    _renderer = inject(Renderer2);
    _dir = inject(Directionality);
    _tooltipHovered = false;
    _listenersRefs = [];
    _delaySubject = undefined;
    _componentRef = undefined;
    _overlayRef = undefined;
    _ariaEffectRef = undefined;
    tooltipDisabled = input(false, ...(ngDevMode ? [{ debugName: "tooltipDisabled", transform: booleanAttribute }] : [{ transform: booleanAttribute }]));
    mutableTooltipDisabled = linkedSignal(this.tooltipDisabled, ...(ngDevMode ? [{ debugName: "mutableTooltipDisabled" }] : []));
    position = input(this._config.position ?? 'top', ...(ngDevMode ? [{ debugName: "position" }] : []));
    brnTooltip = input(null, ...(ngDevMode ? [{ debugName: "brnTooltip" }] : []));
    showDelay = input(this._config.showDelay, ...(ngDevMode ? [{ debugName: "showDelay", transform: numberAttribute }] : [{ transform: numberAttribute }]));
    hideDelay = input(this._config.hideDelay, ...(ngDevMode ? [{ debugName: "hideDelay", transform: numberAttribute }] : [{ transform: numberAttribute }]));
    show = output();
    hide = output();
    _tooltipText = computed(() => {
        let tooltipText = this.brnTooltip();
        if (!tooltipText) {
            return '';
        }
        else if (typeof tooltipText === 'string') {
            tooltipText = tooltipText.trim();
        }
        return tooltipText;
    }, ...(ngDevMode ? [{ debugName: "_tooltipText" }] : []));
    constructor() {
        afterNextRender(() => {
            this._overlayRef = this._overlay.create({
                direction: this._dir,
                positionStrategy: this._buildPositionStrategy(),
            });
            this._dir.change.pipe(takeUntilDestroyed(this._destroyRef)).subscribe(() => {
                if (this._overlayRef) {
                    this._updatePosition();
                    if (this._overlayRef.hasAttached()) {
                        this._overlayRef.updatePosition();
                    }
                }
            });
            runInInjectionContext(this._injector, () => {
                if (!this._overlayRef)
                    return;
                this._setupDelayMechanism();
                this._cleanupTriggerEvents();
                this._initTriggers();
                this._listenersRefs = [
                    ...this._listenersRefs,
                    this._renderer.listen(this._overlayRef.hostElement, 'mouseenter', () => (this._tooltipHovered = true)),
                    this._renderer.listen(this._overlayRef.hostElement, 'mouseleave', () => {
                        this._tooltipHovered = false;
                        this.delay(false, this.hideDelay());
                    }),
                ];
            });
        });
        this._destroyRef.onDestroy(() => {
            this._clearAriaDescribedBy();
            this._delaySubject?.complete();
            this._cleanupTriggerEvents();
            this._overlayRef?.dispose();
        });
    }
    _updatePosition() {
        const strategy = this._overlayRef?.getConfig().positionStrategy;
        if (strategy) {
            strategy.withPositions([this._getAdjustedPosition()]);
        }
    }
    _buildPositionStrategy() {
        return this._overlayPositionBuilder
            .flexibleConnectedTo(this._elementRef)
            .withPositions([this._getAdjustedPosition()]);
    }
    _getAdjustedPosition() {
        const position = BRN_TOOLTIP_POSITIONS_MAP[this.position()];
        const isLtr = this._dir.value !== 'rtl';
        return {
            ...position,
            offsetX: position.offsetX != null ? (isLtr ? position.offsetX : -position.offsetX) : undefined,
        };
    }
    _initTriggers() {
        this._initScrollListener();
        this._initHoverListeners();
    }
    _initHoverListeners() {
        this._listenersRefs = [
            ...this._listenersRefs,
            this._renderer.listen(this._elementRef.nativeElement, 'mouseenter', () => this.delay(true, this.showDelay())),
            this._renderer.listen(this._elementRef.nativeElement, 'mouseleave', () => this.delay(false, this.hideDelay())),
            this._renderer.listen(this._elementRef.nativeElement, 'focus', () => this.delay(true, this.showDelay())),
            this._renderer.listen(this._elementRef.nativeElement, 'blur', () => this.delay(false, this.hideDelay())),
        ];
    }
    _initScrollListener() {
        this._listenersRefs = [
            ...this._listenersRefs,
            this._renderer.listen(this._document.defaultView, 'scroll', () => this._hide()),
        ];
    }
    _cleanupTriggerEvents() {
        for (const eventRef of this._listenersRefs) {
            eventRef();
        }
        this._listenersRefs = [];
    }
    delay(isShow, delay = -1) {
        this._delaySubject?.next({ isShow, delay });
    }
    _setupDelayMechanism() {
        this._delaySubject?.complete();
        this._delaySubject = new Subject();
        this._delaySubject
            .pipe(switchMap((config) => (config.delay < 0 ? of(config) : timer(config.delay).pipe(map(() => config)))), takeUntilDestroyed(this._destroyRef))
            .subscribe((config) => {
            if (config.isShow) {
                this._show();
            }
            else {
                this._hide();
            }
        });
    }
    _show() {
        if (this._componentRef || !this._tooltipText() || this.mutableTooltipDisabled()) {
            return;
        }
        const tooltipPortal = new ComponentPortal(BrnTooltipContent);
        this._componentRef = this._overlayRef?.attach(tooltipPortal);
        this._componentRef?.onDestroy(() => {
            this._componentRef = undefined;
        });
        this._componentRef?.instance.state.set('opened');
        this._componentRef?.instance.setProps(this._tooltipText(), this.position(), this._config.tooltipContentClasses, this._config.arrowClasses(this.position()), this._config.svgClasses);
        runInInjectionContext(this._injector, () => {
            this._ariaEffectRef = effect(() => {
                const tooltipId = this._componentRef?.instance.id();
                if (tooltipId) {
                    this._renderer.setAttribute(this._elementRef.nativeElement, 'aria-describedby', tooltipId);
                    this._ariaEffectRef?.destroy();
                    this._ariaEffectRef = undefined;
                }
            }, ...(ngDevMode ? [{ debugName: "_ariaEffectRef" }] : []));
        });
        this.show.emit();
    }
    _hide() {
        if (!this._componentRef || this._tooltipHovered)
            return;
        this._clearAriaDescribedBy();
        this._renderer.removeAttribute(this._elementRef.nativeElement, 'aria-describedby');
        this._componentRef.instance.state.set('closed');
        this.hide.emit();
        this._overlayRef?.detach();
    }
    _clearAriaDescribedBy() {
        if (this._ariaEffectRef) {
            this._ariaEffectRef.destroy();
            this._ariaEffectRef = undefined;
        }
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnTooltip, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.3.17", type: BrnTooltip, isStandalone: true, selector: "[brnTooltip]", inputs: { tooltipDisabled: { classPropertyName: "tooltipDisabled", publicName: "tooltipDisabled", isSignal: true, isRequired: false, transformFunction: null }, position: { classPropertyName: "position", publicName: "position", isSignal: true, isRequired: false, transformFunction: null }, brnTooltip: { classPropertyName: "brnTooltip", publicName: "brnTooltip", isSignal: true, isRequired: false, transformFunction: null }, showDelay: { classPropertyName: "showDelay", publicName: "showDelay", isSignal: true, isRequired: false, transformFunction: null }, hideDelay: { classPropertyName: "hideDelay", publicName: "hideDelay", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { show: "show", hide: "hide" }, exportAs: ["brnTooltip"], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnTooltip, decorators: [{
            type: Directive,
            args: [{ selector: '[brnTooltip]', exportAs: 'brnTooltip' }]
        }], ctorParameters: () => [], propDecorators: { tooltipDisabled: [{ type: i0.Input, args: [{ isSignal: true, alias: "tooltipDisabled", required: false }] }], position: [{ type: i0.Input, args: [{ isSignal: true, alias: "position", required: false }] }], brnTooltip: [{ type: i0.Input, args: [{ isSignal: true, alias: "brnTooltip", required: false }] }], showDelay: [{ type: i0.Input, args: [{ isSignal: true, alias: "showDelay", required: false }] }], hideDelay: [{ type: i0.Input, args: [{ isSignal: true, alias: "hideDelay", required: false }] }], show: [{ type: i0.Output, args: ["show"] }], hide: [{ type: i0.Output, args: ["hide"] }] } });

const BrnTooltipImports = [BrnTooltip, BrnTooltipContent];

/**
 * Generated bundle index. Do not edit.
 */

export { BrnTooltip, BrnTooltipContent, BrnTooltipImports, defaultOptions, injectBrnTooltipDefaultOptions, provideBrnTooltipDefaultOptions };
