// @spartan-ng/brain/fesm2022/spartan-ng-brain-dialog.mjs
import { OverlayOutsideClickDispatcher, OverlayPositionBuilder, ScrollStrategyOptions } from '@angular/cdk/overlay';
import * as i0 from '@angular/core';
import { InjectionToken, inject, signal, RendererFactory2, Injector, computed, runInInjectionContext, effect, Injectable, DestroyRef, ViewContainerRef, output, input, booleanAttribute, linkedSignal, untracked, numberAttribute, Directive, TemplateRef } from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { take, takeUntil, filter } from 'rxjs/operators';
import { DIALOG_DATA, Dialog } from '@angular/cdk/dialog';
import { Subject, merge } from 'rxjs';
import { coerceNumberProperty } from '@angular/cdk/coercion';
import { provideExposesStateProviderExisting, provideCustomClassSettableExisting } from '@spartan-ng/brain/core';

const defaultOptions = {
    ariaLabel: undefined,
    ariaModal: true,
    attachPositions: [],
    attachTo: null,
    autoFocus: 'first-tabbable',
    backdropClass: 'bg-transparent',
    closeDelay: 100,
    closeOnBackdropClick: true,
    closeOnOutsidePointerEvents: false,
    disableClose: false,
    hasBackdrop: true,
    panelClass: '',
    positionStrategy: null,
    restoreFocus: true,
    role: 'dialog',
    scrollStrategy: null,
};
const BRN_DIALOG_DEFAULT_OPTIONS = new InjectionToken('brn-dialog-default-options', {
    providedIn: 'root',
    factory: () => defaultOptions,
});
function provideBrnDialogDefaultOptions(options) {
    return { provide: BRN_DIALOG_DEFAULT_OPTIONS, useValue: { ...defaultOptions, ...options } };
}
function injectBrnDialogDefaultOptions() {
    return inject(BRN_DIALOG_DEFAULT_OPTIONS, { optional: true }) ?? defaultOptions;
}

// brn-dialog-utils.ts
const cssClassesToArray = (classes, defaultClass = '') => {
    if (typeof classes === 'string') {
        const splitClasses = classes.trim().split(' ');
        if (splitClasses.length === 0) {
            return [defaultClass];
        }
        return splitClasses;
    }
    return classes ?? [];
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
class BrnDialogRef {
    _cdkDialogRef;
    _open;
    state;
    dialogId;
    _closing$ = new Subject();
    closing$ = this._closing$.asObservable();
    closed$;
    _previousTimeout;
    get open() {
        return this.state() === 'open';
    }
    _options = signal(undefined, ...(ngDevMode ? [{ debugName: "_options" }] : []));
    options = this._options.asReadonly();
    constructor(_cdkDialogRef, _open, state, dialogId, _options) {
        this._cdkDialogRef = _cdkDialogRef;
        this._open = _open;
        this.state = state;
        this.dialogId = dialogId;
        if (_options) {
            this._options.set(_options);
        }
        this.closed$ = this._cdkDialogRef.closed.pipe(take(1));
    }
    updateOptions(options) {
        this._options.update((prev) => ({ ...(prev ?? {}), ...options }));
    }
    close(result, delay = this._options()?.closeDelay ?? 0) {
        if (!this.open || this._options()?.disableClose)
            return;
        this._closing$.next();
        this._open.set(false);
        if (this._previousTimeout) {
            clearTimeout(this._previousTimeout);
        }
        this._previousTimeout = setTimeout(() => {
            this._cdkDialogRef.close(result);
        }, delay);
    }
    setPanelClass(paneClass) {
        this._cdkDialogRef.config.panelClass = cssClassesToArray(paneClass);
    }
    setOverlayClass(overlayClass) {
        this._cdkDialogRef.config.backdropClass = cssClassesToArray(overlayClass);
    }
    setAriaDescribedBy(ariaDescribedBy) {
        this._cdkDialogRef.config.ariaDescribedBy = ariaDescribedBy;
    }
    setAriaLabelledBy(ariaLabelledBy) {
        this._cdkDialogRef.config.ariaLabelledBy = ariaLabelledBy;
    }
    setAriaLabel(ariaLabel) {
        this._cdkDialogRef.config.ariaLabel = ariaLabel;
    }
}

let dialogSequence = 0;
/** @deprecated `injectBrnDialogCtx` will no longer be supported once components are stable. Use `injectBrnDialogContext` instead.  */
const injectBrnDialogCtx = () => {
    return inject(DIALOG_DATA);
};
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const injectBrnDialogContext = (options = {}) => {
    return inject(DIALOG_DATA, options);
};
class BrnDialogService {
    _overlayCloseDispatcher = inject(OverlayOutsideClickDispatcher);
    _cdkDialog = inject(Dialog);
    _rendererFactory = inject(RendererFactory2);
    _renderer = this._rendererFactory.createRenderer(null, null);
    _positionBuilder = inject(OverlayPositionBuilder);
    _sso = inject(ScrollStrategyOptions);
    _injector = inject(Injector);
    _defaultOptions = injectBrnDialogDefaultOptions();
    open(content, vcr, context, options) {
        if (options?.id && this._cdkDialog.getDialogById(options.id)) {
            throw new Error(`Dialog with ID: ${options.id} already exists`);
        }
        const attachTo = options?.attachTo ?? this._defaultOptions.attachTo;
        const positionStrategy = options?.positionStrategy ??
            this._defaultOptions.positionStrategy ??
            (attachTo && options?.attachPositions && options?.attachPositions?.length > 0
                ? this._positionBuilder?.flexibleConnectedTo(attachTo).withPositions(options.attachPositions ?? [])
                : this._positionBuilder.global().centerHorizontally().centerVertically());
        let brnDialogRef;
        const effectRefs = [];
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const contextOrData = {
            ...context,
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            close: (result = undefined) => brnDialogRef.close(result, options?.closeDelay),
        };
        const destroyed$ = new Subject();
        const optionsChanged$ = new Subject();
        const open = signal(true, ...(ngDevMode ? [{ debugName: "open" }] : []));
        const state = computed(() => (open() ? 'open' : 'closed'), ...(ngDevMode ? [{ debugName: "state" }] : []));
        const dialogId = ++dialogSequence;
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        const cdkDialogRef = this._cdkDialog.open(content, {
            id: options?.id ?? `brn-dialog-${dialogId}`,
            role: options?.role ?? this._defaultOptions.role,
            viewContainerRef: vcr,
            templateContext: () => ({
                $implicit: contextOrData,
            }),
            data: contextOrData,
            hasBackdrop: options?.hasBackdrop ?? this._defaultOptions.hasBackdrop,
            panelClass: cssClassesToArray(options?.panelClass) ?? cssClassesToArray(this._defaultOptions.panelClass),
            backdropClass: cssClassesToArray(options?.backdropClass) ?? cssClassesToArray(this._defaultOptions.backdropClass),
            positionStrategy,
            scrollStrategy: options?.scrollStrategy ?? this._defaultOptions.scrollStrategy ?? this._sso?.block(),
            restoreFocus: options?.restoreFocus ?? this._defaultOptions.restoreFocus,
            disableClose: true,
            autoFocus: options?.autoFocus ?? this._defaultOptions.autoFocus,
            ariaDescribedBy: options?.ariaDescribedBy ?? `brn-dialog-description-${dialogId}`,
            ariaLabelledBy: options?.ariaLabelledBy ?? `brn-dialog-title-${dialogId}`,
            ariaLabel: options?.ariaLabel ?? this._defaultOptions.ariaLabel,
            ariaModal: options?.ariaModal ?? this._defaultOptions.ariaModal,
            providers: (cdkDialogRef) => {
                brnDialogRef = new BrnDialogRef(cdkDialogRef, open, state, dialogId, options);
                runInInjectionContext(this._injector, () => {
                    const ref = effect(() => {
                        if (overlay) {
                            this._renderer.setAttribute(overlay, 'data-state', state());
                        }
                        if (backdrop) {
                            this._renderer.setAttribute(backdrop, 'data-state', state());
                        }
                    }, ...(ngDevMode ? [{ debugName: "ref" }] : []));
                    effectRefs.push(ref);
                });
                const providers = [
                    {
                        provide: BrnDialogRef,
                        useValue: brnDialogRef,
                    },
                ];
                if (options?.providers) {
                    if (typeof options.providers === 'function') {
                        providers.push(...options.providers());
                    }
                    if (Array.isArray(options.providers)) {
                        providers.push(...options.providers);
                    }
                }
                return providers;
            },
        });
        const overlay = cdkDialogRef.overlayRef.overlayElement;
        const backdrop = cdkDialogRef.overlayRef.backdropElement;
        runInInjectionContext(this._injector, () => {
            const optionChangeEffect = effect(() => {
                const options = brnDialogRef.options();
                optionsChanged$.next();
                const closeOnOutsidePointerEvents = options?.closeOnOutsidePointerEvents ?? this._defaultOptions.closeOnOutsidePointerEvents;
                if (closeOnOutsidePointerEvents) {
                    cdkDialogRef.outsidePointerEvents.pipe(takeUntil(merge(destroyed$, optionsChanged$))).subscribe(() => {
                        const overlays = this._overlayCloseDispatcher._attachedOverlays;
                        const index = overlays.indexOf(cdkDialogRef.overlayRef);
                        // close if this is the topmost overlay
                        // but prevent closing if this overlay contains the top overlay (which will be closed)
                        if (index === overlays.length - 1 ||
                            (overlays.length > 1 && !this.isNested(cdkDialogRef.overlayRef, overlays.at(-1)))) {
                            brnDialogRef.close(undefined, options?.closeDelay);
                        }
                    });
                }
                const closeOnBackdropClick = options?.closeOnBackdropClick ?? this._defaultOptions.closeOnBackdropClick;
                if (closeOnBackdropClick) {
                    cdkDialogRef.backdropClick.pipe(takeUntil(merge(destroyed$, optionsChanged$))).subscribe(() => {
                        brnDialogRef.close(undefined, options?.closeDelay);
                    });
                }
                const disableClose = options?.disableClose ?? this._defaultOptions.disableClose;
                if (!disableClose) {
                    cdkDialogRef.keydownEvents
                        .pipe(filter((e) => e.key === 'Escape'), takeUntil(merge(destroyed$, optionsChanged$)))
                        .subscribe(() => {
                        brnDialogRef.close(undefined, options?.closeDelay);
                    });
                }
            }, ...(ngDevMode ? [{ debugName: "optionChangeEffect" }] : []));
            effectRefs.push(optionChangeEffect);
        });
        cdkDialogRef.closed.pipe(takeUntil(destroyed$)).subscribe(() => {
            effectRefs.forEach((a) => a.destroy());
            destroyed$.next();
            optionsChanged$.next();
        });
        // this is not ideal, but we need to force change detection to run in the dialog,
        // otherwise in some conditions (like menu launching a dialog) the dialog content
        // does not render.
        if ('_changeDetectorRef' in cdkDialogRef.containerInstance) {
            const containerInstance = cdkDialogRef.containerInstance;
            containerInstance._changeDetectorRef.detectChanges();
        }
        return brnDialogRef;
    }
    isNested(parent, child) {
        const childOrigin = child.getConfig().positionStrategy._origin;
        if (!childOrigin) {
            return false;
        }
        else if ('width' in childOrigin && 'height' in childOrigin) {
            const rect = parent.hostElement.getBoundingClientRect();
            return (childOrigin.x >= rect.left &&
                childOrigin.x <= rect.right &&
                childOrigin.y >= rect.top &&
                childOrigin.y <= rect.bottom);
        }
        else {
            const element = childOrigin.nativeElement || childOrigin;
            return parent.hostElement.contains(element);
        }
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnDialogService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
    /** @nocollapse */ static ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnDialogService, providedIn: 'root' });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnDialogService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }] });

class BrnDialog {
    _dialogService = inject(BrnDialogService);
    _destroyRef = inject(DestroyRef);
    _vcr = inject(ViewContainerRef);
    positionBuilder = inject(OverlayPositionBuilder);
    ssos = inject(ScrollStrategyOptions);
    _injector = inject(Injector);
    _defaultOptions = injectBrnDialogDefaultOptions();
    _context = {};
    stateComputed = computed(() => this._dialogRef()?.state() ?? 'closed', ...(ngDevMode ? [{ debugName: "stateComputed" }] : []));
    _contentTemplate;
    _dialogRef = signal(undefined, ...(ngDevMode ? [{ debugName: "_dialogRef" }] : []));
    _dialogStateEffectRefs = [];
    _backdropClass = signal(null, ...(ngDevMode ? [{ debugName: "_backdropClass" }] : []));
    _panelClass = signal(null, ...(ngDevMode ? [{ debugName: "_panelClass" }] : []));
    closed = output();
    stateChanged = output();
    state = input(null, ...(ngDevMode ? [{ debugName: "state" }] : []));
    role = input(this._defaultOptions.role, ...(ngDevMode ? [{ debugName: "role" }] : []));
    hasBackdrop = input(this._defaultOptions.hasBackdrop, ...(ngDevMode ? [{ debugName: "hasBackdrop", transform: booleanAttribute }] : [{
            transform: booleanAttribute,
        }]));
    positionStrategy = input(this._defaultOptions.positionStrategy, ...(ngDevMode ? [{ debugName: "positionStrategy" }] : []));
    mutablePositionStrategy = linkedSignal(() => this.positionStrategy(), ...(ngDevMode ? [{ debugName: "mutablePositionStrategy" }] : []));
    scrollStrategy = input(this._defaultOptions.scrollStrategy, ...(ngDevMode ? [{ debugName: "scrollStrategy" }] : []));
    _options = computed(() => {
        const scrollStrategyInput = this.scrollStrategy();
        let scrollStrategy;
        if (scrollStrategyInput === 'close') {
            scrollStrategy = this.ssos.close();
        }
        else if (scrollStrategyInput === 'reposition') {
            scrollStrategy = this.ssos.reposition();
        }
        else {
            scrollStrategy = scrollStrategyInput;
        }
        return {
            role: this.role(),
            hasBackdrop: this.hasBackdrop(),
            positionStrategy: this.mutablePositionStrategy(),
            scrollStrategy,
            restoreFocus: this.restoreFocus(),
            closeOnOutsidePointerEvents: this.mutableCloseOnOutsidePointerEvents(),
            closeOnBackdropClick: this.closeOnBackdropClick(),
            attachTo: this.mutableAttachTo(),
            attachPositions: this.mutableAttachPositions(),
            autoFocus: this.autoFocus(),
            closeDelay: this.closeDelay(),
            disableClose: this.disableClose(),
            backdropClass: this._backdropClass() ?? '',
            panelClass: this._panelClass() ?? '',
            ariaDescribedBy: this._mutableAriaDescribedBy(),
            ariaLabelledBy: this._mutableAriaLabelledBy(),
            ariaLabel: this._mutableAriaLabel(),
            ariaModal: this._mutableAriaModal(),
        };
    }, ...(ngDevMode ? [{ debugName: "_options" }] : []));
    constructor() {
        effect(() => {
            const state = this.state();
            if (state === 'open') {
                untracked(() => this.open());
            }
            if (state === 'closed') {
                untracked(() => this.close());
            }
        });
    }
    restoreFocus = input(this._defaultOptions.restoreFocus, ...(ngDevMode ? [{ debugName: "restoreFocus" }] : []));
    closeOnOutsidePointerEvents = input(this._defaultOptions.closeOnOutsidePointerEvents, ...(ngDevMode ? [{ debugName: "closeOnOutsidePointerEvents", transform: booleanAttribute }] : [{
            transform: booleanAttribute,
        }]));
    mutableCloseOnOutsidePointerEvents = linkedSignal(() => this.closeOnOutsidePointerEvents(), ...(ngDevMode ? [{ debugName: "mutableCloseOnOutsidePointerEvents" }] : []));
    closeOnBackdropClick = input(this._defaultOptions.closeOnBackdropClick, ...(ngDevMode ? [{ debugName: "closeOnBackdropClick", transform: booleanAttribute }] : [{
            transform: booleanAttribute,
        }]));
    attachTo = input(null, ...(ngDevMode ? [{ debugName: "attachTo" }] : []));
    mutableAttachTo = linkedSignal(() => this.attachTo(), ...(ngDevMode ? [{ debugName: "mutableAttachTo" }] : []));
    attachPositions = input(this._defaultOptions.attachPositions, ...(ngDevMode ? [{ debugName: "attachPositions" }] : []));
    mutableAttachPositions = linkedSignal(() => this.attachPositions(), ...(ngDevMode ? [{ debugName: "mutableAttachPositions" }] : []));
    autoFocus = input(this._defaultOptions.autoFocus, ...(ngDevMode ? [{ debugName: "autoFocus" }] : []));
    closeDelay = input(this._defaultOptions.closeDelay, ...(ngDevMode ? [{ debugName: "closeDelay", transform: numberAttribute }] : [{
            transform: numberAttribute,
        }]));
    disableClose = input(this._defaultOptions.disableClose, ...(ngDevMode ? [{ debugName: "disableClose", transform: booleanAttribute }] : [{
            transform: booleanAttribute,
        }]));
    ariaDescribedBy = input(null, ...(ngDevMode ? [{ debugName: "ariaDescribedBy", alias: 'aria-describedby' }] : [{
            alias: 'aria-describedby',
        }]));
    _mutableAriaDescribedBy = linkedSignal(() => this.ariaDescribedBy(), ...(ngDevMode ? [{ debugName: "_mutableAriaDescribedBy" }] : []));
    ariaLabelledBy = input(null, ...(ngDevMode ? [{ debugName: "ariaLabelledBy", alias: 'aria-labelledby' }] : [{ alias: 'aria-labelledby' }]));
    _mutableAriaLabelledBy = linkedSignal(() => this.ariaLabelledBy(), ...(ngDevMode ? [{ debugName: "_mutableAriaLabelledBy" }] : []));
    ariaLabel = input(null, ...(ngDevMode ? [{ debugName: "ariaLabel", alias: 'aria-label' }] : [{ alias: 'aria-label' }]));
    _mutableAriaLabel = linkedSignal(() => this.ariaLabel(), ...(ngDevMode ? [{ debugName: "_mutableAriaLabel" }] : []));
    ariaModal = input(true, ...(ngDevMode ? [{ debugName: "ariaModal", alias: 'aria-modal', transform: booleanAttribute }] : [{ alias: 'aria-modal', transform: booleanAttribute }]));
    _mutableAriaModal = linkedSignal(() => this.ariaModal(), ...(ngDevMode ? [{ debugName: "_mutableAriaModal" }] : []));
    open() {
        if (!this._contentTemplate || this._dialogRef())
            return;
        this._dialogStateEffectRefs.forEach((ref) => ref.destroy());
        const dialogRef = this._dialogService.open(this._contentTemplate, this._vcr, this._context, this._options());
        this._dialogRef.set(dialogRef);
        runInInjectionContext(this._injector, () => {
            this._dialogStateEffectRefs.push(effect(() => {
                const state = dialogRef.state();
                untracked(() => this.stateChanged.emit(state));
            }), effect(() => dialogRef.updateOptions(this._options())));
        });
        dialogRef.closed$.pipe(take(1), takeUntilDestroyed(this._destroyRef)).subscribe((result) => {
            this._dialogRef.set(undefined);
            this.closed.emit(result);
        });
    }
    close(result, delay) {
        this._dialogRef()?.close(result, delay ?? this._options().closeDelay);
    }
    registerTemplate(template) {
        this._contentTemplate = template;
    }
    setOverlayClass(overlayClass) {
        this._backdropClass.set(overlayClass);
        this._dialogRef()?.setOverlayClass(overlayClass);
    }
    setPanelClass(panelClass) {
        this._panelClass.set(panelClass ?? '');
        this._dialogRef()?.setPanelClass(panelClass);
    }
    setContext(context) {
        this._context = { ...this._context, ...context };
    }
    setAriaDescribedBy(ariaDescribedBy) {
        this._mutableAriaDescribedBy.set(ariaDescribedBy);
        this._dialogRef()?.setAriaDescribedBy(ariaDescribedBy);
    }
    setAriaLabelledBy(ariaLabelledBy) {
        this._mutableAriaLabelledBy.set(ariaLabelledBy);
        this._dialogRef()?.setAriaLabelledBy(ariaLabelledBy);
    }
    setAriaLabel(ariaLabel) {
        this._mutableAriaLabel.set(ariaLabel);
        this._dialogRef()?.setAriaLabel(ariaLabel);
    }
    setAriaModal(ariaModal) {
        this._mutableAriaModal.set(ariaModal);
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnDialog, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.3.17", type: BrnDialog, isStandalone: true, selector: "[brnDialog],brn-dialog", inputs: { state: { classPropertyName: "state", publicName: "state", isSignal: true, isRequired: false, transformFunction: null }, role: { classPropertyName: "role", publicName: "role", isSignal: true, isRequired: false, transformFunction: null }, hasBackdrop: { classPropertyName: "hasBackdrop", publicName: "hasBackdrop", isSignal: true, isRequired: false, transformFunction: null }, positionStrategy: { classPropertyName: "positionStrategy", publicName: "positionStrategy", isSignal: true, isRequired: false, transformFunction: null }, scrollStrategy: { classPropertyName: "scrollStrategy", publicName: "scrollStrategy", isSignal: true, isRequired: false, transformFunction: null }, restoreFocus: { classPropertyName: "restoreFocus", publicName: "restoreFocus", isSignal: true, isRequired: false, transformFunction: null }, closeOnOutsidePointerEvents: { classPropertyName: "closeOnOutsidePointerEvents", publicName: "closeOnOutsidePointerEvents", isSignal: true, isRequired: false, transformFunction: null }, closeOnBackdropClick: { classPropertyName: "closeOnBackdropClick", publicName: "closeOnBackdropClick", isSignal: true, isRequired: false, transformFunction: null }, attachTo: { classPropertyName: "attachTo", publicName: "attachTo", isSignal: true, isRequired: false, transformFunction: null }, attachPositions: { classPropertyName: "attachPositions", publicName: "attachPositions", isSignal: true, isRequired: false, transformFunction: null }, autoFocus: { classPropertyName: "autoFocus", publicName: "autoFocus", isSignal: true, isRequired: false, transformFunction: null }, closeDelay: { classPropertyName: "closeDelay", publicName: "closeDelay", isSignal: true, isRequired: false, transformFunction: null }, disableClose: { classPropertyName: "disableClose", publicName: "disableClose", isSignal: true, isRequired: false, transformFunction: null }, ariaDescribedBy: { classPropertyName: "ariaDescribedBy", publicName: "aria-describedby", isSignal: true, isRequired: false, transformFunction: null }, ariaLabelledBy: { classPropertyName: "ariaLabelledBy", publicName: "aria-labelledby", isSignal: true, isRequired: false, transformFunction: null }, ariaLabel: { classPropertyName: "ariaLabel", publicName: "aria-label", isSignal: true, isRequired: false, transformFunction: null }, ariaModal: { classPropertyName: "ariaModal", publicName: "aria-modal", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { closed: "closed", stateChanged: "stateChanged" }, exportAs: ["brnDialog"], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnDialog, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnDialog],brn-dialog',
                    exportAs: 'brnDialog',
                }]
        }], ctorParameters: () => [], propDecorators: { closed: [{ type: i0.Output, args: ["closed"] }], stateChanged: [{ type: i0.Output, args: ["stateChanged"] }], state: [{ type: i0.Input, args: [{ isSignal: true, alias: "state", required: false }] }], role: [{ type: i0.Input, args: [{ isSignal: true, alias: "role", required: false }] }], hasBackdrop: [{ type: i0.Input, args: [{ isSignal: true, alias: "hasBackdrop", required: false }] }], positionStrategy: [{ type: i0.Input, args: [{ isSignal: true, alias: "positionStrategy", required: false }] }], scrollStrategy: [{ type: i0.Input, args: [{ isSignal: true, alias: "scrollStrategy", required: false }] }], restoreFocus: [{ type: i0.Input, args: [{ isSignal: true, alias: "restoreFocus", required: false }] }], closeOnOutsidePointerEvents: [{ type: i0.Input, args: [{ isSignal: true, alias: "closeOnOutsidePointerEvents", required: false }] }], closeOnBackdropClick: [{ type: i0.Input, args: [{ isSignal: true, alias: "closeOnBackdropClick", required: false }] }], attachTo: [{ type: i0.Input, args: [{ isSignal: true, alias: "attachTo", required: false }] }], attachPositions: [{ type: i0.Input, args: [{ isSignal: true, alias: "attachPositions", required: false }] }], autoFocus: [{ type: i0.Input, args: [{ isSignal: true, alias: "autoFocus", required: false }] }], closeDelay: [{ type: i0.Input, args: [{ isSignal: true, alias: "closeDelay", required: false }] }], disableClose: [{ type: i0.Input, args: [{ isSignal: true, alias: "disableClose", required: false }] }], ariaDescribedBy: [{ type: i0.Input, args: [{ isSignal: true, alias: "aria-describedby", required: false }] }], ariaLabelledBy: [{ type: i0.Input, args: [{ isSignal: true, alias: "aria-labelledby", required: false }] }], ariaLabel: [{ type: i0.Input, args: [{ isSignal: true, alias: "aria-label", required: false }] }], ariaModal: [{ type: i0.Input, args: [{ isSignal: true, alias: "aria-modal", required: false }] }] } });

class BrnDialogClose {
    _brnDialogRef = inject(BrnDialogRef);
    delay = input(undefined, ...(ngDevMode ? [{ debugName: "delay", transform: coerceNumberProperty }] : [{ transform: coerceNumberProperty }]));
    close() {
        this._brnDialogRef.close(undefined, this.delay());
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnDialogClose, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.3.17", type: BrnDialogClose, isStandalone: true, selector: "button[brnDialogClose]", inputs: { delay: { classPropertyName: "delay", publicName: "delay", isSignal: true, isRequired: false, transformFunction: null } }, host: { listeners: { "click": "close()" } }, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnDialogClose, decorators: [{
            type: Directive,
            args: [{
                    selector: 'button[brnDialogClose]',
                    host: {
                        '(click)': 'close()',
                    },
                }]
        }], propDecorators: { delay: [{ type: i0.Input, args: [{ isSignal: true, alias: "delay", required: false }] }] } });

class BrnDialogContent {
    _brnDialog = inject(BrnDialog, { optional: true });
    _brnDialogRef = inject(BrnDialogRef, { optional: true });
    _template = inject(TemplateRef);
    state = computed(() => this._brnDialog?.stateComputed() ?? this._brnDialogRef?.state() ?? 'closed', ...(ngDevMode ? [{ debugName: "state" }] : []));
    className = input(undefined, ...(ngDevMode ? [{ debugName: "className", alias: 'class' }] : [{ alias: 'class' }]));
    context = input(undefined, ...(ngDevMode ? [{ debugName: "context" }] : []));
    constructor() {
        if (!this._brnDialog)
            return;
        this._brnDialog.registerTemplate(this._template);
        effect(() => {
            const context = this.context();
            if (!this._brnDialog || !context)
                return;
            untracked(() => this._brnDialog?.setContext(context));
        });
        effect(() => {
            if (!this._brnDialog)
                return;
            const newClass = this.className();
            untracked(() => this._brnDialog?.setPanelClass(newClass));
        });
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnDialogContent, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.3.17", type: BrnDialogContent, isStandalone: true, selector: "[brnDialogContent]", inputs: { className: { classPropertyName: "className", publicName: "class", isSignal: true, isRequired: false, transformFunction: null }, context: { classPropertyName: "context", publicName: "context", isSignal: true, isRequired: false, transformFunction: null } }, providers: [provideExposesStateProviderExisting((() => BrnDialogContent))], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnDialogContent, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnDialogContent]',
                    providers: [provideExposesStateProviderExisting((() => BrnDialogContent))],
                }]
        }], ctorParameters: () => [], propDecorators: { className: [{ type: i0.Input, args: [{ isSignal: true, alias: "class", required: false }] }], context: [{ type: i0.Input, args: [{ isSignal: true, alias: "context", required: false }] }] } });

class BrnDialogDescription {
    _brnDialogRef = inject(BrnDialogRef);
    _id = signal(`brn-dialog-description-${this._brnDialogRef?.dialogId}`, ...(ngDevMode ? [{ debugName: "_id" }] : []));
    constructor() {
        effect(() => {
            this._brnDialogRef.setAriaDescribedBy(this._id());
        });
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnDialogDescription, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "20.3.17", type: BrnDialogDescription, isStandalone: true, selector: "[brnDialogDescription]", host: { properties: { "id": "_id()" } }, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnDialogDescription, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnDialogDescription]',
                    host: {
                        '[id]': '_id()',
                    },
                }]
        }], ctorParameters: () => [] });

class BrnDialogOverlay {
    _brnDialog = inject(BrnDialog);
    className = input(undefined, ...(ngDevMode ? [{ debugName: "className", alias: 'class' }] : [{ alias: 'class' }]));
    constructor() {
        effect(() => {
            if (!this._brnDialog)
                return;
            const newClass = this.className();
            untracked(() => this._brnDialog.setOverlayClass(newClass));
        });
    }
    setClassToCustomElement(newClass) {
        this._brnDialog.setOverlayClass(newClass);
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnDialogOverlay, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.3.17", type: BrnDialogOverlay, isStandalone: true, selector: "[brnDialogOverlay],brn-dialog-overlay", inputs: { className: { classPropertyName: "className", publicName: "class", isSignal: true, isRequired: false, transformFunction: null } }, providers: [provideCustomClassSettableExisting((() => BrnDialogOverlay))], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnDialogOverlay, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnDialogOverlay],brn-dialog-overlay',
                    providers: [provideCustomClassSettableExisting((() => BrnDialogOverlay))],
                }]
        }], ctorParameters: () => [], propDecorators: { className: [{ type: i0.Input, args: [{ isSignal: true, alias: "class", required: false }] }] } });

class BrnDialogTitle {
    _brnDialogRef = inject(BrnDialogRef);
    _id = signal(`brn-dialog-title-${this._brnDialogRef?.dialogId}`, ...(ngDevMode ? [{ debugName: "_id" }] : []));
    constructor() {
        effect(() => {
            this._brnDialogRef.setAriaLabelledBy(this._id());
        });
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnDialogTitle, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "20.3.17", type: BrnDialogTitle, isStandalone: true, selector: "[brnDialogTitle]", host: { properties: { "id": "_id()" } }, ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnDialogTitle, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnDialogTitle]',
                    host: {
                        '[id]': '_id()',
                    },
                }]
        }], ctorParameters: () => [] });

let idSequence = 0;
class BrnDialogTrigger {
    _brnDialog = inject(BrnDialog, { optional: true });
    _brnDialogRef = inject(BrnDialogRef, { optional: true });
    id = input(`brn-dialog-trigger-${++idSequence}`, ...(ngDevMode ? [{ debugName: "id" }] : []));
    type = input('button', ...(ngDevMode ? [{ debugName: "type" }] : []));
    state = computed(() => {
        // If we have a dialog component from the input, use its state (the highest priority)
        const dialogFromInput = this.brnDialogTriggerForState();
        if (dialogFromInput) {
            return dialogFromInput.stateComputed();
        }
        // If we have a dialog component, use its state
        if (this._brnDialog) {
            return this._brnDialog.stateComputed();
        }
        // If we have a dialog ref, use its state
        if (this._brnDialogRef) {
            return this._brnDialogRef.state();
        }
        return 'closed';
    }, ...(ngDevMode ? [{ debugName: "state" }] : []));
    dialogId = `brn-dialog-${this._brnDialogRef?.dialogId ?? ++idSequence}`;
    brnDialogTriggerFor = input(undefined, ...(ngDevMode ? [{ debugName: "brnDialogTriggerFor", alias: 'brnDialogTriggerFor' }] : [{
            alias: 'brnDialogTriggerFor',
        }]));
    mutableBrnDialogTriggerFor = computed(() => signal(this.brnDialogTriggerFor()), ...(ngDevMode ? [{ debugName: "mutableBrnDialogTriggerFor" }] : []));
    brnDialogTriggerForState = computed(() => this.mutableBrnDialogTriggerFor()(), ...(ngDevMode ? [{ debugName: "brnDialogTriggerForState" }] : []));
    constructor() {
        effect(() => {
            const brnDialog = this.brnDialogTriggerForState();
            if (!brnDialog)
                return;
            this._brnDialog = brnDialog;
        });
    }
    open() {
        this._brnDialog?.open();
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnDialogTrigger, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.1.0", version: "20.3.17", type: BrnDialogTrigger, isStandalone: true, selector: "button[brnDialogTrigger],button[brnDialogTriggerFor]", inputs: { id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null }, type: { classPropertyName: "type", publicName: "type", isSignal: true, isRequired: false, transformFunction: null }, brnDialogTriggerFor: { classPropertyName: "brnDialogTriggerFor", publicName: "brnDialogTriggerFor", isSignal: true, isRequired: false, transformFunction: null } }, host: { attributes: { "aria-haspopup": "dialog" }, listeners: { "click": "open()" }, properties: { "id": "id()", "attr.aria-expanded": "state() === 'open' ? 'true': 'false'", "attr.data-state": "state()", "attr.aria-controls": "dialogId", "type": "type()" } }, exportAs: ["brnDialogTrigger"], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnDialogTrigger, decorators: [{
            type: Directive,
            args: [{
                    selector: 'button[brnDialogTrigger],button[brnDialogTriggerFor]',
                    exportAs: 'brnDialogTrigger',
                    host: {
                        '[id]': 'id()',
                        '(click)': 'open()',
                        'aria-haspopup': 'dialog',
                        '[attr.aria-expanded]': "state() === 'open' ? 'true': 'false'",
                        '[attr.data-state]': 'state()',
                        '[attr.aria-controls]': 'dialogId',
                        '[type]': 'type()',
                    },
                }]
        }], ctorParameters: () => [], propDecorators: { id: [{ type: i0.Input, args: [{ isSignal: true, alias: "id", required: false }] }], type: [{ type: i0.Input, args: [{ isSignal: true, alias: "type", required: false }] }], brnDialogTriggerFor: [{ type: i0.Input, args: [{ isSignal: true, alias: "brnDialogTriggerFor", required: false }] }] } });

const BrnDialogImports = [
    BrnDialog,
    BrnDialogOverlay,
    BrnDialogTrigger,
    BrnDialogClose,
    BrnDialogContent,
    BrnDialogTitle,
    BrnDialogDescription,
];

/**
 * Generated bundle index. Do not edit.
 */

export { BrnDialog, BrnDialogClose, BrnDialogContent, BrnDialogDescription, BrnDialogImports, BrnDialogOverlay, BrnDialogRef, BrnDialogService, BrnDialogTitle, BrnDialogTrigger, cssClassesToArray, defaultOptions, injectBrnDialogContext, injectBrnDialogCtx, injectBrnDialogDefaultOptions, provideBrnDialogDefaultOptions };
