// @spartan-ng/brain/fesm2022/spartan-ng-brain-radio-group.mjs
import { FocusMonitor } from '@angular/cdk/a11y';
import * as i0 from '@angular/core';
import { InjectionToken, inject, ElementRef, input, booleanAttribute, computed, output, viewChild, ChangeDetectionStrategy, Component, forwardRef, model, linkedSignal, contentChildren, Directive } from '@angular/core';
import * as i1 from '@spartan-ng/brain/field';
import { BrnField, provideBrnLabelable, BrnFieldControl } from '@spartan-ng/brain/field';
import { Directionality } from '@angular/cdk/bidi';
import { NG_VALUE_ACCESSOR } from '@angular/forms';

const BrnRadioGroupToken = new InjectionToken('BrnRadioGroupToken');
function provideBrnRadioGroupToken(directive) {
    return { provide: BrnRadioGroupToken, useExisting: directive };
}
function injectBrnRadioGroup() {
    return inject(BrnRadioGroupToken);
}

class BrnRadioChange {
    source;
    value;
    constructor(source, value) {
        this.source = source;
        this.value = value;
    }
}
const CONTAINER_POST_FIX = '-radio';
const INPUT_POST_FIX = '-input';
class BrnRadio {
    static _nextUniqueId = 0;
    _focusMonitor = inject(FocusMonitor);
    _elementRef = inject(ElementRef);
    _field = inject(BrnField, { optional: true });
    _radioGroup = injectBrnRadioGroup();
    /**
     * Whether the radio button is disabled.
     */
    disabled = input(false, ...(ngDevMode ? [{ debugName: "disabled", transform: booleanAttribute }] : [{ transform: booleanAttribute }]));
    /**
     * Whether the radio button is disabled or the radio group is disabled.
     */
    _disabledState = computed(() => this.disabled() || this._radioGroup.disabledState(), ...(ngDevMode ? [{ debugName: "_disabledState" }] : []));
    /**
     * Whether the radio button is checked.
     */
    _checked = computed(() => this._radioGroup.value() === this.value(), ...(ngDevMode ? [{ debugName: "_checked" }] : []));
    _tabIndex = computed(() => {
        const disabled = this._disabledState();
        const checked = this._checked();
        const hasSelectedRadio = this._radioGroup.value() !== undefined;
        const isFirstRadio = this._radioGroup.radioButtons()[0] === this;
        if (disabled || (!checked && (hasSelectedRadio || !isFirstRadio))) {
            return -1;
        }
        return 0;
    }, ...(ngDevMode ? [{ debugName: "_tabIndex" }] : []));
    /**
     * The unique ID for the radio button input. If none is supplied, it will be auto-generated.
     */
    id = input(undefined, ...(ngDevMode ? [{ debugName: "id" }] : []));
    ariaLabel = input(undefined, ...(ngDevMode ? [{ debugName: "ariaLabel", alias: 'aria-label' }] : [{ alias: 'aria-label' }]));
    ariaLabelledby = input(undefined, ...(ngDevMode ? [{ debugName: "ariaLabelledby", alias: 'aria-labelledby' }] : [{ alias: 'aria-labelledby' }]));
    ariaDescribedby = input(undefined, ...(ngDevMode ? [{ debugName: "ariaDescribedby", alias: 'aria-describedby' }] : [{ alias: 'aria-describedby' }]));
    /**
     * The value this radio button represents.
     */
    value = input.required(...(ngDevMode ? [{ debugName: "value" }] : []));
    /**
     * Whether the radio button is required.
     */
    required = input(false, ...(ngDevMode ? [{ debugName: "required", transform: booleanAttribute }] : [{
            transform: booleanAttribute,
        }]));
    /**
     * Event emitted when the checked state of this radio button changes.
     */
    change = output();
    _hostId = computed(() => {
        return this.id() ? this.id() + CONTAINER_POST_FIX : `brn-radio-${++BrnRadio._nextUniqueId}`;
    }, ...(ngDevMode ? [{ debugName: "_hostId" }] : []));
    _inputId = computed(() => {
        return this.id() ?? `brn-radio${INPUT_POST_FIX}-${++BrnRadio._nextUniqueId}`;
    }, ...(ngDevMode ? [{ debugName: "_inputId" }] : []));
    _inputElement = viewChild.required('input');
    labelableId = this._inputId;
    constructor() {
        this._field?.registerLabelable(this);
        this._focusMonitor.monitor(this._elementRef, true);
    }
    ngOnDestroy() {
        this._focusMonitor.stopMonitoring(this._elementRef);
    }
    /** Dispatch change event with current value. */
    emitChangeEvent() {
        this.change.emit(new BrnRadioChange(this, this.value()));
    }
    onInputClick(event) {
        // We have to stop propagation for click events on the visual hidden input element.
        // By default, when a user clicks on a label element, a generated click event will be
        // dispatched on the associated input element. Since we are using a label element as our
        // root container, the click event on the `radio-button` will be executed twice.
        // The real click event will bubble up, and the generated click event also tries to bubble up.
        // This will lead to multiple click events.
        // Preventing bubbling for the second event will solve that issue.
        event.stopPropagation();
    }
    onInputInteraction(event) {
        // We always have to stop propagation on the change event.
        // Otherwise the change event, from the input element, will bubble up and
        // emit its event object to the `change` output.
        event.stopPropagation();
        if (!this._checked() && !this._disabledState()) {
            this.emitChangeEvent();
            this._radioGroup.select(this, this.value());
        }
    }
    /** Triggered when the user clicks on the touch target. */
    onTouchTargetClick(event) {
        this.onInputInteraction(event);
        if (!this._disabledState()) {
            // Normally the input should be focused already, but if the click
            // comes from the touch target, then we might have to focus it ourselves.
            this._inputElement().nativeElement.focus();
        }
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnRadio, deps: [], target: i0.ɵɵFactoryTarget.Component });
    /** @nocollapse */ static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "17.2.0", version: "20.3.17", type: BrnRadio, isStandalone: true, selector: "brn-radio", inputs: { disabled: { classPropertyName: "disabled", publicName: "disabled", isSignal: true, isRequired: false, transformFunction: null }, id: { classPropertyName: "id", publicName: "id", isSignal: true, isRequired: false, transformFunction: null }, ariaLabel: { classPropertyName: "ariaLabel", publicName: "aria-label", isSignal: true, isRequired: false, transformFunction: null }, ariaLabelledby: { classPropertyName: "ariaLabelledby", publicName: "aria-labelledby", isSignal: true, isRequired: false, transformFunction: null }, ariaDescribedby: { classPropertyName: "ariaDescribedby", publicName: "aria-describedby", isSignal: true, isRequired: false, transformFunction: null }, value: { classPropertyName: "value", publicName: "value", isSignal: true, isRequired: true, transformFunction: null }, required: { classPropertyName: "required", publicName: "required", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { change: "change" }, host: { listeners: { "focus": "_inputElement().nativeElement.focus()" }, properties: { "attr.id": "_hostId()", "class.brn-radio-checked": "_checked()", "class.brn-radio-disabled": "_disabledState()", "attr.data-checked": "_checked()", "attr.data-disabled": "_disabledState()", "attr.data-value": "value()", "attr.tabindex": "null", "attr.aria-label": "null", "attr.aria-labelledby": "null", "attr.aria-describedby": "null" }, classAttribute: "brn-radio" }, providers: [provideBrnLabelable(BrnRadio)], viewQueries: [{ propertyName: "_inputElement", first: true, predicate: ["input"], descendants: true, isSignal: true }], exportAs: ["brnRadio"], ngImport: i0, template: `
		<div data-slot="indicator" class="flex h-fit w-fit empty:hidden" (click)="onTouchTargetClick($event)">
			<ng-content select="[target],[indicator]" />
		</div>
		<input
			#input
			style="position: absolute; width: 1px; height: 1px; padding: 0; margin: -1px; overflow: hidden; clip: rect(0, 0, 0, 0); white-space: nowrap; border-width: 0;"
			type="radio"
			[attr.id]="_inputId()"
			[checked]="_checked()"
			[disabled]="_disabledState()"
			[tabIndex]="_tabIndex()"
			[attr.name]="_radioGroup.name()"
			[attr.value]="value()"
			[required]="required()"
			[attr.aria-checked]="_checked()"
			[attr.aria-label]="ariaLabel()"
			[attr.aria-labelledby]="ariaLabelledby()"
			[attr.aria-describedby]="ariaDescribedby()"
			[attr.aria-disabled]="_disabledState()"
			(change)="onInputInteraction($event)"
			(click)="onInputClick($event)"
		/>
		<ng-content />
	`, isInline: true, changeDetection: i0.ChangeDetectionStrategy.OnPush });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnRadio, decorators: [{
            type: Component,
            args: [{
                    selector: 'brn-radio',
                    exportAs: 'brnRadio',
                    providers: [provideBrnLabelable(BrnRadio)],
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    host: {
                        class: 'brn-radio',
                        '[attr.id]': '_hostId()',
                        '[class.brn-radio-checked]': '_checked()',
                        '[class.brn-radio-disabled]': '_disabledState()',
                        '[attr.data-checked]': '_checked()',
                        '[attr.data-disabled]': '_disabledState()',
                        '[attr.data-value]': 'value()',
                        // Needs to be removed since it causes some a11y issues (see #21266).
                        '[attr.tabindex]': 'null',
                        '[attr.aria-label]': 'null',
                        '[attr.aria-labelledby]': 'null',
                        '[attr.aria-describedby]': 'null',
                        // Note: under normal conditions focus shouldn't land on this element, however it may be
                        // programmatically set, for example inside of a focus trap, in this case we want to forward
                        // the focus to the native element.
                        '(focus)': '_inputElement().nativeElement.focus()',
                    },
                    template: `
		<div data-slot="indicator" class="flex h-fit w-fit empty:hidden" (click)="onTouchTargetClick($event)">
			<ng-content select="[target],[indicator]" />
		</div>
		<input
			#input
			style="position: absolute; width: 1px; height: 1px; padding: 0; margin: -1px; overflow: hidden; clip: rect(0, 0, 0, 0); white-space: nowrap; border-width: 0;"
			type="radio"
			[attr.id]="_inputId()"
			[checked]="_checked()"
			[disabled]="_disabledState()"
			[tabIndex]="_tabIndex()"
			[attr.name]="_radioGroup.name()"
			[attr.value]="value()"
			[required]="required()"
			[attr.aria-checked]="_checked()"
			[attr.aria-label]="ariaLabel()"
			[attr.aria-labelledby]="ariaLabelledby()"
			[attr.aria-describedby]="ariaDescribedby()"
			[attr.aria-disabled]="_disabledState()"
			(change)="onInputInteraction($event)"
			(click)="onInputClick($event)"
		/>
		<ng-content />
	`,
                }]
        }], ctorParameters: () => [], propDecorators: { disabled: [{ type: i0.Input, args: [{ isSignal: true, alias: "disabled", required: false }] }], id: [{ type: i0.Input, args: [{ isSignal: true, alias: "id", required: false }] }], ariaLabel: [{ type: i0.Input, args: [{ isSignal: true, alias: "aria-label", required: false }] }], ariaLabelledby: [{ type: i0.Input, args: [{ isSignal: true, alias: "aria-labelledby", required: false }] }], ariaDescribedby: [{ type: i0.Input, args: [{ isSignal: true, alias: "aria-describedby", required: false }] }], value: [{ type: i0.Input, args: [{ isSignal: true, alias: "value", required: true }] }], required: [{ type: i0.Input, args: [{ isSignal: true, alias: "required", required: false }] }], change: [{ type: i0.Output, args: ["change"] }], _inputElement: [{ type: i0.ViewChild, args: ['input', { isSignal: true }] }] } });

/* eslint-disable @typescript-eslint/no-empty-function */
const BRN_RADIO_GROUP_CONTROL_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(() => BrnRadioGroup),
    multi: true,
};
class BrnRadioGroup {
    _dir = inject(Directionality);
    _fieldControl = inject(BrnFieldControl, { optional: true });
    static _nextUniqueId = 0;
    onChange = () => { };
    onTouched = () => { };
    name = input(`brn-radio-group-${++BrnRadioGroup._nextUniqueId}`, ...(ngDevMode ? [{ debugName: "name" }] : []));
    /**
     * The value of the selected radio button.
     */
    value = model(...(ngDevMode ? [undefined, { debugName: "value" }] : []));
    /** Emits when the value changes. */
    valueChange = output();
    /**
     * Whether the radio group is disabled.
     */
    disabled = input(false, ...(ngDevMode ? [{ debugName: "disabled", transform: booleanAttribute }] : [{
            transform: booleanAttribute,
        }]));
    /**
     * Whether the radio group should be required.
     */
    required = input(false, ...(ngDevMode ? [{ debugName: "required", transform: booleanAttribute }] : [{
            transform: booleanAttribute,
        }]));
    /**
     * The direction of the radio group.
     */
    direction = this._dir.valueSignal;
    /**
     * Event emitted when the group value changes.
     */
    change = output();
    /**
     * The internal disabled state of the radio group.
     * @internal
     */
    disabledState = linkedSignal(() => this.disabled(), ...(ngDevMode ? [{ debugName: "disabledState" }] : []));
    controlState = this._fieldControl?.controlState;
    _ariaInvalid = computed(() => this._fieldControl?.invalid(), ...(ngDevMode ? [{ debugName: "_ariaInvalid" }] : []));
    _spartanInvalid = computed(() => this._fieldControl?.spartanInvalid(), ...(ngDevMode ? [{ debugName: "_spartanInvalid" }] : []));
    _dirty = computed(() => this._fieldControl?.dirty(), ...(ngDevMode ? [{ debugName: "_dirty" }] : []));
    _touched = computed(() => this._fieldControl?.touched(), ...(ngDevMode ? [{ debugName: "_touched" }] : []));
    /**
     * Access the radio buttons within the group.
     * @internal
     */
    radioButtons = contentChildren(BrnRadio, ...(ngDevMode ? [{ debugName: "radioButtons", descendants: true }] : [{ descendants: true }]));
    writeValue(value) {
        this.value.set(value);
    }
    registerOnChange(fn) {
        this.onChange = fn;
    }
    registerOnTouched(fn) {
        this.onTouched = fn;
    }
    setDisabledState(isDisabled) {
        this.disabledState.set(isDisabled);
    }
    /**
     * Select a radio button.
     * @internal
     */
    select(radioButton, value) {
        if (this.value() === value) {
            return;
        }
        this.value.set(value);
        this.valueChange.emit(value);
        this.onChange(value);
        this.change.emit(new BrnRadioChange(radioButton, value));
    }
    /** @nocollapse */ static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnRadioGroup, deps: [], target: i0.ɵɵFactoryTarget.Directive });
    /** @nocollapse */ static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "17.2.0", version: "20.3.17", type: BrnRadioGroup, isStandalone: true, selector: "[brnRadioGroup]", inputs: { name: { classPropertyName: "name", publicName: "name", isSignal: true, isRequired: false, transformFunction: null }, value: { classPropertyName: "value", publicName: "value", isSignal: true, isRequired: false, transformFunction: null }, disabled: { classPropertyName: "disabled", publicName: "disabled", isSignal: true, isRequired: false, transformFunction: null }, required: { classPropertyName: "required", publicName: "required", isSignal: true, isRequired: false, transformFunction: null } }, outputs: { value: "valueChange", valueChange: "valueChange", change: "change" }, host: { attributes: { "role": "radiogroup" }, listeners: { "focusout": "onTouched()" }, properties: { "dir": "direction()", "attr.aria-invalid": "_ariaInvalid() ? \"true\" : null", "attr.data-invalid": "_ariaInvalid() ? \"true\" : null", "attr.data-dirty": "_dirty() ? \"true\" : null", "attr.data-touched": "_touched() ? \"true\" : null", "attr.data-matches-spartan-invalid": "_spartanInvalid() ? \"true\" : null" } }, providers: [BRN_RADIO_GROUP_CONTROL_VALUE_ACCESSOR, provideBrnRadioGroupToken(BrnRadioGroup)], queries: [{ propertyName: "radioButtons", predicate: BrnRadio, descendants: true, isSignal: true }], hostDirectives: [{ directive: i1.BrnFieldControl }], ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "20.3.17", ngImport: i0, type: BrnRadioGroup, decorators: [{
            type: Directive,
            args: [{
                    selector: '[brnRadioGroup]',
                    providers: [BRN_RADIO_GROUP_CONTROL_VALUE_ACCESSOR, provideBrnRadioGroupToken(BrnRadioGroup)],
                    hostDirectives: [BrnFieldControl],
                    host: {
                        role: 'radiogroup',
                        '[dir]': 'direction()',
                        '(focusout)': 'onTouched()',
                        '[attr.aria-invalid]': '_ariaInvalid() ? "true" : null',
                        '[attr.data-invalid]': '_ariaInvalid() ? "true" : null',
                        '[attr.data-dirty]': '_dirty() ? "true" : null',
                        '[attr.data-touched]': '_touched() ? "true" : null',
                        '[attr.data-matches-spartan-invalid]': '_spartanInvalid() ? "true" : null',
                    },
                }]
        }], propDecorators: { name: [{ type: i0.Input, args: [{ isSignal: true, alias: "name", required: false }] }], value: [{ type: i0.Input, args: [{ isSignal: true, alias: "value", required: false }] }, { type: i0.Output, args: ["valueChange"] }], valueChange: [{ type: i0.Output, args: ["valueChange"] }], disabled: [{ type: i0.Input, args: [{ isSignal: true, alias: "disabled", required: false }] }], required: [{ type: i0.Input, args: [{ isSignal: true, alias: "required", required: false }] }], change: [{ type: i0.Output, args: ["change"] }], radioButtons: [{ type: i0.ContentChildren, args: [i0.forwardRef(() => BrnRadio), { ...{ descendants: true }, isSignal: true }] }] } });

const BrnRadioGroupImports = [BrnRadioGroup, BrnRadio];

/**
 * Generated bundle index. Do not edit.
 */

export { BRN_RADIO_GROUP_CONTROL_VALUE_ACCESSOR, BrnRadio, BrnRadioChange, BrnRadioGroup, BrnRadioGroupImports, injectBrnRadioGroup, provideBrnRadioGroupToken };
