// @spartan-ng/cli/src/generators/theme/libs/add-theme-to-application-styles.js
"use strict";
// All credit goes to the incredible folks at Nx who use this code to update the app styles when adding tailwind
// Check out the code here: https://github.com/nrwl/nx/blob/master/packages/angular/src/generators/setup-tailwind/lib/update-application-styles.ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.addThemeToApplicationStyles = addThemeToApplicationStyles;
const devkit_1 = require("@nx/devkit");
const enquirer_1 = require("enquirer");
const get_tailwind_version_1 = require("../../../utils/get-tailwind-version");
const colors_1 = require("./colors");
async function addThemeToApplicationStyles(tree, options, project) {
    const tailwindVersion = (0, get_tailwind_version_1.getTailwindVersion)(tree);
    if (tailwindVersion === 3) {
        await (0, enquirer_1.prompt)({
            type: 'confirm',
            name: 'Tailwind 3 detected',
            initial: true,
            message: `Please note that we cannot guarantee full compatibility of the components with Tailwind CSS version 3, and some features may not function as expected. Are you sure you want to proceed with Tailwind CSS v3?`,
        });
    }
    const spartantTailwindPresetImport = tailwindVersion === 4 ? '@import "@spartan-ng/brain/hlm-tailwind-preset.css";' : '';
    const prefix = options.prefix ? ` .${options.prefix}` : '';
    let stylesEntryPoint = options.stylesEntryPoint;
    if (stylesEntryPoint && !tree.exists(stylesEntryPoint)) {
        throw new Error(`The provided styles entry point "${stylesEntryPoint}" could not be found.`);
    }
    if (!stylesEntryPoint) {
        stylesEntryPoint = findStylesEntryPoint(tree, project);
        if (!stylesEntryPoint) {
            throw new Error((0, devkit_1.stripIndents) `Could not find a styles entry point for project "${options.project}".
        Please specify a styles entry point using the "stylesEntryPoint" option.`);
        }
    }
    const stylesEntryPointContent = tree.read(stylesEntryPoint, 'utf-8');
    const CDK_IMPORT = `@import "@angular/cdk/overlay-prebuilt.css";`;
    const ckdOverlayImport = tailwindVersion === 4 ? '' : CDK_IMPORT;
    const twSetup = options.setupTailwindCss
        ? `@layer theme, base, components, utilities;
@import 'tailwindcss/theme.css' layer(theme);
@import 'tailwindcss/preflight.css' layer(base);
@import 'tailwindcss/utilities.css';
		`
        : '';
    const fontSans = stylesEntryPointContent.includes('--font-sans')
        ? ''
        : `--font-sans: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji";`;
    const colorScheme = tailwindVersion === 4 ? colors_1.themes[options.theme] : colors_1.legacyThemes[options.theme];
    let stylesCssContentWithPotentialTailwindCssImport = (0, devkit_1.stripIndents) `
		${twSetup}
    ${ckdOverlayImport}

    ${stylesEntryPointContent}
    ${spartantTailwindPresetImport}

		:root${prefix} {
			color-scheme: light;

			${fontSans}

			${Object.entries(colorScheme.light)
        .map(([key, value]) => `--${key}: ${value};`)
        .join('\n')}
		}

		:root.dark${prefix} {
			color-scheme: dark;

			${Object.entries(colorScheme.dark)
        .map(([key, value]) => `--${key}: ${value};`)
        .join('\n')}
		}

    @layer base {
      * {
    		@apply border-border outline-ring/50;
 			}
  		body {
  		  @apply bg-background text-foreground;
  		}
    }`;
    if (options.setupTailwindCss) {
        stylesCssContentWithPotentialTailwindCssImport = stylesCssContentWithPotentialTailwindCssImport.replace(/^\s*@import\s+['"]tailwindcss['"]\s*;\s*$/gm, '');
    }
    tree.write(stylesEntryPoint, stylesCssContentWithPotentialTailwindCssImport);
}
function findStylesEntryPoint(tree, project) {
    // first check for common names
    const possibleStylesEntryPoints = [
        (0, devkit_1.joinPathFragments)(project.sourceRoot ?? project.root, 'styles.css'),
        (0, devkit_1.joinPathFragments)(project.sourceRoot ?? project.root, 'styles.scss'),
        (0, devkit_1.joinPathFragments)(project.sourceRoot ?? project.root, 'styles.sass'),
        (0, devkit_1.joinPathFragments)(project.sourceRoot ?? project.root, 'styles.less'),
    ];
    const stylesEntryPoint = possibleStylesEntryPoints.find((s) => tree.exists(s));
    if (stylesEntryPoint) {
        return stylesEntryPoint;
    }
    // then check for the specified styles in the build configuration if it exists
    const styles = project.targets?.build.options?.styles;
    if (!styles) {
        return undefined;
    }
    // find the first style that belongs to the project source
    const style = styles.find((s) => typeof s === 'string'
        ? s.startsWith(project.root) && tree.exists(s)
        : s.input.startsWith(project.root) && s.inject !== false && tree.exists(s.input));
    if (!style) {
        return undefined;
    }
    return typeof style === 'string' ? style : style.input;
}
