// @spartan-ng/cli/src/generators/theme/generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = addThemeToApplicationGenerator;
const enquirer_1 = require("enquirer");
const get_project_names_1 = require("../../utils/get-project-names");
const add_theme_to_application_styles_1 = require("./libs/add-theme-to-application-styles");
const colors_1 = require("./libs/colors");
async function addThemeToApplicationGenerator(tree, setupTailwindCss = false) {
    const { projects, projectNames } = (0, get_project_names_1.getProjectsAndNames)(tree);
    const response = await (0, enquirer_1.prompt)({
        type: 'select',
        required: true,
        name: 'app',
        message: 'Choose which application you want to add the theme to:',
        choices: projectNames,
    });
    const project = projects.get(response.app);
    if (!project)
        return;
    const themeOptions = await (0, enquirer_1.prompt)([
        {
            type: 'select',
            required: true,
            name: 'theme',
            message: 'Choose which theme to apply. You can always re-run this generator and add a custom prefix to add other themes.',
            choices: colors_1.themeNames,
        },
        {
            type: 'input',
            name: 'stylesEntryPoint',
            message: "Path to the styles entry point relative to the workspace root. If not provided the generator will do its best to find it and it will error if it can't.",
        },
        {
            type: 'input',
            name: 'prefix',
            message: "Prefix class name applied to your theme's style definitions: e.g., theme-zinc. Leave empty for global theme.",
        },
    ]);
    await (0, add_theme_to_application_styles_1.addThemeToApplicationStyles)(tree, {
        setupTailwindCss,
        project: project.name,
        theme: themeOptions.theme,
        addCdkStyles: themeOptions.addCdkStyles,
        stylesEntryPoint: themeOptions.stylesEntryPoint,
        prefix: themeOptions.prefix,
    }, project);
}
