// @spartan-ng/cli/src/registry/schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.configJsonSchema = exports.presetSchema = exports.registriesSchema = exports.searchResultItemSchema = exports.iconsSchema = exports.stylesSchema = exports.registryIndexSchema = exports.registrySchema = exports.registryItemSchema = exports.registryItemCommonSchema = exports.registryItemFontSchema = exports.registryItemEnvVarsSchema = exports.registryItemCssSchema = exports.registryItemCssVarsSchema = exports.registryItemTailwindSchema = exports.registryItemFileSchema = exports.registryItemTypeSchema = exports.workspaceConfigSchema = exports.configSchema = exports.rawConfigSchema = exports.registryConfigSchema = exports.registryConfigItemSchema = void 0;
const zod_1 = require("zod");
// Note: if you edit the schema here, you must also edit the schema in the
// apps/v4/public/schema/registry-item.json file.
exports.registryConfigItemSchema = zod_1.z.union([
    // Simple string format: "https://example.com/{name}.json"
    zod_1.z.string().refine((s) => s.includes('{name}'), {
        message: 'Registry URL must include {name} placeholder',
    }),
    // Advanced object format with auth options
    zod_1.z.object({
        url: zod_1.z.string().refine((s) => s.includes('{name}'), {
            message: 'Registry URL must include {name} placeholder',
        }),
        params: zod_1.z.record(zod_1.z.string(), zod_1.z.string()).optional(),
        headers: zod_1.z.record(zod_1.z.string(), zod_1.z.string()).optional(),
    }),
]);
exports.registryConfigSchema = zod_1.z.record(zod_1.z.string().refine((key) => key.startsWith('@'), {
    message: 'Registry names must start with @ (e.g., @v0, @acme)',
}), exports.registryConfigItemSchema);
exports.rawConfigSchema = zod_1.z
    .object({
    $schema: zod_1.z.string().optional(),
    style: zod_1.z.string(),
    tailwind: zod_1.z.object({
        css: zod_1.z.string(),
        baseColor: zod_1.z.string(),
        prefix: zod_1.z.string().default('').optional(),
    }),
    aliases: zod_1.z.object({
        ui: zod_1.z.string().optional(),
    }),
    registries: exports.registryConfigSchema.optional(),
})
    .strict();
exports.configSchema = exports.rawConfigSchema.extend({
    resolvedPaths: zod_1.z.object({
        cwd: zod_1.z.string(),
        tailwindConfig: zod_1.z.string(),
        tailwindCss: zod_1.z.string(),
        utils: zod_1.z.string(),
        components: zod_1.z.string(),
        lib: zod_1.z.string(),
        hooks: zod_1.z.string(),
        ui: zod_1.z.string(),
    }),
});
// TODO: type the key.
// Okay for now since I don't want a breaking change.
exports.workspaceConfigSchema = zod_1.z.record(exports.configSchema);
exports.registryItemTypeSchema = zod_1.z.enum([
    'registry:lib',
    'registry:block',
    'registry:component',
    'registry:ui',
    'registry:hook',
    'registry:page',
    'registry:file',
    'registry:theme',
    'registry:style',
    'registry:item',
    'registry:base',
    'registry:font',
    // Internal use only.
    'registry:example',
    'registry:internal',
]);
exports.registryItemFileSchema = zod_1.z.discriminatedUnion('type', [
    // Target is required for registry:file and registry:page
    zod_1.z.object({
        path: zod_1.z.string(),
        content: zod_1.z.string().optional(),
        type: zod_1.z.enum(['registry:file', 'registry:page']),
        target: zod_1.z.string(),
    }),
    zod_1.z.object({
        path: zod_1.z.string(),
        content: zod_1.z.string().optional(),
        type: exports.registryItemTypeSchema.exclude(['registry:file', 'registry:page']),
        target: zod_1.z.string().optional(),
    }),
]);
exports.registryItemTailwindSchema = zod_1.z.object({
    config: zod_1.z
        .object({
        content: zod_1.z.array(zod_1.z.string()).optional(),
        theme: zod_1.z.record(zod_1.z.string(), zod_1.z.any()).optional(),
        plugins: zod_1.z.array(zod_1.z.string()).optional(),
    })
        .optional(),
});
exports.registryItemCssVarsSchema = zod_1.z.object({
    theme: zod_1.z.record(zod_1.z.string(), zod_1.z.string()).optional(),
    light: zod_1.z.record(zod_1.z.string(), zod_1.z.string()).optional(),
    dark: zod_1.z.record(zod_1.z.string(), zod_1.z.string()).optional(),
});
// Recursive type for CSS properties that supports empty objects at any level.
const cssValueSchema = zod_1.z.lazy(() => zod_1.z.union([
    zod_1.z.string(),
    zod_1.z.array(zod_1.z.union([zod_1.z.string(), zod_1.z.record(zod_1.z.string(), zod_1.z.string())])),
    zod_1.z.record(zod_1.z.string(), cssValueSchema),
]));
exports.registryItemCssSchema = zod_1.z.record(zod_1.z.string(), cssValueSchema);
exports.registryItemEnvVarsSchema = zod_1.z.record(zod_1.z.string(), zod_1.z.string());
// Font metadata schema for registry:font items.
exports.registryItemFontSchema = zod_1.z.object({
    family: zod_1.z.string(),
    provider: zod_1.z.literal('google'),
    import: zod_1.z.string(),
    variable: zod_1.z.string(),
    weight: zod_1.z.array(zod_1.z.string()).optional(),
    subsets: zod_1.z.array(zod_1.z.string()).optional(),
});
// Common fields shared by all registry items.
exports.registryItemCommonSchema = zod_1.z.object({
    name: zod_1.z.string(),
    title: zod_1.z.string().optional(),
    description: zod_1.z.string().optional(),
    dependencies: zod_1.z.array(zod_1.z.string()).optional(),
    devDependencies: zod_1.z.array(zod_1.z.string()).optional(),
    files: zod_1.z.array(exports.registryItemFileSchema).optional(),
    cssVars: exports.registryItemCssVarsSchema.optional(),
});
// registry:base has a config field, registry:font has a font field.
exports.registryItemSchema = zod_1.z.discriminatedUnion('type', [
    exports.registryItemCommonSchema.extend({
        type: zod_1.z.literal('registry:base'),
        config: exports.rawConfigSchema.deepPartial().optional(),
    }),
    exports.registryItemCommonSchema.extend({
        type: zod_1.z.literal('registry:font'),
        font: exports.registryItemFontSchema,
    }),
    exports.registryItemCommonSchema.extend({
        type: exports.registryItemTypeSchema.exclude(['registry:base', 'registry:font']),
    }),
]);
exports.registrySchema = zod_1.z.object({
    name: zod_1.z.string(),
    homepage: zod_1.z.string(),
    items: zod_1.z.array(exports.registryItemSchema),
});
exports.registryIndexSchema = zod_1.z.array(exports.registryItemSchema);
exports.stylesSchema = zod_1.z.array(zod_1.z.object({
    name: zod_1.z.string(),
    label: zod_1.z.string(),
}));
exports.iconsSchema = zod_1.z.record(zod_1.z.string(), zod_1.z.record(zod_1.z.string(), zod_1.z.string()));
exports.searchResultItemSchema = zod_1.z.object({
    name: zod_1.z.string(),
    type: zod_1.z.string().optional(),
    description: zod_1.z.string().optional(),
    registry: zod_1.z.string(),
    addCommandArgument: zod_1.z.string(),
});
// New schema for getRegistries().
exports.registriesSchema = zod_1.z.array(zod_1.z.object({
    name: zod_1.z.string(),
    homepage: zod_1.z.string().optional(),
    url: zod_1.z.string(),
    description: zod_1.z.string().optional(),
}));
exports.presetSchema = zod_1.z.object({
    name: zod_1.z.string(),
    title: zod_1.z.string(),
    description: zod_1.z.string(),
    base: zod_1.z.string(),
    style: zod_1.z.string(),
    baseColor: zod_1.z.string(),
    theme: zod_1.z.string(),
    iconLibrary: zod_1.z.string(),
    font: zod_1.z.string(),
    rtl: zod_1.z.coerce.boolean().default(false),
    menuAccent: zod_1.z.enum(['subtle', 'bold']),
    menuColor: zod_1.z.enum(['default', 'inverted']),
    radius: zod_1.z.string(),
});
exports.configJsonSchema = zod_1.z.object({
    presets: zod_1.z.array(exports.presetSchema),
});
