// @spartan-ng/cli/src/generators/base/lib/initialize-angular-library.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initializeAngularEntrypoint = initializeAngularEntrypoint;
exports.initializeAngularLibrary = initializeAngularLibrary;
const tslib_1 = require("tslib");
const generators_1 = require("@nx/angular/generators");
const devkit_1 = require("@nx/devkit");
const path = tslib_1.__importStar(require("path"));
const ts_morph_1 = require("ts-morph");
const single_lib_name_1 = require("./single-lib-name");
function addRulesToEslintOverride(tree, libRoot, filePattern, newRules) {
    const eslintPath = (0, devkit_1.joinPathFragments)(libRoot, 'eslint.config.mjs');
    if (!tree.exists(eslintPath))
        return;
    const eslintFullPath = path.join(process.cwd(), eslintPath);
    const project = new ts_morph_1.Project({
        tsConfigFilePath: path.join(process.cwd(), 'tsconfig.base.json'),
        skipAddingFilesFromTsConfig: true,
    });
    const sourceFile = project.createSourceFile(eslintFullPath, tree.read(eslintPath, 'utf-8'), { overwrite: true });
    const exportArray = sourceFile.getFirstDescendantByKind(ts_morph_1.SyntaxKind.ArrayLiteralExpression);
    if (!exportArray)
        return;
    const overrideObj = exportArray.getElements().find((el) => {
        if (!el || !el.isKind(ts_morph_1.SyntaxKind.ObjectLiteralExpression))
            return false;
        const obj = el;
        const filesProp = obj.getProperty('files');
        return filesProp?.getText().includes(filePattern);
    });
    if (!overrideObj)
        return;
    let rulesProp = overrideObj.getProperty('rules');
    if (!rulesProp) {
        overrideObj.addPropertyAssignment({ name: 'rules', initializer: '{}' });
        rulesProp = overrideObj.getPropertyOrThrow('rules');
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const rulesObj = rulesProp.getInitializerIfKindOrThrow(ts_morph_1.SyntaxKind.ObjectLiteralExpression);
    for (const [ruleName, ruleValue] of Object.entries(newRules)) {
        if (!rulesObj.getProperty(ruleName)) {
            rulesObj.addPropertyAssignment({
                name: ruleName,
                initializer: ruleValue,
            });
        }
    }
    tree.write(eslintPath, sourceFile.getFullText());
}
/**
 * Add TypeScript rules to ESLint flat config
 */
function addTsRules(tree, libRoot) {
    addRulesToEslintOverride(tree, libRoot, '"**/*.ts"', {
        '"@angular-eslint/no-input-rename"': '"off"',
        '"@nx/enforce-module-boundaries"': `
		(() => {
  const r = baseConfig.find(c => c.rules && c.rules["@nx/enforce-module-boundaries"])?.rules["@nx/enforce-module-boundaries"];
  return r ? [r[0], { ...r[1], allowCircularSelfDependency: true }] : undefined;
})()
		`,
        '"@angular-eslint/directive-class-suffix"': '"off"',
        '"@angular-eslint/component-class-suffix"': '"off"',
        '"@typescript-eslint/naming-convention"': `[
			"error",
			{
				"selector": "classProperty",
				"modifiers": ["protected"],
				"format": ["camelCase"],
				"leadingUnderscore": "require"
			}
		]`,
    });
}
/**
 * Add HTML template rules to ESLint flat config
 */
function addHtmlRules(tree, libRoot) {
    addRulesToEslintOverride(tree, libRoot, '"**/*.html"', {
        '"@angular-eslint/template/interactive-supports-focus"': '"off"',
        '"@angular-eslint/template/click-events-have-key-events"': '"off"',
    });
}
function addRules(tree, libRoot) {
    addTsRules(tree, libRoot);
    addHtmlRules(tree, libRoot);
}
function cleanupSingleLibFolder(tree, dir) {
    // delete everything in the src directory except the index.ts file
    const srcDir = (0, devkit_1.joinPathFragments)(dir, 'src');
    const files = tree.children(srcDir).filter((file) => file !== 'index.ts');
    for (const file of files) {
        const filePath = (0, devkit_1.joinPathFragments)(srcDir, file);
        if (tree.exists(filePath)) {
            tree.delete(filePath);
        }
    }
    // replace the content of the index.ts with "export default null;"
    const indexPath = (0, devkit_1.joinPathFragments)(srcDir, 'index.ts');
    if (tree.exists(indexPath)) {
        tree.write(indexPath, '// This file is intentionally left empty. Import from the respective entrypoint.\nexport default null;');
    }
}
const defaultSchema = {
    skipFormat: true,
    simpleName: true,
    prefix: 'hlm',
    skipModule: true,
    skipTests: true,
    strict: true,
    unitTestRunner: generators_1.UnitTestRunner.None,
};
async function initializeAngularEntrypoint(tree, options) {
    const { libraryGenerator } = await Promise.resolve().then(() => tslib_1.__importStar(require('@nx/angular/generators')));
    const dir = (0, devkit_1.joinPathFragments)(options.directory);
    const callback = await libraryGenerator(tree, {
        ...((0, devkit_1.readNxJson)(tree).generators?.['@nx/angular:library'] || {}),
        ...defaultSchema,
        flat: true,
        name: single_lib_name_1.singleLibName,
        buildable: options.buildable,
        importPath: options.importAlias,
        directory: dir,
        tags: options.tags,
    });
    cleanupSingleLibFolder(tree, dir);
    addRules(tree, dir);
    (0, devkit_1.updateJson)(tree, (0, devkit_1.joinPathFragments)(dir, 'tsconfig.lib.json'), (json) => {
        const include = json.include || [];
        if (!include.includes('**/*.ts')) {
            include.push('**/*.ts');
        }
        json.include = include;
        return json;
    });
    return callback;
}
async function initializeAngularLibrary(tree, options) {
    const { libraryGenerator } = await Promise.resolve().then(() => tslib_1.__importStar(require('@nx/angular/generators')));
    const dir = (0, devkit_1.joinPathFragments)(options.directory, options.name);
    const callback = await libraryGenerator(tree, {
        ...((0, devkit_1.readNxJson)(tree).generators?.['@nx/angular:library'] || {}),
        ...defaultSchema,
        name: options.name,
        buildable: options.buildable,
        importPath: `${options.importAlias}/${options.name}`,
        directory: dir,
        tags: options.tags,
    });
    addRules(tree, dir);
    return callback;
}
