// @spartan-ng/cli/src/generators/migrate-brain-accordion-trigger/generator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = migrateAccordionTriggerGenerator;
const devkit_1 = require("@nx/devkit");
const node_html_parser_1 = require("node-html-parser");
const ts_morph_1 = require("ts-morph");
const visit_files_1 = require("../../utils/visit-files");
const html_utils_1 = require("./utils/html-utils");
const shared_utils_1 = require("./utils/shared-utils");
async function migrateAccordionTriggerGenerator(tree, { skipFormat }) {
    const issues = [];
    const project = new ts_morph_1.Project({ useInMemoryFileSystem: true });
    wrapAccordionTriggers(tree, project, issues);
    if (issues.length > 0) {
        reportManualFixesNeeded(issues);
    }
    if (!skipFormat) {
        await (0, devkit_1.formatFiles)(tree);
    }
}
function wrapAccordionTriggers(tree, project, issues) {
    (0, visit_files_1.visitFiles)(tree, '.', (path) => {
        console.log('Processing ' + path + '...');
        if (!(0, shared_utils_1.shouldProcessFile)(path))
            return;
        const content = tree.read(path, 'utf-8');
        if (!content || !(0, shared_utils_1.hasAccordionTriggers)(content))
            return;
        const result = path.endsWith('.html')
            ? fixTemplate(content, path, issues)
            : fixComponentFile(content, path, project, issues);
        if (result?.modified) {
            tree.write(path, result.content);
            console.log(`Fixed ${result.fixedCount} accordion trigger(s) in ${path}`);
        }
    });
}
function fixTemplate(content, filePath, issues, options = { noContentsClass: false }) {
    const root = (0, html_utils_1.parseHtmlForAccordionMigration)(content);
    const buttons = (0, html_utils_1.findAccordionButtons)(root);
    let fixedCount = 0;
    for (const button of buttons) {
        const analysis = (0, shared_utils_1.analyzeButton)(button);
        switch (analysis.issue) {
            case 'NotInHeading':
                (0, html_utils_1.wrapButtonInHeading)(button, !options.noContentsClass);
                fixedCount++;
                break;
            case 'HasSiblings':
                logSiblingIssue(analysis, filePath, issues);
                break;
            case 'NoIssue':
                break;
        }
    }
    if (fixedCount > 0) {
        return {
            content: formatHtmlOutput(root),
            modified: true,
            fixedCount,
        };
    }
    return { content, modified: false, fixedCount: 0 };
}
function fixComponentFile(content, filePath, project, issues) {
    const sourceFile = project.createSourceFile(filePath, content, { overwrite: true });
    const templateInfo = (0, shared_utils_1.extractInlineTemplate)(sourceFile);
    if (!templateInfo) {
        return { content, modified: false, fixedCount: 0 };
    }
    // Fix the template content
    const result = fixTemplate(templateInfo.text, filePath, issues);
    if (result.modified && templateInfo.propertyAssignment.getInitializer()) {
        // Replace template in source file
        const quoteStyle = templateInfo.originalQuoteStyle || '`';
        let newTemplateValue;
        if (quoteStyle === '`') {
            newTemplateValue = `\`${result.content}\``;
        }
        else if (quoteStyle === '"') {
            newTemplateValue = `"${result.content.replace(/"/g, '\\"')}"`;
        }
        else {
            newTemplateValue = `'${result.content.replace(/'/g, "\\'")}'`;
        }
        templateInfo.propertyAssignment.getInitializer().replaceWithText(newTemplateValue);
        return {
            content: sourceFile.getFullText(),
            modified: true,
            fixedCount: result.fixedCount,
        };
    }
    return { content, modified: false, fixedCount: 0 };
}
function logSiblingIssue(analysis, filePath, issues) {
    const parent = analysis.parent;
    const siblingContent = getSiblingPreview(analysis.button, parent);
    issues.push(`! ${filePath}\n` +
        ` - Button "${analysis.buttonText}" in <${parent.tagName.toLowerCase()}> ` +
        `has siblings that need to be moved: ${siblingContent}...\n`);
}
function getSiblingPreview(button, parent) {
    return parent.childNodes
        .filter((child) => child !== button)
        .map((child) => {
        if (child.nodeType === node_html_parser_1.NodeType.TEXT_NODE) {
            return child.text?.trim();
        }
        return child.toString();
    })
        .filter(Boolean)
        .join(', ')
        .substring(0, 50);
}
function formatHtmlOutput(root) {
    const output = root.toString();
    // Convert empty tags to self-closing
    return output.replace(/<([^>\s]+)([^>]*)><\/\1>/g, '<$1$2/>');
}
function reportManualFixesNeeded(issues) {
    console.log(`We also detected ${issues.length} accordion trigger buttons ` +
        `that have siblings that need to be manually moved:`);
    issues.forEach((issue) => console.log(issue));
    console.log('To learn more about this pattern see: ' + 'https://www.w3.org/WAI/ARIA/apg/patterns/accordion/');
}
