// @spartan-ng/cli/src/generators/healthcheck/healthchecks/brn-accordion-trigger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.brainAccordionTriggerHealthcheck = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nx/devkit");
const ts_morph_1 = require("ts-morph");
const generator_1 = tslib_1.__importDefault(require("../../migrate-brain-accordion-trigger/generator"));
const html_utils_1 = require("../../migrate-brain-accordion-trigger/utils/html-utils");
const shared_utils_1 = require("../../migrate-brain-accordion-trigger/utils/shared-utils");
const healthchecks_1 = require("../healthchecks");
exports.brainAccordionTriggerHealthcheck = {
    name: 'Brain Accordion Trigger',
    async detect(tree, failure) {
        const project = new ts_morph_1.Project({ useInMemoryFileSystem: true });
        const violations = [];
        (0, devkit_1.visitNotIgnoredFiles)(tree, '/', (file) => {
            if (!(0, shared_utils_1.shouldProcessFile)(file))
                return;
            const contents = tree.read(file, 'utf-8');
            if (!contents || !(0, shared_utils_1.hasAccordionTriggers)(contents))
                return;
            const fileViolations = file.endsWith('.html')
                ? findTemplateViolations(contents, file)
                : findComponentViolations(contents, file, project);
            violations.push(...fileViolations);
        });
        if (violations.length > 0) {
            const summary = generateViolationSummary(violations);
            failure(summary, healthchecks_1.HealthcheckSeverity.Error, true);
        }
    },
    fix: async (tree) => {
        await (0, generator_1.default)(tree, { skipFormat: true });
        return true;
    },
    prompt: 'Would you like to migrate your brain accordion triggers?',
};
function findTemplateViolations(content, filePath) {
    const violations = [];
    const root = (0, html_utils_1.parseHtmlForAccordionMigration)(content);
    const buttons = (0, html_utils_1.findAccordionButtons)(root);
    for (const button of buttons) {
        const violation = checkButton(button, filePath);
        if (violation)
            violations.push(violation);
    }
    return violations;
}
function checkButton(button, filePath) {
    if (!(0, html_utils_1.isInHeading)(button)) {
        return {
            file: filePath,
            line: button.range?.[0] || 0,
            type: 'not-in-heading',
            buttonText: (0, html_utils_1.getButtonText)(button),
        };
    }
    if ((0, html_utils_1.hasSignificantSiblings)(button)) {
        return {
            file: filePath,
            line: button.range?.[0] || 0,
            type: 'not-only-child',
            buttonText: (0, html_utils_1.getButtonText)(button),
        };
    }
    return null;
}
function findComponentViolations(content, filePath, project) {
    const sourceFile = project.createSourceFile(filePath, content, { overwrite: true });
    const templateContent = (0, shared_utils_1.extractInlineTemplate)(sourceFile);
    if (!templateContent)
        return [];
    const templateViolations = findTemplateViolations(templateContent.text, filePath);
    for (const violation of templateViolations) {
        violation.line += templateContent.startLine - 1;
    }
    return templateViolations;
}
function generateViolationSummary(violations) {
    const notInHeading = violations.filter((v) => v.type === 'not-in-heading');
    const byFile = groupViolationsByFile(notInHeading);
    let summary = `Found ${notInHeading.length} accordion trigger(s) that need to be wrapped in headings:\n\n`;
    for (const [file, fileViolations] of byFile) {
        summary += `${file}\n`;
        summary += formatFileViolations(fileViolations);
    }
    return summary;
}
function groupViolationsByFile(violations) {
    const byFile = new Map();
    for (const violation of violations) {
        if (!byFile.has(violation.file)) {
            byFile.set(violation.file, []);
        }
        byFile.get(violation.file).push(violation);
    }
    return byFile;
}
function formatFileViolations(violations) {
    return violations
        .map((v) => {
        let line = ` - Line ${v.line}`;
        if (v.buttonText) {
            line += ` - (button: "${v.buttonText}")`;
        }
        return line + '\n';
    })
        .join('');
}
